<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Bailey Andrew">
<meta name="dcterms.date" content="2023-01-08">
<meta name="description" content="This was originally meant to be a whole-workflow blog post, but understanding dataset procurement is hard…">

<title>Lorem Ipsomething - Analyzing scRNA-seq Data – Dataset Acquisition</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../styles.css">
<meta property="og:title" content="Lorem Ipsomething - Analyzing scRNA-seq Data – Dataset Acquisition">
<meta property="og:description" content="This was originally meant to be a whole-workflow blog post, but understanding dataset procurement is hard…">
<meta property="og:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/cell-expression-atlas-species.png">
<meta property="og:site-name" content="Lorem Ipsomething">
<meta property="og:image:height" content="876">
<meta property="og:image:width" content="2780">
<meta name="twitter:title" content="Lorem Ipsomething - Analyzing scRNA-seq Data – Dataset Acquisition">
<meta name="twitter:description" content="This was originally meant to be a whole-workflow blog post, but understanding dataset procurement is hard…">
<meta name="twitter:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/cell-expression-atlas-species.png">
<meta name="twitter:image-height" content="876">
<meta name="twitter:image-width" content="2780">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Lorem Ipsomething</span>
    </a>
  </div>
        <div class="quarto-navbar-tools ms-auto">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Analyzing scRNA-seq Data – Dataset Acquisition</li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Home</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../blog.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Blog</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../staticposts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Static Posts</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Puzzles</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../500_Puzzles/bnv.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Boulders in Valleys</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slideshowindex.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slideshows</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../otherblogs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Other People’s Blogs</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#the-dataset" id="toc-the-dataset" class="nav-link active" data-scroll-target="#the-dataset">The Dataset</a>
  <ul class="collapse">
  <li><a href="#experimental-design" id="toc-experimental-design" class="nav-link" data-scroll-target="#experimental-design">Experimental Design</a>
  <ul class="collapse">
  <li><a href="#sample-selection" id="toc-sample-selection" class="nav-link" data-scroll-target="#sample-selection">Sample Selection</a></li>
  <li><a href="#tissue-preparation" id="toc-tissue-preparation" class="nav-link" data-scroll-target="#tissue-preparation">Tissue Preparation</a></li>
  <li><a href="#cell-isolation" id="toc-cell-isolation" class="nav-link" data-scroll-target="#cell-isolation">Cell Isolation</a></li>
  <li><a href="#library-preparation-and-sequencing" id="toc-library-preparation-and-sequencing" class="nav-link" data-scroll-target="#library-preparation-and-sequencing">Library Preparation and Sequencing</a></li>
  </ul></li>
  <li><a href="#acquiring-and-exploring-the-dataset" id="toc-acquiring-and-exploring-the-dataset" class="nav-link" data-scroll-target="#acquiring-and-exploring-the-dataset">Acquiring and Exploring the Dataset</a>
  <ul class="collapse">
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/BaileyAndrew/Blog/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Analyzing scRNA-seq Data – Dataset Acquisition</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Work</div>
    <div class="quarto-category">Useful</div>
    <div class="quarto-category">Omics</div>
    <div class="quarto-category">Paper Replication</div>
  </div>
  </div>

<div>
  <div class="description">
    This was originally meant to be a whole-workflow blog post, but understanding dataset procurement is hard…
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Bailey Andrew </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 8, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p>In this walkthrough, we’re going to start playing around with real scRNA-seq data. I had no end goal in mind when starting this post; let’s see what happens!</p>
<section id="the-dataset" class="level1">
<h1>The Dataset</h1>
<p>We’ll use the <a href="https://www.ebi.ac.uk/gxa/sc/home">Single Cell Expression Atlas</a> to find a suitable dataset to explore. The website has a nice design, and there’re a lot of species to choose from.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/cell-expression-atlas-species.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Species Options</figcaption>
</figure>
</div>
<p>I chose to look at <i style="color:#EB1960">Danio rerio</i> (<b style="color:#EB1960">Zebrafish</b>) because it was the animal with the most experiments that I hadn’t really heard about before<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>. I didn’t want to look at plants/fungi/protists because they might require additional considerations I’m not aware of (especially protists).</p>
<details>
<summary style="color:#C0CF95">
<b>&lt;i&gt; vs &lt;em&gt;</b>
</summary>
<p>
This is very tangential, but in trying to learn the right way to represent scientific names such as <i style="color:#EB1960">Danio rerio</i> I stumbled onto a debate about <span style="color:#757575">&lt;i&gt;</span> vs <span style="color:#757575">&lt;em&gt;</span> (html tags to represent italics), and analogously about <span style="color:#757575">&lt;b&gt;</span> vs <span style="color:#757575">&lt;strong&gt;</span> (to represent bolds). In my investigation into this debate, I’ve encountered that <span style="color:#757575">&lt;i&gt;</span>/<span style="color:#757575">&lt;b&gt;</span> are to be used when there is no semantic emphasis on the words, whereas <span style="color:#757575">&lt;em&gt;</span>/<span style="color:#757575">&lt;strong&gt;</span> contain semantic emphasis <span class="citation" data-cites="b-vs-strong">(<a href="#ref-b-vs-strong" role="doc-biblioref">Lucero 2008</a>)</span>. This unfortunately means I’ve been using them wrong 😅 - since I’ve just been <span style="color:#757575">&lt;strong&gt;</span>-ing all my bolds even though I basically never bold for emphasis these days. (I’m not a very emphatic person).
</p>
</details>
<p>I chose the dataset from the paper “Single-cell transcriptional analysis reveals innate lymphoid cell (ILC)-like cells in zebrafish” <span class="citation" data-cites="zebrafish-data">(<a href="#ref-zebrafish-data" role="doc-biblioref">Hernández et al. 2018</a>)</span>, because it was one of the most recent but also not exceedingly large (&lt;1000 cells).</p>
<section id="experimental-design" class="level2">
<h2 class="anchored" data-anchor-id="experimental-design">Experimental Design</h2>
<p>Before downloading the data, we want to check if it’s actually useful to us - is it mRNA, what type of cells are they, were specific genes targeted, etc. Since the previous blog posts were talking about the wetlab generation of the data, let’s do a deep dive into what they did. The paper <span class="citation" data-cites="zebrafish-data">(<a href="#ref-zebrafish-data" role="doc-biblioref">Hernández et al. 2018</a>)</span> is freely available online, and the pertinent information will be contained in the <b style="color:#EB1960">Materials and Methods</b> section. The paper is the source of two datasets on the <b style="color:#EB1960">Single Cell Expression Atlas</b> so we will have to keep that in mind when reading about the methods.</p>
<section id="sample-selection" class="level3">
<h3 class="anchored" data-anchor-id="sample-selection">Sample Selection</h3>
<blockquote class="blockquote">
<p>The aim of this study was to characterised innate and adaptive <strong style="color:#A6A440">lymphocytes</strong> in zebrafish in steady state and following the immune challenge, using scRNA-seq. <strong style="color:#C0CF95">Multiple zebrafish</strong>, either in <strong style="color:#A6A440">steady state</strong> or <strong style="color:#537FBF">exposed to immune challenge</strong>, were used to collect cells for sequencing.</p>
<p>– <cite><b style="color:#EB1960">Study Design Subsection; Materials and Methods Section;</b> <span class="citation" data-cites="zebrafish-data">Hernández et al. (<a href="#ref-zebrafish-data" role="doc-biblioref">2018</a>)</span> </cite></p>
</blockquote>
<p>A <b style="color:#A6A440">lymphocyte</b> is a specific type of cell that is part of your immune system; B, T, and NK cells are all <b style="color:#A6A440">lymphocytes</b>. We can also see that the experiment was done on multiple zebrafish, rather than just one - and that some were “exposed to immune challenge”, which I assume means they were made sick to try to trigger interesting processes in their immune system.</p>
<p>The paper goes into more depth into this immune challenge (it involves <i style="color:#EB1960">Vibrio anguillarum</i>, a “fish pathogen” <span class="citation" data-cites="vibrio-anguillarum">(details: <a href="#ref-vibrio-anguillarum" role="doc-biblioref">Frans et al. 2011</a>)</span>), however this does not seem to be relevant for our dataset. The paper seems to refer to our dataset as the <b style="color:#EB1960">Smart-seq2 experiment</b>, whereas the other dataset is the <b style="color:#EB1960">10x experiment</b>. The immune challenge was only applied to the <b style="color:#EB1960">10x experiment</b>. We can double-check this by comparing the sample characteristics of <a href="https://www.ebi.ac.uk/gxa/sc/experiments/E-MTAB-7117/experiment-design">our dataset</a> and <a href="https://www.ebi.ac.uk/gxa/sc/experiments/E-MTAB-7159/experiment-design">the other dataset</a> on the <b style="color:#EB1960">Single Cell Expression Atlas</b>, noting that <b style="color:#757575">infect</b> is not one of the experimental variables for our dataset.</p>
<p>A triple-check, if the above was not convincing enough<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>, can be found when we read the whole paper:</p>
<blockquote class="blockquote">
<p>To allow easy retrieval of sequencing data from zebrafish innate and adaptive lymphocytes we generated a cloud repository (https://www.sanger.ac.uk/science/tools/lymphocytes/ lymphocytes/) with transcriptional profiles of over 14,000 single cells collected from healthy and immune challenged zebrafish using 10x genomics and Smart-seq2 methodology <strong style="color:#C0CF96">(please see Explanatory Note in Supplementary Material)</strong>.</p>
<p>– <cite><b style="color:#EB1960">“Single cell atlas of innate and adaptive lymphocytes in zebrafish” section;</b> <span class="citation" data-cites="zebrafish-data">Hernández et al. (<a href="#ref-zebrafish-data" role="doc-biblioref">2018</a>)</span> </cite></p>
</blockquote>
<p>We can then track down this explanatory note<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a> to see what it has to say;</p>
<blockquote class="blockquote">
<p>Available datasets includes <strong style="color:#C0CF96"><b style="color:#EB1960">Smart-seq2 data</b> from kidney, thymus, spleen, guts and gills of healthy, unstimulated <b style="color:#A6A440">wild-type</b> zebrafish</strong> as well as <strong style="color:#C0CF96"><b style="color:#EB1960">10x datasets</b> from gut of unstimulated zebrafish both <b style="color:#A6A440">wild-type</b> and <b style="color:#A6A440">rag1<sup>-/-</sup> mutant</b> and <b style="color:#537FBF">immune-challenged</b> (V. anguillarum- or A. simplex-injected) <b style="color:#A6A440">rag1<sup>-/-</sup> mutant</b></strong>.</p>
<p>– <cite><b style="color:#EB1960">Supplementary Material; Explanatory Note</b> <span class="citation" data-cites="zebrafish-data">Hernández et al. (<a href="#ref-zebrafish-data" role="doc-biblioref">2018</a>)</span> </cite></p>
</blockquote>
<p>We know the <b style="color:#EB1960">10x experiment</b> contains over 10,000 cells, and experienced <b style="color:#537FBF">immune challenges</b>. Ours doesn’t. Our dataset should be from multiple body parts (corroborated by the sample characteristics of <a href="https://www.ebi.ac.uk/gxa/sc/experiments/E-MTAB-7117/experiment-design">our dataset</a>) and the <b style="color:#EB1960">10x dataset</b> should only be from the gut (corroborated by the sample characteristics of <a href="https://www.ebi.ac.uk/gxa/sc/experiments/E-MTAB-7159/experiment-design">the other dataset</a>).</p>
<p>The above quote also points out that all our zebrafish were <b style="color:#A6A440">wild-type</b>. Interestingly, the sample characteristics of <a href="https://www.ebi.ac.uk/gxa/sc/experiments/E-MTAB-7117/experiment-design">our dataset</a> do list multiple genotypes, including <b style="color:#A6A440">rag1<sup>-/-</sup> mutants</b>. The quote seems to imply that <b style="color:#A6A440">rag1<sup>-/-</sup></b> is not “<b style="color:#A6A440">wild-type</b>”, as does the rest of the paper. However, the rest of the paper does explicitly say that the <b style="color:#EB1960">Smart-seq2 data</b> contains <b style="color:#A6A440">rag1<sup>-/-</sup> mutants</b>, so I assume this was either an oversight by the authors or a minor misinterpretation on my part<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>. Either way, I’m satisfied with this understanding; we can move on to the tissue preparation.</p>
<details>
<summary style="color:#C0CF95">
<b>About rag1<sup>-/-</sup> mutants</b>
</summary>
<p>
Well, maybe not move on just yet; I’m just a bit curious as to what a <b style="color:#A6A440">rag1<sup>-/-</sup> mutant</b> is! According to <span class="citation" data-cites="medline-rag1"><span>“RAG1 Gene”</span> (<a href="#ref-medline-rag1" role="doc-biblioref">2017</a>)</span>, it’s related to <b style="color:#537FBF">VDJ-recombination</b> (something that was briefly mentioned, but not expanded on, in a <a href="./004_scRNA2.html">prior blogpost</a>)). It stands for “recombination activating gene 1”, and when it is absent it can really mess up your immune system as VDJ-recombination is essential for B and T cells to adapt to new pathogens. For a paper specifically on this mutation’s effects in zebrafish, see <span class="citation" data-cites="rag1-zebrafish">Hohn and Petrie-Hanson (<a href="#ref-rag1-zebrafish" role="doc-biblioref">2012</a>)</span> and <span class="citation" data-cites="rag1-zebrafish-2">Petrie-Hanson, Hohn, and Hanson (<a href="#ref-rag1-zebrafish-2" role="doc-biblioref">2009</a>)</span>.
</p>
</details>
</section>
<section id="tissue-preparation" class="level3">
<h3 class="anchored" data-anchor-id="tissue-preparation">Tissue Preparation</h3>
<blockquote class="blockquote">
<p>Kidneys from heterozygote transgenic zebrafish either <b style="color:#A6A440">wild-type</b> or <b style="color:#A6A440">rag1-/- mutant</b>, were dissected and processed as previously described <span class="citation" data-cites="tissue-preparation">(<a href="#ref-tissue-preparation" role="doc-biblioref">Athanasiadis et al. 2017</a>)</span>. The guts, spleens, gills and thymuses were dissected and placed in ice cold PBS/5% foetal bovine serum.</p>
<p>– <cite><b style="color:#EB1960">FACS Sorting Subsection; Materials and Methods Section;</b> <span class="citation" data-cites="zebrafish-data">Hernández et al. (<a href="#ref-zebrafish-data" role="doc-biblioref">2018</a>)</span> </cite></p>
</blockquote>
<p>Which gives us another paper to read! Also, I never thought I’d read the phrase “ice cold fetal bovine serum”…</p>
<blockquote class="blockquote">
<p>A single kidney from heterozygote transgenic or wild-type fish was dissected and placed in ice cold PBS/5% foetal bovine serum.</p>
<p>– <cite><b style="color:#EB1960">Single-Cell Sorting Subsection; Methods Section;</b> <span class="citation" data-cites="tissue-preparation">Athanasiadis et al. (<a href="#ref-tissue-preparation" role="doc-biblioref">2017</a>)</span> </cite></p>
</blockquote>
<p>Despite being given another paper to read, it seems that the same dissection type was used for all organs. That reduces the amount of work for us!</p>
</section>
<section id="cell-isolation" class="level3">
<h3 class="anchored" data-anchor-id="cell-isolation">Cell Isolation</h3>
<p>They used <b style="color:#537FBF">FACS (Fluorescence-Activated Cell Sorting)</b>. There’s a brief overview of it available from <span class="citation" data-cites="FACS"><span>“Fluorescence-Activated Cell Sorting (FACS)”</span> (<a href="#ref-FACS" role="doc-biblioref">n.d.</a>)</span>.</p>
<blockquote class="blockquote">
<p>For [the] <b style="color:#EB1960">Smart-seq2 experiment</b> individual cells were index sorted into 96 well plates using a <b style="color:#EB1960">BD Influx Index Sorter</b></p>
<p>– <cite><b style="color:#EB1960">FACS Sorting Subsection; Materials and Methods Section;</b> <span class="citation" data-cites="zebrafish-data">Hernández et al. (<a href="#ref-zebrafish-data" role="doc-biblioref">2018</a>)</span> </cite></p>
</blockquote>
<details>
<summary style="color:#C0CF95">
<b>96</b>
</summary>
<p>
Why does the number 96 keep popping up??? I’ll never know.
</p>
</details>
</section>
<section id="library-preparation-and-sequencing" class="level3">
<h3 class="anchored" data-anchor-id="library-preparation-and-sequencing">Library Preparation and Sequencing</h3>
<blockquote class="blockquote">
<p>The <b style="color:#EB1960">Smart-seq2 protocol</b> was used for <b style="color:#537FBF">whole transcriptome amplification</b> and <b style="color:#537FBF">library preparation</b> as previously described. Generated libraries were sequenced in <b style="color:#A6A440">pair-end mode</b> on [a] <b style="color:#EB1960">Hi-Seq4000 platform</b>.</p>
<p>– <cite><b style="color:#EB1960">Plate-Based Single-Cell RNA processing Subsection; Materials and Methods Section;</b> <span class="citation" data-cites="zebrafish-data">Hernández et al. (<a href="#ref-zebrafish-data" role="doc-biblioref">2018</a>)</span> </cite></p>
</blockquote>
<p>This is short and straight to the point. They say a bit more about sequencing later:</p>
<blockquote class="blockquote">
<p>For the samples that were processed using the <b style="color:#EB1960">Smart-seq2 protocol</b>, the reads were aligned to the zebrafish reference genome (<b style="color:#EB1960">Ensemble BioMart version 89</b>) combined with the sequences for EGFP, mCherry, mhc2dab and ERCC spike-ins. <b style="color:#EB1960">Salmon v0.8.2</b> <span class="citation" data-cites="salmon">(<a href="#ref-salmon" role="doc-biblioref">Patro et al. 2017</a>)</span> was used for both alignment and quantification of reads with the default paired-end parameters, while library type was set to inward (I) relative orientation (reads face each other) with unstranded (U) protocol (parameter –l IU).</p>
<p>– <cite><b style="color:#EB1960">Alignment and Quantification of Single-Cell RNA-Sequencing Data; Materials and Methods Section;</b> <span class="citation" data-cites="zebrafish-data">Hernández et al. (<a href="#ref-zebrafish-data" role="doc-biblioref">2018</a>)</span> </cite></p>
</blockquote>
</section>
</section>
<section id="acquiring-and-exploring-the-dataset" class="level2">
<h2 class="anchored" data-anchor-id="acquiring-and-exploring-the-dataset">Acquiring and Exploring the Dataset</h2>
<p>The page for this dataset on the <b style="color:#EB1960">Single Cell Expression Atlas</b> contains a <a href="https://www.ebi.ac.uk/gxa/sc/experiments/E-MTAB-7117/downloads">download tab</a>. We want the <strong style="color:#C0CF96">raw counts</strong> matrix (we don’t want to use the normalized counts, since we intend to recreate their analysis and that includes normalizing the counts ourselves!).</p>
<p>The file we’re interested in, after downloading and unzipping, is <code>E-MTAB-7117.aggregated_filtered_counts.mtx</code>. (The other files are important too; they’re the row and column names!) We can read the file in using <b style="color:#EB1960">R</b>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>raw.counts <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">readMM</span>(<span class="st">'./localdata/E-MTAB-7117.aggregated_filtered_counts.mtx'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s take a peak at this matrix to see what it’s like.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">dim</span>(raw.counts))</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>raw.counts[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">max</span>(raw.counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 21797   966</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre><code>10 x 10 sparse Matrix of class "dgTMatrix"
                                          
 [1,] .   .         .      . .   . . . . .
 [2,] .   .         .      . .   . . . . .
 [3,] .   .         .      . .   . . . . .
 [4,] .   .        48.000 77 .   . . . . .
 [5,] 1 680.8777 1151.611  1 .   . . . . .
 [6,] .   .         .      . .   . . . . .
 [7,] .   .         .      . .   . . . . .
 [8,] .   .         .      . . 111 . . . .
 [9,] .   .         .      . .   . . . . .
[10,] .   .         .      . .   . . . . .</code></pre>
</div>
<div class="cell-output cell-output-display">
292841
</div>
</div>
<p>We can see that the matrix is 21,797 by 966 - thus, there are 21,797 genes in the dataset and each column corresponds to a specific cell.</p>
<p>Strangely, <strong style="color:#C0CF96">some values seem to be noninteger</strong>? Let’s consult the paper!</p>
<blockquote class="blockquote">
<p>For each of the <strong style="color:#C0CF96">542 single cells</strong>, counts reported by <b style="color:#EB1960">Salmon</b> were transformed into normalised counts per million (CPM) and used for the further analysis. This was performed by <strong style="color:#C0CF96">dividing the number of counts for each gene with the total number of counts for each cell and by multiplying the resulting number by a factor of 1,000,000</strong>. Genes that were expressed in less than 1% of cells (e.g.&nbsp;5 single cells with CPM &gt; 1) <strong style="color:#C0CF96">were filtered out</strong>. In the final step we ended up using 16,059 genes across the 542 single cells. The <b style="color:#EB1960">scran R package</b> (version 1.6.7) <span class="citation" data-cites="scran">Lun, Bach, and Marioni (<a href="#ref-scran" role="doc-biblioref">2016</a>)</span> was then used to <b style="color:#537FBF">normalise</b> the data and remove differences due to the library size or capture efficiency and sequencing depth.</p>
<p>– <cite><b style="color:#EB1960">Downstream Analysis of Smart-seq2 Data; Materials and Methods Section;</b> <span class="citation" data-cites="zebrafish-data">Hernández et al. (<a href="#ref-zebrafish-data" role="doc-biblioref">2018</a>)</span> </cite></p>
</blockquote>
<p>This explains the non-integer values; there’s division involved in producing them.</p>
<p>However, we’ve found a new discrepancy! Why are there 966 cells in the dataset, when we should only have 542? Let’s read a bit more of the paper:</p>
<blockquote class="blockquote">
<p>For the <b style="color:#EB1960">Smart-seq2 protocol</b> transcript per million (TPM) values reported by Salmon were used for the quality control (QC). Wells with fewer than 900 expressed genes (TPM &gt; 1) or having more than either 60% of ERCC or 45% of mitochondrial content were annotated as <strong style="color:#C0CF96">poor quality cells</strong>. As a result, 322 cells failed QC and 542 single cells were selected for the further study.</p>
<p>– <cite><b style="color:#EB1960">Quality Control of Single-Cell Data; Materials and Methods Section;</b> <span class="citation" data-cites="zebrafish-data">Hernández et al. (<a href="#ref-zebrafish-data" role="doc-biblioref">2018</a>)</span> </cite></p>
</blockquote>
<p>So 322 cells failed quality control - but that still only brings us up to 864 cells, 102 too few. Let’s download the <a href="https://www.ebi.ac.uk/gxa/sc/experiments/E-MTAB-7117/experiment-design">experiment design table</a> from the <b style="color:#EB1960">Single Cell Expression Atlas</b>, as we’ll need it for later anyways (it contains non-gene info about the samples) and because it contains info on sample quality via the fields “well information” and “single cell quality”. The name of the file once downloaded is <code>ExpDesign-E-MTAB-7117.tsv</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>sample.data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"./localdata/ExpDesign-E-MTAB-7117.tsv"</span>, <span class="at">sep=</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">header=</span><span class="cn">TRUE</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sample.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 33</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">Assay</th>
<th data-quarto-table-cell-role="th" scope="col">Sample.Characteristic.organism.</th>
<th data-quarto-table-cell-role="th" scope="col">Sample.Characteristic.Ontology.Term.organism.</th>
<th data-quarto-table-cell-role="th" scope="col">Sample.Characteristic.strain.</th>
<th data-quarto-table-cell-role="th" scope="col">Sample.Characteristic.Ontology.Term.strain.</th>
<th data-quarto-table-cell-role="th" scope="col">Sample.Characteristic.age.</th>
<th data-quarto-table-cell-role="th" scope="col">Sample.Characteristic.Ontology.Term.age.</th>
<th data-quarto-table-cell-role="th" scope="col">Sample.Characteristic.developmental.stage.</th>
<th data-quarto-table-cell-role="th" scope="col">Sample.Characteristic.Ontology.Term.developmental.stage.</th>
<th data-quarto-table-cell-role="th" scope="col">Sample.Characteristic.sex.</th>
<th data-quarto-table-cell-role="th" scope="col">⋯</th>
<th data-quarto-table-cell-role="th" scope="col">Sample.Characteristic.single.cell.quality.</th>
<th data-quarto-table-cell-role="th" scope="col">Sample.Characteristic.Ontology.Term.single.cell.quality.</th>
<th data-quarto-table-cell-role="th" scope="col">Sample.Characteristic.cluster.</th>
<th data-quarto-table-cell-role="th" scope="col">Sample.Characteristic.Ontology.Term.cluster.</th>
<th data-quarto-table-cell-role="th" scope="col">Factor.Value.genotype.</th>
<th data-quarto-table-cell-role="th" scope="col">Factor.Value.Ontology.Term.genotype.</th>
<th data-quarto-table-cell-role="th" scope="col">Factor.Value.organism.part.</th>
<th data-quarto-table-cell-role="th" scope="col">Factor.Value.Ontology.Term.organism.part.</th>
<th data-quarto-table-cell-role="th" scope="col">Factor.Value.single.cell.identifier.</th>
<th data-quarto-table-cell-role="th" scope="col">Factor.Value.Ontology.Term.single.cell.identifier.</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;lgl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;lgl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">⋯</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;lgl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;lgl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;lgl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;lgl&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">1</td>
<td>ERR2722968</td>
<td>Danio rerio</td>
<td>http://purl.obolibrary.org/obo/NCBITaxon_7955</td>
<td>AB</td>
<td>NA</td>
<td>6 month</td>
<td>NA</td>
<td>adult</td>
<td>http://www.ebi.ac.uk/efo/EFO_0001272</td>
<td>male</td>
<td>⋯</td>
<td>OK</td>
<td>NA</td>
<td>1</td>
<td>NA</td>
<td>Tg(lck:EGFP)</td>
<td>NA</td>
<td>kidney</td>
<td></td>
<td>LCK_kidney_A12</td>
<td>NA</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">2</td>
<td>ERR2722969</td>
<td>Danio rerio</td>
<td>http://purl.obolibrary.org/obo/NCBITaxon_7955</td>
<td>AB</td>
<td>NA</td>
<td>6 month</td>
<td>NA</td>
<td>adult</td>
<td>http://www.ebi.ac.uk/efo/EFO_0001272</td>
<td>male</td>
<td>⋯</td>
<td>OK</td>
<td>NA</td>
<td>1</td>
<td>NA</td>
<td>Tg(lck:EGFP)</td>
<td>NA</td>
<td>kidney</td>
<td></td>
<td>LCK_kidney_A4</td>
<td>NA</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">3</td>
<td>ERR2722970</td>
<td>Danio rerio</td>
<td>http://purl.obolibrary.org/obo/NCBITaxon_7955</td>
<td>AB</td>
<td>NA</td>
<td>6 month</td>
<td>NA</td>
<td>adult</td>
<td>http://www.ebi.ac.uk/efo/EFO_0001272</td>
<td>male</td>
<td>⋯</td>
<td>OK</td>
<td>NA</td>
<td>1</td>
<td>NA</td>
<td>Tg(lck:EGFP)</td>
<td>NA</td>
<td>kidney</td>
<td></td>
<td>LCK_kidney_B1</td>
<td>NA</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">4</td>
<td>ERR2722971</td>
<td>Danio rerio</td>
<td>http://purl.obolibrary.org/obo/NCBITaxon_7955</td>
<td>AB</td>
<td>NA</td>
<td>6 month</td>
<td>NA</td>
<td>adult</td>
<td>http://www.ebi.ac.uk/efo/EFO_0001272</td>
<td>male</td>
<td>⋯</td>
<td>OK</td>
<td>NA</td>
<td>1</td>
<td>NA</td>
<td>Tg(lck:EGFP)</td>
<td>NA</td>
<td>kidney</td>
<td></td>
<td>LCK_kidney_B2</td>
<td>NA</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">5</td>
<td>ERR2722972</td>
<td>Danio rerio</td>
<td>http://purl.obolibrary.org/obo/NCBITaxon_7955</td>
<td>AB</td>
<td>NA</td>
<td>6 month</td>
<td>NA</td>
<td>adult</td>
<td>http://www.ebi.ac.uk/efo/EFO_0001272</td>
<td>male</td>
<td>⋯</td>
<td>OK</td>
<td>NA</td>
<td>1</td>
<td>NA</td>
<td>Tg(lck:EGFP)</td>
<td>NA</td>
<td>kidney</td>
<td></td>
<td>LCK_kidney_C10</td>
<td>NA</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">6</td>
<td>ERR2722973</td>
<td>Danio rerio</td>
<td>http://purl.obolibrary.org/obo/NCBITaxon_7955</td>
<td>AB</td>
<td>NA</td>
<td>6 month</td>
<td>NA</td>
<td>adult</td>
<td>http://www.ebi.ac.uk/efo/EFO_0001272</td>
<td>male</td>
<td>⋯</td>
<td>OK</td>
<td>NA</td>
<td>1</td>
<td>NA</td>
<td>Tg(lck:EGFP)</td>
<td>NA</td>
<td>kidney</td>
<td></td>
<td>LCK_kidney_C11</td>
<td>NA</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sample.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>1056</li><li>33</li></ol>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sample.data[<span class="fu">which</span>(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    sample.data[<span class="st">'Sample.Characteristic.single.cell.quality.'</span>] <span class="sc">==</span> <span class="st">"OK"</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    <span class="sc">&amp;</span> sample.data[<span class="st">'Sample.Characteristic.well.information.'</span>] <span class="sc">==</span> <span class="st">"single cell"</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>),])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>976</li><li>33</li></ol>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>filtered <span class="ot">&lt;-</span> sample.data[<span class="fu">which</span>(</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    sample.data[<span class="st">'Sample.Characteristic.cluster.'</span>] <span class="sc">!=</span> <span class="st">"unknown"</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>),]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(filtered)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>542</li><li>33</li></ol>
</div>
</div>
<p>The mystery keeps compounding! For some reason, we start with 90 extra rows (cells) in this dataset! We can filter out the obviously bad entries, but we still have 10 extra rows. There’s a field “cluster” which we can use to get the post-filtered value of 542, although I’m fairly certain that these clusters were created after the analysis rather than as part of the data creation, hence why there’s only 542.</p>
<p>Since we know the names of all ‘names’ of all the samples in our experiment design table, we can compare this with the list of row names for our gene count matrix to hopefully shed some light on what’s going wrong:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Grab the lists of genes in the experiment</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co"># and in the gene counts, for comparison</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>experiment.assays <span class="ot">&lt;-</span> sample.data<span class="sc">$</span>Assay</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>gene.assays <span class="ot">&lt;-</span> <span class="fu">read.table</span>(</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">'./localdata/E-MTAB-7117.aggregated_filtered_counts.mtx_cols'</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>)<span class="sc">$</span>V1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the list of genes that appear in the experiment data</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co"># but not the gene counts</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>discrepancies <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(experiment.assays, gene.assays)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(discrepancies)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
90
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Now get the experiment info of the 10 of these cells that</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co"># aren't filtered out by our previous filtering method</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>extra.cells <span class="ot">&lt;-</span> sample.data[</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">which</span>(<span class="fu">sapply</span>(experiment.assays, <span class="st">`</span><span class="at">%in%</span><span class="st">`</span>, discrepancies)),</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>extra.cells <span class="ot">&lt;-</span> extra.cells[</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">which</span>(</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>        extra.cells[<span class="st">'Sample.Characteristic.single.cell.quality.'</span>] <span class="sc">==</span> <span class="st">"OK"</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Show all columns.  There's probably a better way,</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co"># but oh well...</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># (I've hidden the outputs of this code block on the blog,</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co"># because it's just a bunch of ugly, unenlightening tables.)</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>extra.cells[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>extra.cells[,<span class="dv">11</span><span class="sc">:</span><span class="dv">20</span>]</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>extra.cells[,<span class="dv">21</span><span class="sc">:</span><span class="dv">30</span>]</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>extra.cells[,<span class="dv">31</span><span class="sc">:</span><span class="dv">33</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The above contains the 10 outstanding cells. Frustratingly, there doesn’t seem to be any reason I can find as to why these cells were excluded at this stage!</p>
<p><b>To summarize</b>:</p>
<ul>
<li>The website reports 966 cells</li>
<li>The gene count table also reports 966 cells</li>
<li>The experiment table reports 1056 cells</li>
<li>Flitering the experiment table yields 976 cells</li>
<li>The paper reports 864 cells pre-quality control</li>
</ul>
<p>This isn’t an uncommon occurance; in fact, every time I’ve read a paper like this (which, admittedly, is only 3ish times) I’ve been unable to get a consistent answer on how many cells were actually used! <strong style="color:#C0CF96">It’s a major nuisance</strong>.</p>
<p>We’ll pick up where we left off in the next blog post. While annoying, this discrepancy isn’t fatal - we can perform quality control on the 976 cells we get from filtering.</p>
<section id="references" class="level3">
<h3 class="anchored" data-anchor-id="references">References</h3>
<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-tissue-preparation" class="csl-entry" role="listitem">
Athanasiadis, E. I., J. G. Botthof, H. Andres, L. Ferreira, P. Lio, and A. Cvejic. 2017. <span>“<span class="nocase"><span>S</span>ingle-cell <span>R</span><span>N</span><span>A</span>-sequencing uncovers transcriptional states and fate decisions in haematopoiesis</span>”</span> 8 (1): 2045.
</div>
<div id="ref-FACS" class="csl-entry" role="listitem">
<span>“Fluorescence-Activated Cell Sorting (FACS).”</span> n.d. SinoBiological. <a href="https://www.sinobiological.com/category/fcm-facs-facs">https://www.sinobiological.com/category/fcm-facs-facs</a>.
</div>
<div id="ref-vibrio-anguillarum" class="csl-entry" role="listitem">
Frans, I., C. W. Michiels, P. Bossier, K. A. Willems, B. Lievens, and H. Rediers. 2011. <span>“<span class="nocase"><span>V</span>ibrio anguillarum as a fish pathogen: virulence factors, diagnosis and prevention</span>.”</span> <em>Journal of Fish Diseases</em> 34 (9): 643–61.
</div>
<div id="ref-zebrafish-data" class="csl-entry" role="listitem">
Hernández, P. P., P. M. Strzelecka, E. I. Athanasiadis, D. Hall, A. F. Robalo, C. M. Collins, P. Boudinot, J. P. Levraud, and A. Cvejic. 2018. <span>“<span class="nocase"><span>S</span>ingle-cell transcriptional analysis reveals <span>I</span><span>L</span><span>C</span>-like cells in zebrafish</span>.”</span> <em>Science Immunology</em> 3 (29).
</div>
<div id="ref-rag1-zebrafish" class="csl-entry" role="listitem">
Hohn, C., and L. Petrie-Hanson. 2012. <span>“<span class="nocase"><span>R</span>ag1-/- mutant zebrafish demonstrate specific protection following bacterial re-exposure</span>.”</span> <em>PLoS One</em> 7 (9): e44451.
</div>
<div id="ref-b-vs-strong" class="csl-entry" role="listitem">
Lucero, Sophia. 2008. <span>“HTML5: &lt;B&gt; and &lt;i&gt; Tags Are Going to Be Useful (Read: Semantic) Again!”</span> <a href="https://stellify.net/html5-b-and-i-tags-are-going-to-be-useful-read-semantic-again/">https://stellify.net/html5-b-and-i-tags-are-going-to-be-useful-read-semantic-again/</a>.
</div>
<div id="ref-scran" class="csl-entry" role="listitem">
Lun, A. T., K. Bach, and J. C. Marioni. 2016. <span>“<span class="nocase"><span>P</span>ooling across cells to normalize single-cell <span>R</span><span>N</span><span>A</span> sequencing data with many zero counts</span>.”</span> <em>Genome Biol</em> 17 (April): 75.
</div>
<div id="ref-salmon" class="csl-entry" role="listitem">
Patro, R., G. Duggal, M. I. Love, R. A. Irizarry, and C. Kingsford. 2017. <span>“<span class="nocase"><span>S</span>almon provides fast and bias-aware quantification of transcript expression</span>.”</span> <em>Nat Methods</em> 14 (4): 417–19.
</div>
<div id="ref-rag1-zebrafish-2" class="csl-entry" role="listitem">
Petrie-Hanson, L., C. Hohn, and L. Hanson. 2009. <span>“<span class="nocase"><span>C</span>haracterization of rag1 mutant zebrafish leukocytes</span>.”</span> <em>BMC Immunology</em> 10 (February): 8.
</div>
<div id="ref-medline-rag1" class="csl-entry" role="listitem">
<span>“RAG1 Gene.”</span> 2017, February. <a href="https://medlineplus.gov/genetics/gene/rag1/">https://medlineplus.gov/genetics/gene/rag1/</a>.
</div>
</div>
<script src="https://giscus.app/client.js" data-repo="baileyandrew/blog" data-repo-id="R_kgDOInJwKg" data-category="Announcements" data-category-id="DIC_kwDOInJwKs4CTGOQ" data-mapping="title" data-strict="0" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="dark_protanopia" data-lang="en" data-loading="lazy" crossorigin="anonymous" async="">
</script>


</section>
</section>
</section>


<div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>Apparently it does seem to be one of the ‘useful’ animals in biology - you learn new things every day! Shoutout to Sam’s blog for introducing me to being able to use footnotes.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>It wasn’t really, at least for me - I like to be sure!<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>I found it on <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6258902/">NIH</a>; warning - you’ll have to download a 16MB file to read it!<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p>I’m leaning towards misinterpretation, as I don’t see why wild populations would be incapable of having a <b style="color:#A6A440">rag1<sup>-/-</sup> mutation</b> - unless maybe the <b style="color:#A6A440">rag1<sup>-/-</sup> mutants</b> used in the study were specifically bred in the lab?<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>