<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Bailey Andrew">
<meta name="dcterms.date" content="2023-01-10">
<meta name="description" content="A continuation of yesterday; what can we learn from the data?">

<title>Lorem Ipsomething - Analyzing scRNA-seq Data – Exploration</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../styles.css">
<link rel="stylesheet" href="../../html_scripts/collapse.css">
<link rel="stylesheet" href="../../html_scripts/pretty_shortcuts.css">
<meta property="og:title" content="Lorem Ipsomething - Analyzing scRNA-seq Data – Exploration">
<meta property="og:description" content="A continuation of yesterday; what can we learn from the data?">
<meta property="og:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/r-help.png">
<meta property="og:site-name" content="Lorem Ipsomething">
<meta property="og:image:height" content="1038">
<meta property="og:image:width" content="1960">
<meta name="twitter:title" content="Lorem Ipsomething - Analyzing scRNA-seq Data – Exploration">
<meta name="twitter:description" content="A continuation of yesterday; what can we learn from the data?">
<meta name="twitter:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/r-help.png">
<meta name="twitter:image-height" content="1038">
<meta name="twitter:image-width" content="1960">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Lorem Ipsomething</span>
    </a>
  </div>
        <div class="quarto-navbar-tools ms-auto">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Analyzing scRNA-seq Data – Exploration</li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Home</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../blog.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Blog</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../staticposts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Static Posts</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Puzzles</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../500_Puzzles/bnv.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Boulders in Valleys</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slideshowindex.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slideshows</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../otherblogs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Other People’s Blogs</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#quality-control" id="toc-quality-control" class="nav-link active" data-scroll-target="#quality-control">Quality Control</a></li>
  <li><a href="#analysis" id="toc-analysis" class="nav-link" data-scroll-target="#analysis">Analysis</a>
  <ul class="collapse">
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/BaileyAndrew/Blog/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Analyzing scRNA-seq Data – Exploration</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Work</div>
    <div class="quarto-category">Omics</div>
    <div class="quarto-category">Paper Replication</div>
  </div>
  </div>

<div>
  <div class="description">
    A continuation of yesterday; what can we learn from the data?
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Bailey Andrew </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 10, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p>Yesterday, we looked at a scRNA-seq dataset of <i style="color:#EB1960">Danio rerio</i> cells <span class="citation" data-cites="zebrafish-data">(<a href="#ref-zebrafish-data" role="doc-biblioref">Hernández et al. 2018</a>)</span>. We spent a lot of time understanding how the dataset was created, and ended with a bit of a mystery: <i style="color:#C0CF96">why did the cell counts not match up between our datasets?</i> I don’t have an answer to that mystery, unfortunately - but in my experience this mismatch happens in a lot of papers. We might as well continue with the analysis, rather than getting hung up on a minor anomaly.</p>
<p>I realized since yesterday that there was another gene counts file, (the filtered tpms file). I downloaded this and the “experiment metadata” files; I suspect these may be useful as I know that the TPM values were pre-quality-control:</p>
<blockquote class="blockquote">
<p>For the <b style="color:#EB1960">Smart-seq2 protocol</b> transcript per million (TPM) values reported by Salmon were used for the quality control (QC). Wells with fewer than 900 expressed genes (TPM &gt; 1) or having more than either 60% of ERCC or 45% of mitochondrial content were annotated as <strong style="color:#C0CF96">poor quality cells</strong>. As a result, 322 cells failed QC and 542 single cells were selected for the further study.</p>
<p>– <cite><b style="color:#EB1960">Quality Control of Single-Cell Data; Materials and Methods Section;</b> <span class="citation" data-cites="zebrafish-data">Hernández et al. (<a href="#ref-zebrafish-data" role="doc-biblioref">2018</a>)</span> </cite></p>
</blockquote>
<details>
<summary style="color:#C0CF96">
<b>Can I mix R and Python in the same notebook?</b>
</summary>
Yes! I sometimes use <b style="color:#EB1960">SOS Kernel</b> which allows me to swap between kernels at ease. You can change kernels (R vs Python) using <span style="color:#757575">%use</span> magic commands; I’ve done this before but the syntax highlighting gets messed up often, so I’ve chosen not to. do that here.
</details>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>raw.counts <span class="op">&lt;-</span> Matrix::readMM(<span class="st">'./localdata/E-MTAB-7117.expression_tpm.mtx'</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>dim(raw.counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>21797</li><li>966</li></ol>
</div>
</div>
<details>
<summary style="color:#C0CF96">
<b>Does the metadata file help us resolve the cell count mystery?</b>
</summary>
<p>What does this extra metadata file tell us? Let’s see:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>meta.data <span class="op">&lt;-</span> read.table(<span class="st">"./localdata/E-MTAB-7117.sdrf.txt"</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, header <span class="op">=</span> TRUE)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>head(meta.data)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>dim(meta.data)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>length(which(meta.data[<span class="st">'Characteristics.single.cell.quality.'</span>] <span class="op">!=</span> <span class="st">"not OK"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 6 × 55</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">Source.Name</th>
<th data-quarto-table-cell-role="th" scope="col">Comment.ENA_SAMPLE.</th>
<th data-quarto-table-cell-role="th" scope="col">Comment.BioSD_SAMPLE.</th>
<th data-quarto-table-cell-role="th" scope="col">Characteristics.organism.</th>
<th data-quarto-table-cell-role="th" scope="col">Characteristics.strain.</th>
<th data-quarto-table-cell-role="th" scope="col">Characteristics.age.</th>
<th data-quarto-table-cell-role="th" scope="col">Unit.time.unit.</th>
<th data-quarto-table-cell-role="th" scope="col">Characteristics.developmental.stage.</th>
<th data-quarto-table-cell-role="th" scope="col">Characteristics.sex.</th>
<th data-quarto-table-cell-role="th" scope="col">Characteristics.genotype.</th>
<th data-quarto-table-cell-role="th" scope="col">⋯</th>
<th data-quarto-table-cell-role="th" scope="col">Comment.ENA_EXPERIMENT.</th>
<th data-quarto-table-cell-role="th" scope="col">Scan.Name</th>
<th data-quarto-table-cell-role="th" scope="col">Comment.SUBMITTED_FILE_NAME.</th>
<th data-quarto-table-cell-role="th" scope="col">Comment.ENA_RUN.</th>
<th data-quarto-table-cell-role="th" scope="col">Comment.FASTQ_URI.</th>
<th data-quarto-table-cell-role="th" scope="col">Comment.SPOT_LENGTH.</th>
<th data-quarto-table-cell-role="th" scope="col">Comment.READ_INDEX_1_BASE_COORD.</th>
<th data-quarto-table-cell-role="th" scope="col">Factor.Value.genotype.</th>
<th data-quarto-table-cell-role="th" scope="col">Factor.Value.organism.part.</th>
<th data-quarto-table-cell-role="th" scope="col">Factor.Value.single.cell.identifier.</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">⋯</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">1</td>
<td>CD4_gill_A1</td>
<td>ERS2634491</td>
<td>SAMEA4814592</td>
<td>Danio rerio</td>
<td>AB</td>
<td>6</td>
<td>month</td>
<td>adult</td>
<td>male</td>
<td>Tg(cd4-1:mCherry)</td>
<td>⋯</td>
<td>ERX2737040</td>
<td>SLX-10875.N701_S513.C9FTNANXX.s_5.r_1.fq.gz</td>
<td>SLX-10875.N701_S513.C9FTNANXX.s_5.r_1.fq.gz</td>
<td>ERR2723271</td>
<td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR272/001/ERR2723271/ERR2723271_1.fastq.gz</td>
<td>250</td>
<td>126</td>
<td>Tg(cd4-1:mCherry)</td>
<td>gill</td>
<td>CD4_gill_A1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">2</td>
<td>CD4_gill_A1</td>
<td>ERS2634491</td>
<td>SAMEA4814592</td>
<td>Danio rerio</td>
<td>AB</td>
<td>6</td>
<td>month</td>
<td>adult</td>
<td>male</td>
<td>Tg(cd4-1:mCherry)</td>
<td>⋯</td>
<td>ERX2737040</td>
<td>SLX-10875.N701_S513.C9FTNANXX.s_5.r_2.fq.gz</td>
<td>SLX-10875.N701_S513.C9FTNANXX.s_5.r_2.fq.gz</td>
<td>ERR2723271</td>
<td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR272/001/ERR2723271/ERR2723271_2.fastq.gz</td>
<td>250</td>
<td>126</td>
<td>Tg(cd4-1:mCherry)</td>
<td>gill</td>
<td>CD4_gill_A1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">3</td>
<td>CD4_gill_A10</td>
<td>ERS2634691</td>
<td>SAMEA4814793</td>
<td>Danio rerio</td>
<td>AB</td>
<td>6</td>
<td>month</td>
<td>adult</td>
<td>male</td>
<td>Tg(cd4-1:mCherry)</td>
<td>⋯</td>
<td>ERX2737240</td>
<td>SLX-10875.N712_S513.C9FTNANXX.s_5.r_1.fq.gz</td>
<td>SLX-10875.N712_S513.C9FTNANXX.s_5.r_1.fq.gz</td>
<td>ERR2723471</td>
<td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR272/001/ERR2723471/ERR2723471_1.fastq.gz</td>
<td>250</td>
<td>126</td>
<td>Tg(cd4-1:mCherry)</td>
<td>gill</td>
<td>CD4_gill_A10</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">4</td>
<td>CD4_gill_A10</td>
<td>ERS2634691</td>
<td>SAMEA4814793</td>
<td>Danio rerio</td>
<td>AB</td>
<td>6</td>
<td>month</td>
<td>adult</td>
<td>male</td>
<td>Tg(cd4-1:mCherry)</td>
<td>⋯</td>
<td>ERX2737240</td>
<td>SLX-10875.N712_S513.C9FTNANXX.s_5.r_2.fq.gz</td>
<td>SLX-10875.N712_S513.C9FTNANXX.s_5.r_2.fq.gz</td>
<td>ERR2723471</td>
<td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR272/001/ERR2723471/ERR2723471_2.fastq.gz</td>
<td>250</td>
<td>126</td>
<td>Tg(cd4-1:mCherry)</td>
<td>gill</td>
<td>CD4_gill_A10</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">5</td>
<td>CD4_gill_A11</td>
<td>ERS2635221</td>
<td>SAMEA4815323</td>
<td>Danio rerio</td>
<td>AB</td>
<td>6</td>
<td>month</td>
<td>adult</td>
<td>male</td>
<td>Tg(cd4-1:mCherry)</td>
<td>⋯</td>
<td>ERX2737770</td>
<td>SLX-10875.N714_S513.C9FTNANXX.s_5.r_1.fq.gz</td>
<td>SLX-10875.N714_S513.C9FTNANXX.s_5.r_1.fq.gz</td>
<td>ERR2724001</td>
<td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR272/001/ERR2724001/ERR2724001_1.fastq.gz</td>
<td>250</td>
<td>126</td>
<td>Tg(cd4-1:mCherry)</td>
<td>gill</td>
<td>CD4_gill_A11</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">6</td>
<td>CD4_gill_A11</td>
<td>ERS2635221</td>
<td>SAMEA4815323</td>
<td>Danio rerio</td>
<td>AB</td>
<td>6</td>
<td>month</td>
<td>adult</td>
<td>male</td>
<td>Tg(cd4-1:mCherry)</td>
<td>⋯</td>
<td>ERX2737770</td>
<td>SLX-10875.N714_S513.C9FTNANXX.s_5.r_2.fq.gz</td>
<td>SLX-10875.N714_S513.C9FTNANXX.s_5.r_2.fq.gz</td>
<td>ERR2724001</td>
<td>ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR272/001/ERR2724001/ERR2724001_2.fastq.gz</td>
<td>250</td>
<td>126</td>
<td>Tg(cd4-1:mCherry)</td>
<td>gill</td>
<td>CD4_gill_A11</td>
</tr>
</tbody>
</table>
</div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>2112</li><li>55</li></ol>
</div>
<div class="cell-output cell-output-display">
1952
</div>
</div>
<p>At first glance, <b style="color:#C0CF96">no</b> - it does not. The metadata has 2112 lines, and removing poor quality cells leaves us with 1952.</p>
<p>But, it does seem like some cells have multiple names, so let’s filter by unique values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>dim(unique(meta.data[<span class="st">'Source.Name'</span>]))</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>length(</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    unique(</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>        meta.data[</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>            which(meta.data[<span class="st">'Characteristics.single.cell.quality.'</span>] <span class="op">!=</span> <span class="st">"not OK"</span>),</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Source.Name"</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>        ]</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>1056</li><li>1</li></ol>
</div>
<div class="cell-output cell-output-display">
976
</div>
</div>
<p>Gee whiz, that 976 number returns… We can re-use the list of 10 extra cells calculated yesterday to investigate them:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>discrepancies <span class="op">&lt;-</span> c(</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ERR2723217"</span>, <span class="st">"ERR2723218"</span>, <span class="st">"ERR2723236"</span>, <span class="st">"ERR2723237"</span>, <span class="st">"ERR2723551"</span>,</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ERR2723578"</span>, <span class="st">"ERR2723789"</span>, <span class="st">"ERR2723794"</span>, <span class="st">"ERR2723974"</span>, <span class="st">"ERR2723990"</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>extra.cells <span class="op">&lt;-</span> meta.data[</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    which(sapply(meta.data[<span class="st">"Comment.ENA_RUN."</span>], `<span class="op">%</span><span class="kw">in</span><span class="op">%</span>`, discrepancies)),</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Output hidden to reduce spam</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>extra.cells[<span class="dv">1</span>:<span class="dv">10</span>]</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>extra.cells[<span class="dv">11</span>:<span class="dv">20</span>]</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>extra.cells[<span class="dv">21</span>:<span class="dv">30</span>]</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>extra.cells[<span class="dv">31</span>:<span class="dv">40</span>]</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>extra.cells[<span class="dv">41</span>:<span class="dv">55</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>But as before there doesn’t really seem to be any outstanding features. So the mystery continues!</p>
</details>
<section id="quality-control" class="level1">
<h1>Quality Control</h1>
<p>The <b style="color:#757575">raw.counts</b> matrix is a <b style="color:#A6A440">Sparse Matrix</b> with 21,797 genes and 966 cells - the values are expressed in TPM. Our goal is to filter out the low quality cells.</p>
<blockquote class="blockquote">
<p>For the <b style="color:#EB1960">Smart-seq2 protocol</b> transcript per million (TPM) values reported by Salmon were used for the quality control (QC). Wells with fewer than 900 expressed genes (TPM &gt; 1) or having more than either 60% of ERCC or 45% of mitochondrial content were annotated as <strong style="color:#C0CF96">poor quality cells</strong>. As a result, 322 cells failed QC and 542 single cells were selected for the further study.</p>
<p>– <cite><b style="color:#EB1960">Quality Control of Single-Cell Data; Materials and Methods Section;</b> <span class="citation" data-cites="zebrafish-data">Hernández et al. (<a href="#ref-zebrafish-data" role="doc-biblioref">2018</a>)</span> </cite></p>
</blockquote>
<p>At the time, I didn’t think we could find the mitochondrial genes with the data we had (and same with ERCC). I’ve gone back and edited this though, in the dropdowns.</p>
<details>
<summary style="color:#C0CF96">
<b>Getting mitochondrial genes</b>
</summary>
<p>I used Python since chronologically I did this ater I gave up on R. The work to create the <code>mito-genes</code> csv was done in the blog post <a href="./004_005_ensemble.html">Exploring Ensembl</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>mito_genes <span class="op">=</span> pd.read_csv(<span class="st">"./localdata/mito-genes.csv"</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>mito_genes[<span class="st">"gene_id"</span>].head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>0    ENSDARG00000083480
1    ENSDARG00000082753
2    ENSDARG00000081443
3    ENSDARG00000080337
4    ENSDARG00000083046
Name: gene_id, dtype: object</code></pre>
</div>
</div>
<p>Now we want to calculate the percent of genomic material per cell that is contained in just the genes in <code>mito_genes</code>, which should not be too hard! Left as excersize to the reader.</p>
</details>
<details>
<summary style="color:#C0CF96">
<b>Getting ERCC genes</b>
</summary>
<p>I could not figure out how to do this; I think they have already been removed from the dataset that we have.</p>
</details>
<p>But yesterday we noticed that we could get the 542 futher-study cells by looking at the metadata and only grabbing the cells which were eventually assigned a cluster by the researchers.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the cells we want to study</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>sample.data <span class="op">&lt;-</span> read.table(<span class="st">"./localdata/ExpDesign-E-MTAB-7117.tsv"</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, header<span class="op">=</span>TRUE)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>filtered <span class="op">&lt;-</span> sample.data[which(</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    sample.data[<span class="st">'Sample.Characteristic.cluster.'</span>] <span class="op">!=</span> <span class="st">"unknown"</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>),]</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>cells.to.study <span class="op">&lt;-</span> filtered$Assay</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>length(cells.to.study)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
542
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get which rows correspond to which assay (cell)</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>assay.to.row.<span class="bu">map</span> <span class="op">&lt;-</span> read.table(</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">'./localdata/E-MTAB-7117.expression_tpm.mtx_cols'</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>)$V1</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>length(assay.to.row.<span class="bu">map</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
966
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset the raw.counts matrix</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>raw.counts.filtered <span class="op">&lt;-</span> raw.counts[,</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    which(sapply(assay.to.row.<span class="bu">map</span>, `<span class="op">%</span><span class="kw">in</span><span class="op">%</span>`, cells.to.study)),</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>dim(raw.counts.filtered)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Save the matrix, for posterity</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>writeMM(raw.counts.filtered, <span class="st">"./localdata/542_cells_21797_genes.tpm.mtx"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>21797</li><li>542</li></ol>
</div>
<div class="cell-output cell-output-display">
<pre><code>NULL</code></pre>
</div>
</div>
<p>Now we need to perform quality control on the genes;</p>
<blockquote class="blockquote">
<p>For each of the <strong style="color:#C0CF96">542 single cells</strong>, counts reported by <b style="color:#EB1960">Salmon</b> were transformed into normalised counts per million (CPM) and used for the further analysis. This was performed by <strong style="color:#C0CF96">dividing the number of counts for each gene with the total number of counts for each cell and by multiplying the resulting number by a factor of 1,000,000</strong>. Genes that were expressed in less than 1% of cells (e.g.&nbsp;5 single cells with CPM &gt; 1) <strong style="color:#C0CF96">were filtered out</strong>. In the final step we ended up using 16,059 genes across the 542 single cells. The <b style="color:#EB1960">scran R package</b> (version 1.6.7) <span class="citation" data-cites="scran">Lun, Bach, and Marioni (<a href="#ref-scran" role="doc-biblioref">2016</a>)</span> was then used to <b style="color:#537FBF">normalise</b> the data and remove differences due to the library size or capture efficiency and sequencing depth.</p>
<p>– <cite><b style="color:#EB1960">Downstream Analysis of Smart-seq2 Data; Materials and Methods Section;</b> <span class="citation" data-cites="zebrafish-data">Hernández et al. (<a href="#ref-zebrafish-data" role="doc-biblioref">2018</a>)</span> </cite></p>
</blockquote>
<p>It was harder than expected, as I am less-than-fluent with R and their sparse matrix operations weren’t as analogous to dense matrices as I would have liked. First let’s take a look at our the sparse datatype we have. I found the blogpost by <span class="citation" data-cites="dgTMatrix">Slowikowski (<a href="#ref-dgTMatrix" role="doc-biblioref">2020</a>)</span> useful, but ultimately got fed up with R and decided to transition to python.</p>
<details>
<summary>
<span class="false-starts" style="font-weight:bold;">False Start</span>: <b style="color:#C0CF96">Trying to use R</b>
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="bu">str</span>(raw.counts.filtered)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Formal class 'dgTMatrix' [package "Matrix"] with 6 slots
  ..@ i       : int [1:1156509] 4 10 19 36 56 57 58 59 62 69 ...
  ..@ j       : int [1:1156509] 0 0 0 0 0 0 0 0 0 0 ...
  ..@ Dim     : int [1:2] 21797 542
  ..@ Dimnames:List of 2
  .. ..$ : NULL
  .. ..$ : NULL
  ..@ x       : num [1:1156509] 0.593 145.541 361.733 0.178 0.107 ...
  ..@ factors : list()</code></pre>
</div>
</div>
<p><code>(i, j)</code> gives us the rows and columns, with the value being the corresponding entry in <code>x</code>. The <code>T</code> in <code>dgTMatrix</code> stands for triplet, because it’s essentially just a list of triplets <code>(i, j, x)</code>. Another common format is <code>CsparseMatrix</code>, although the explanation is more complicated. It’s explained well by <span class="citation" data-cites="dgTMatrix">Slowikowski (<a href="#ref-dgTMatrix" role="doc-biblioref">2020</a>)</span>, I’m just in a rush to get to actual data analysis!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>raw.counts.csparse <span class="op">&lt;-</span> <span class="bu">str</span>(<span class="im">as</span>(raw.counts.filtered, <span class="st">"CsparseMatrix"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  ..@ i       : int [1:1156509] 4 10 19 36 56 57 58 59 62 69 ...
  ..@ p       : int [1:543] 0 1610 3532 5491 7585 9273 11182 13157 14623 15552 ...
  ..@ Dim     : int [1:2] 21797 542
  ..@ Dimnames:List of 2
  .. ..$ : NULL
  .. ..$ : NULL
  ..@ x       : num [1:1156509] 0.593 145.541 361.733 0.178 0.107 ...
  ..@ factors : list()</code></pre>
</div>
</div>
<p>Now, we want to transform the data into <b style="color:#C0CF96">normalized counts per million</b>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This will throw an error</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>colSums(raw.counts.csparse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>ERROR: Error in colSums(raw.counts.csparse): 'x' must be an array of at least two dimensions</code></pre>
</div>
</div>
<details>
<summary style="color:#C0CF96;font-weight:bold;">
But we know it should work for sparse matrixes from the docs
</summary>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/r-help.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">What happens when we run <code>?colSums</code> to get the docs</figcaption>
</figure>
</div>
<p>Fun fact: for some reason running the R help command will mess with Quarto’s html, breaking your webpage!</p>
</details>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can try to force it to use the Matrix overload, to no avail.</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>Matrix::colSums(raw.counts.csparse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>ERROR: Error in base::rowSums(x, na.rm = na.rm, dims = dims, ...): 'x' must be an array of at least two dimensions</code></pre>
</div>
</div>
<p>And it’s about here that I get fed up with R.</p>
</details>
<p>Let’s load our <code>raw.counts.filtered</code> matrix in python:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.io <span class="im">import</span> mmread</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy <span class="im">import</span> sparse</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>raw_counts_filtered <span class="op">=</span> sparse.csc_array(mmread(</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">'./localdata/542_cells_21797_genes.tpm.mtx'</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>))</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>raw_counts_filtered</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>&lt;21797x542 sparse array of type '&lt;class 'numpy.float64'&gt;'
    with 1156509 stored elements in Compressed Sparse Column format&gt;</code></pre>
</div>
</div>
<p>And now we’ll copy the process they used to rescale their data in terms of genes per cell.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>genes_per_cell <span class="op">=</span> raw_counts_filtered.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>raw_counts_nonzero <span class="op">=</span> raw_counts_filtered[:, genes_per_cell <span class="op">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>genes_per_cell <span class="op">=</span> genes_per_cell[genes_per_cell <span class="op">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>raw_counts_nonzero <span class="op">=</span> raw_counts_nonzero <span class="op">/</span> genes_per_cell <span class="op">*</span> <span class="fl">1e6</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>raw_counts_normed <span class="op">=</span> sparse.csr_array(raw_counts_nonzero)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>raw_counts_normed</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>&lt;21797x542 sparse array of type '&lt;class 'numpy.float64'&gt;'
    with 1156509 stored elements in Compressed Sparse Row format&gt;</code></pre>
</div>
</div>
<p>Now we need to filter out genes who were expressed in less than 1% of cells. They’re a bit more strict than that - it has to have a CPM of at least 1 to count as being “expressed”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>raw_counts_expressed <span class="op">=</span> raw_counts_normed.copy()</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>raw_counts_expressed[raw_counts_expressed <span class="op">&lt;=</span> <span class="dv">1</span>] <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>raw_counts_expressed[raw_counts_expressed <span class="op">&gt;</span> <span class="dv">1</span>] <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>number_of_cells_gene_appears_in <span class="op">=</span> raw_counts_expressed.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>raw_counts_genes_filtered <span class="op">=</span> raw_counts_normed[</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>    number_of_cells_gene_appears_in <span class="op">&gt;</span> <span class="dv">5</span>,</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>    :</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>].copy()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>raw_counts_genes_filtered</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>&lt;17832x542 sparse array of type '&lt;class 'numpy.float64'&gt;'
    with 1141288 stored elements in Compressed Sparse Row format&gt;</code></pre>
</div>
</div>
<p>We have too many genes and I’m not sure why - however, I can’t find anywhere in their paper a count of the pre-filtered genes. Just like how their dataset contains more cells than reported, it may contain more genes than reported!</p>
</section>
<section id="analysis" class="level1">
<h1>Analysis</h1>
<blockquote class="blockquote">
<p>In order to identify the <b style="color:#A6A440">highly variable genes (HVGs)</b> we utilised the <b style="color:#537FBF">Brennecke Method</b> <span class="citation" data-cites="brennecke">(<a href="#ref-brennecke" role="doc-biblioref">Brennecke et al. 2013</a>)</span>. We inferred the noise model from the <b style="color:#A6A440">ERCCs</b> and <b style="color:#C0CF96">selected genes that vary higher than 20% percentage of variation</b>. This was performed by using the “<b style="color:#757575">BrenneckeGetVariableGenes</b>” command of <b style="color:#EB1960">M3Drop v1.4.0 R package</b> setting fdr equal to 0.01 and minimum percentage of variance due to biological factors (minBiolDisp) equal to 0.2. <b style="color:#C0CF96">In total, 3,374 were annotated as HVGs.</b></p>
<p>– <cite><b style="color:#EB1960">Downstream Analysis of Smart-seq2 Data; Materials and Methods Section;</b> <span class="citation" data-cites="zebrafish-data">Hernández et al. (<a href="#ref-zebrafish-data" role="doc-biblioref">2018</a>)</span> </cite></p>
</blockquote>
<p>The <code>BrenneckeGetVariableGenes</code> method is described in the <b style="color:#EB1960">M3Drop documentation</b> as follows:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/m3drop.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption"><code>BrenneckeGetVariableGenes</code></figcaption>
</figure>
</div>
<p>Thus we can see that we need the ERCCs to deduce the HGVs. However, I can’t find any data on the ERCC spike-ins, so at this point we stop here. Sorry to disappoint.</p>
<section id="references" class="level3">
<h3 class="anchored" data-anchor-id="references">References</h3>
<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-brennecke" class="csl-entry" role="listitem">
Brennecke, P., S. Anders, J. K. Kim, A. A. Kołodziejczyk, X. Zhang, V. Proserpio, B. Baying, et al. 2013. <span>“<span class="nocase"><span>A</span>ccounting for technical noise in single-cell <span>R</span><span>N</span><span>A</span>-seq experiments</span>.”</span> <em>Nat Methods</em> 10 (11): 1093–95.
</div>
<div id="ref-zebrafish-data" class="csl-entry" role="listitem">
Hernández, P. P., P. M. Strzelecka, E. I. Athanasiadis, D. Hall, A. F. Robalo, C. M. Collins, P. Boudinot, J. P. Levraud, and A. Cvejic. 2018. <span>“<span class="nocase"><span>S</span>ingle-cell transcriptional analysis reveals <span>I</span><span>L</span><span>C</span>-like cells in zebrafish</span>.”</span> <em>Science Immunology</em> 3 (29).
</div>
<div id="ref-scran" class="csl-entry" role="listitem">
Lun, A. T., K. Bach, and J. C. Marioni. 2016. <span>“<span class="nocase"><span>P</span>ooling across cells to normalize single-cell <span>R</span><span>N</span><span>A</span> sequencing data with many zero counts</span>.”</span> <em>Genome Biol</em> 17 (April): 75.
</div>
<div id="ref-dgTMatrix" class="csl-entry" role="listitem">
Slowikowski, Kamil. 2020. <span>“Working with a Sparse Matrix in r.”</span> <a href="https://slowkow.com/notes/sparse-matrix/">https://slowkow.com/notes/sparse-matrix/</a>.
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script>
    // Remove background
    elements = document.querySelectorAll('[id=clean-collapse]').forEach(
        function(element) {
            element.classList.remove("callout-style-default");
        }
    )
    
    // Add arrow to front to signify dropdownyness
    arrow = "▶";
    elements = document.querySelectorAll('.callout-caption-container').forEach(
        function(element) {
            element.innerHTML = arrow + element.innerHTML;
        }
    )
</script>
<script src="https://giscus.app/client.js" data-repo="baileyandrew/blog" data-repo-id="R_kgDOInJwKg" data-category="Announcements" data-category-id="DIC_kwDOInJwKs4CTGOQ" data-mapping="title" data-strict="0" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="dark_protanopia" data-lang="en" data-loading="lazy" crossorigin="anonymous" async="">
</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>