<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.27">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Bailey Andrew">
<meta name="dcterms.date" content="2023-01-10">
<meta name="description" content="A side quest in understanding how to work with genetic data.">

<title>Exploring Ensembl ‚Äì Lorem Ipsomething</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.min.js" integrity="sha384-ZvpUoO/+PpLXR1lu4jmpXWu80pZlYUAfxl5NsBMWOEPSjUn/6Z/hRTt8+pR6L4N2" crossorigin="anonymous"></script><script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-4d9afe2b8d18ee9fa5d0d57b5ed4214d.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-10892d65c249a5d8f93b275ef5a48de5.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdn.jsdelivr.net/npm/requirejs@2.3.6/require.min.js" integrity="sha384-c9c+LnTbwQ3aujuU7ULEPVvgLs+Fn6fJUvIGTsuu1ZcCf11fiEubah0ttpca4ntM sha384-6V1/AdqZRWk1KAlWbKBlGhN7VG4iE/yAZcO6NZPMF8od0vukrvr0tg4qY6NSrItx" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../styles.css">
<link rel="stylesheet" href="../../html_scripts/collapse.css">
<link rel="stylesheet" href="../../html_scripts/pretty_shortcuts.css">
<meta property="og:title" content="Exploring Ensembl ‚Äì Lorem Ipsomething">
<meta property="og:description" content="A side quest in understanding how to work with genetic data.">
<meta property="og:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/ensemble.png">
<meta property="og:site_name" content="Lorem Ipsomething">
<meta property="og:image:height" content="1462">
<meta property="og:image:width" content="1852">
<meta name="twitter:title" content="Exploring Ensembl ‚Äì Lorem Ipsomething">
<meta name="twitter:description" content="A side quest in understanding how to work with genetic data.">
<meta name="twitter:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/ensemble.png">
<meta name="twitter:image-height" content="1462">
<meta name="twitter:image-width" content="1852">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Lorem Ipsomething</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Exploring Ensembl</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Home</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../blog.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Blog</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../staticposts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Static Posts</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Puzzles</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../500_Puzzles/bnv.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Boulders in Valleys</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slideshowindex.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slideshows</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../otherblogs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Other People‚Äôs Blogs</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
        
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Exploring Ensembl</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Work</div>
    <div class="quarto-category">Useful</div>
    <div class="quarto-category">Omics</div>
  </div>
  </div>

<div>
  <div class="description">
    A side quest in understanding how to work with genetic data.
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Bailey Andrew </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 10, 2023</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p>So, I‚Äôve been working on <a href="./004_004_exploration.html">blog posts</a> that do a walkthrough on working with genetic data. I keep running into issues and having to go on <b class="sidequests">sidequests</b> (üòÖ) but I think this one deserves its own blog post. Shoutout to Morgan for the help.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/ensemble.png" class="img-fluid figure-img"></p>
<figcaption>The <b style="color:#EB1960">Ensemble</b> homepage</figcaption>
</figure>
</div>
<p>A major resource in bioinformatics is <b style="color:#EB1960">Ensembl</b>. In this blog post, we‚Äôll spend some time exploring it, capping off with using it to accomplish the useful task of how to grab all known mitochondrial genes in Zebrafish (<i style="color:#EB1960">Danio rerio</i>).</p>
<p>Suppose in our dataset we had a gene called <code>ENSDARG00000000001</code>. What can we actually say about it? Well, first of all, this is an <b style="color:#A6A440">Ensembl ID</b>, so it <a href="http://www.ensembl.org/info/genome/stable_ids/index.html">follows the pattern</a> <code>ENS[species prefix][feature type prefix][a unique eleven digit number]</code>. In this case, <code>DAR</code> is the species (<i style="color:#EB1960"><strong>Da</strong>nio <strong>r</strong>erio</i>) and <code>G</code> indicates that it is referring to a gene.</p>
<p>We can look this gene up on <b style="color:#EB1960">Ensembl</b> to learn more about it:</p>
<p><img src="./images/ensemble-0001.png" class="img-fluid"></p>
<p>This gene is also known as <b style="color:#A6A440">slc35a5</b>, but these names can be harder to work with as it is subject to change if our knowledge about its role in biological processes changes; the <b style="color:#A6A440">Ensembl ID</b> is more permanent. On the other hand, this name is arguably more informative, as it is an abbreviation for its role as ‚Äúsolute carrier family 35 member A5‚Äù.</p>
<p>If we want to investigate this gene more, we can click on the <b style="color:#EB1960">ZFIN</b> link <a href="http://zfin.org/ZDB-GENE-030616-55#summary">provided in the summary section</a>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/zfin.png" class="img-fluid figure-img"></p>
<figcaption><b style="color:#EB1960">ZFIN</b> page for ENSDARG00000000001</figcaption>
</figure>
</div>
<p>Here we can see yet another gene ID (<code>ZDB-GENE-030616-55</code>), as well as a link to the naming history of this gene which may be fun to explore. There are some other goodies on this page, but we‚Äôll return to <b style="color:#EB1960">Enbembl</b> as that is the topic of this post.</p>
<p>Our goal is to get a list of all mitochondrial genes in <i style="color:#EB1960">Danio rerio</i>. <code>ENSDARG00000000001</code> is not a mitochondrial gene, because it is located on chromosome 9. One way to find a list of mitochondrial genes is to search for genes with names beginning with <code>mt-</code>, because that is how mitochondrial genes are named.</p>
<p><img src="./images/mt-search.png" class="img-fluid"></p>
<p>However, this isn‚Äôt convenient for manual use!</p>
<details>
<summary style="color:#C0CF96">
<b>Failed attempt at using <code>scanpy</code></b>
</summary>
<p>One way to try to get this data is to use <a href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.queries.mitochondrial_genes.html">scanpy</a>:</p>
<div id="4de1bfc3-9e0c-4f1b-a134-3fcc71dd523e" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># I had to also install pybiomart, which was only on pip</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scanpy <span class="im">import</span> queries</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="ea82c0b6-7648-4fbc-bdf2-5d1605f7a08f" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>queries.mitochondrial_genes(</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"drerio"</span>,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    attrname <span class="op">=</span> <span class="st">"ensembl_gene_id"</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-error">
<div class="ansi-escaped-output">
<pre><span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">HTTPError</span>                                 Traceback (most recent call last)
Cell <span class="ansi-green-fg">In[3], line 1</span>
<span class="ansi-green-fg">----&gt; 1</span> <span class="ansi-yellow-bg">queries</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">mitochondrial_genes</span><span class="ansi-yellow-bg">(</span>
<span class="ansi-green-fg ansi-bold">      2</span> <span class="ansi-yellow-bg">    </span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">"</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">drerio</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">"</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg ansi-bold">      3</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">attrname</span><span class="ansi-yellow-bg"> </span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg"> </span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">"</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">ensembl_gene_id</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">"</span>
<span class="ansi-green-fg ansi-bold">      4</span> <span class="ansi-yellow-bg">)</span>

File <span class="ansi-green-fg">~/opt/anaconda3/envs/Blog-nbdev/lib/python3.9/site-packages/scanpy/queries/_queries.py:191</span>, in <span class="ansi-cyan-fg">mitochondrial_genes</span><span class="ansi-blue-fg">(org, attrname, host, use_cache, chromosome)</span>
<span class="ansi-green-fg ansi-bold">    156</span> <span style="color:rgb(175,0,255)">@_doc_params</span>(doc_org<span style="color:rgb(98,98,98)">=</span>_doc_org, doc_host<span style="color:rgb(98,98,98)">=</span>_doc_host, doc_use_cache<span style="color:rgb(98,98,98)">=</span>_doc_use_cache)
<span class="ansi-green-fg ansi-bold">    157</span> <span style="font-weight:bold;color:rgb(0,135,0)">def</span> <span style="color:rgb(0,0,255)">mitochondrial_genes</span>(
<span class="ansi-green-fg ansi-bold">    158</span>     org: <span style="color:rgb(0,135,0)">str</span>,
<span class="ansi-green-fg">   (...)</span>
<span class="ansi-green-fg ansi-bold">    163</span>     chromosome: <span style="color:rgb(0,135,0)">str</span> <span style="color:rgb(98,98,98)">=</span> <span style="color:rgb(175,0,0)">"</span><span style="color:rgb(175,0,0)">MT</span><span style="color:rgb(175,0,0)">"</span>,
<span class="ansi-green-fg ansi-bold">    164</span> ) <span style="color:rgb(98,98,98)">-</span><span style="color:rgb(98,98,98)">&gt;</span> pd<span style="color:rgb(98,98,98)">.</span>DataFrame:
<span class="ansi-green-fg ansi-bold">    165</span>     <span style="font-style:italic;color:rgb(175,0,0)">"""\</span>
<span class="ansi-green-fg ansi-bold">    166</span> <span style="font-style:italic;color:rgb(175,0,0)">    Mitochondrial gene symbols for specific organism through BioMart.</span>
<span class="ansi-green-fg ansi-bold">    167</span> 
<span class="ansi-green-fg">   (...)</span>
<span class="ansi-green-fg ansi-bold">    189</span> <span style="font-style:italic;color:rgb(175,0,0)">    &gt;&gt;&gt; mito_gene_names_fly = sc.queries.mitochondrial_genes("dmelanogaster", chromosome="mitochondrion_genome")</span>
<span class="ansi-green-fg ansi-bold">    190</span> <span style="font-style:italic;color:rgb(175,0,0)">    """</span>
<span class="ansi-green-fg">--&gt; 191</span>     <span style="font-weight:bold;color:rgb(0,135,0)">return</span> <span class="ansi-yellow-bg">simple_query</span><span class="ansi-yellow-bg">(</span>
<span class="ansi-green-fg ansi-bold">    192</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">org</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg ansi-bold">    193</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">attrs</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">[</span><span class="ansi-yellow-bg">attrname</span><span class="ansi-yellow-bg">]</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg ansi-bold">    194</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">filters</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">{</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">"</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">chromosome_name</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">"</span><span class="ansi-yellow-bg">:</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">[</span><span class="ansi-yellow-bg">chromosome</span><span class="ansi-yellow-bg">]</span><span class="ansi-yellow-bg">}</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg ansi-bold">    195</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">host</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">host</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg ansi-bold">    196</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">use_cache</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">use_cache</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg ansi-bold">    197</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">)</span>

File <span class="ansi-green-fg">~/opt/anaconda3/envs/Blog-nbdev/lib/python3.9/site-packages/scanpy/queries/_queries.py:67</span>, in <span class="ansi-cyan-fg">simple_query</span><span class="ansi-blue-fg">(org, attrs, filters, host, use_cache)</span>
<span class="ansi-green-fg ansi-bold">     63</span>     <span style="font-weight:bold;color:rgb(0,135,0)">raise</span> <span style="font-weight:bold;color:rgb(215,95,95)">ImportError</span>(
<span class="ansi-green-fg ansi-bold">     64</span>         <span style="color:rgb(175,0,0)">"</span><span style="color:rgb(175,0,0)">This method requires the `pybiomart` module to be installed.</span><span style="color:rgb(175,0,0)">"</span>
<span class="ansi-green-fg ansi-bold">     65</span>     )
<span class="ansi-green-fg ansi-bold">     66</span> server <span style="color:rgb(98,98,98)">=</span> Server(host, use_cache<span style="color:rgb(98,98,98)">=</span>use_cache)
<span class="ansi-green-fg">---&gt; 67</span> dataset <span style="color:rgb(98,98,98)">=</span> <span class="ansi-yellow-bg">server</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">marts</span>[<span style="color:rgb(175,0,0)">"</span><span style="color:rgb(175,0,0)">ENSEMBL_MART_ENSEMBL</span><span style="color:rgb(175,0,0)">"</span>]<span style="color:rgb(98,98,98)">.</span>datasets[
<span class="ansi-green-fg ansi-bold">     68</span>     <span style="color:rgb(175,0,0)">"</span><span style="font-weight:bold;color:rgb(175,95,135)">{}</span><span style="color:rgb(175,0,0)">_gene_ensembl</span><span style="color:rgb(175,0,0)">"</span><span style="color:rgb(98,98,98)">.</span>format(org)
<span class="ansi-green-fg ansi-bold">     69</span> ]
<span class="ansi-green-fg ansi-bold">     70</span> res <span style="color:rgb(98,98,98)">=</span> dataset<span style="color:rgb(98,98,98)">.</span>query(attributes<span style="color:rgb(98,98,98)">=</span>attrs, filters<span style="color:rgb(98,98,98)">=</span>filters, use_attr_names<span style="color:rgb(98,98,98)">=</span><span style="font-weight:bold;color:rgb(0,135,0)">True</span>)
<span class="ansi-green-fg ansi-bold">     71</span> <span style="font-weight:bold;color:rgb(0,135,0)">return</span> res

File <span class="ansi-green-fg">~/opt/anaconda3/envs/Blog-nbdev/lib/python3.9/site-packages/pybiomart/server.py:61</span>, in <span class="ansi-cyan-fg">Server.marts</span><span class="ansi-blue-fg">(self)</span>
<span class="ansi-green-fg ansi-bold">     59</span> <span style="font-style:italic;color:rgb(175,0,0)">"""List of available marts."""</span>
<span class="ansi-green-fg ansi-bold">     60</span> <span style="font-weight:bold;color:rgb(0,135,0)">if</span> <span style="color:rgb(0,135,0)">self</span><span style="color:rgb(98,98,98)">.</span>_marts <span style="font-weight:bold;color:rgb(175,0,255)">is</span> <span style="font-weight:bold;color:rgb(0,135,0)">None</span>:
<span class="ansi-green-fg">---&gt; 61</span>     <span style="color:rgb(0,135,0)">self</span><span style="color:rgb(98,98,98)">.</span>_marts <span style="color:rgb(98,98,98)">=</span> <span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">_fetch_marts</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg ansi-bold">     62</span> <span style="font-weight:bold;color:rgb(0,135,0)">return</span> <span style="color:rgb(0,135,0)">self</span><span style="color:rgb(98,98,98)">.</span>_marts

File <span class="ansi-green-fg">~/opt/anaconda3/envs/Blog-nbdev/lib/python3.9/site-packages/pybiomart/server.py:79</span>, in <span class="ansi-cyan-fg">Server._fetch_marts</span><span class="ansi-blue-fg">(self)</span>
<span class="ansi-green-fg ansi-bold">     78</span> <span style="font-weight:bold;color:rgb(0,135,0)">def</span> <span style="color:rgb(0,0,255)">_fetch_marts</span>(<span style="color:rgb(0,135,0)">self</span>):
<span class="ansi-green-fg">---&gt; 79</span>     response <span style="color:rgb(98,98,98)">=</span> <span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">get</span><span class="ansi-yellow-bg">(</span><span style="color:rgb(0,135,0)" class="ansi-yellow-bg">type</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">=</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">'</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">registry</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">'</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg ansi-bold">     81</span>     xml <span style="color:rgb(98,98,98)">=</span> xml_from_string(response<span style="color:rgb(98,98,98)">.</span>content)
<span class="ansi-green-fg ansi-bold">     82</span>     marts <span style="color:rgb(98,98,98)">=</span> [
<span class="ansi-green-fg ansi-bold">     83</span>         <span style="color:rgb(0,135,0)">self</span><span style="color:rgb(98,98,98)">.</span>_mart_from_xml(child)
<span class="ansi-green-fg ansi-bold">     84</span>         <span style="font-weight:bold;color:rgb(0,135,0)">for</span> child <span style="font-weight:bold;color:rgb(175,0,255)">in</span> xml<span style="color:rgb(98,98,98)">.</span>findall(<span style="color:rgb(175,0,0)">'</span><span style="color:rgb(175,0,0)">MartURLLocation</span><span style="color:rgb(175,0,0)">'</span>)
<span class="ansi-green-fg ansi-bold">     85</span>     ]

File <span class="ansi-green-fg">~/opt/anaconda3/envs/Blog-nbdev/lib/python3.9/site-packages/pybiomart/base.py:111</span>, in <span class="ansi-cyan-fg">ServerBase.get</span><span class="ansi-blue-fg">(self, **params)</span>
<span class="ansi-green-fg ansi-bold">    109</span>     <span style="font-weight:bold;color:rgb(0,135,0)">with</span> requests_cache<span style="color:rgb(98,98,98)">.</span>disabled():
<span class="ansi-green-fg ansi-bold">    110</span>         r <span style="color:rgb(98,98,98)">=</span> requests<span style="color:rgb(98,98,98)">.</span>get(<span style="color:rgb(0,135,0)">self</span><span style="color:rgb(98,98,98)">.</span>url, params<span style="color:rgb(98,98,98)">=</span>params)
<span class="ansi-green-fg">--&gt; 111</span> <span class="ansi-yellow-bg">r</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">raise_for_status</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg ansi-bold">    112</span> <span style="font-weight:bold;color:rgb(0,135,0)">return</span> r

File <span class="ansi-green-fg">~/opt/anaconda3/envs/Blog-nbdev/lib/python3.9/site-packages/requests/models.py:1021</span>, in <span class="ansi-cyan-fg">Response.raise_for_status</span><span class="ansi-blue-fg">(self)</span>
<span class="ansi-green-fg ansi-bold">   1016</span>     http_error_msg <span style="color:rgb(98,98,98)">=</span> (
<span class="ansi-green-fg ansi-bold">   1017</span>         <span style="color:rgb(175,0,0)">f</span><span style="color:rgb(175,0,0)">"</span><span style="font-weight:bold;color:rgb(175,95,135)">{</span><span style="color:rgb(0,135,0)">self</span><span style="color:rgb(98,98,98)">.</span>status_code<span style="font-weight:bold;color:rgb(175,95,135)">}</span><span style="color:rgb(175,0,0)"> Server Error: </span><span style="font-weight:bold;color:rgb(175,95,135)">{</span>reason<span style="font-weight:bold;color:rgb(175,95,135)">}</span><span style="color:rgb(175,0,0)"> for url: </span><span style="font-weight:bold;color:rgb(175,95,135)">{</span><span style="color:rgb(0,135,0)">self</span><span style="color:rgb(98,98,98)">.</span>url<span style="font-weight:bold;color:rgb(175,95,135)">}</span><span style="color:rgb(175,0,0)">"</span>
<span class="ansi-green-fg ansi-bold">   1018</span>     )
<span class="ansi-green-fg ansi-bold">   1020</span> <span style="font-weight:bold;color:rgb(0,135,0)">if</span> http_error_msg:
<span class="ansi-green-fg">-&gt; 1021</span>     <span style="font-weight:bold;color:rgb(0,135,0)">raise</span> HTTPError(http_error_msg, response<span style="color:rgb(98,98,98)">=</span><span style="color:rgb(0,135,0)">self</span>)

<span class="ansi-red-fg">HTTPError</span>: 500 Server Error: Internal Server Error for url: http://www.ensembl.org:80/biomart/martservice?type=registry</pre>
</div>
</div>
</div>
<p>Unfortunately, <b style="color:#EB1960">Biomart</b> was down at the time I tried to do this! (<b style="color:#EB1960">Biomart</b> seems to be the api for this type of stuff).</p>
<blockquote class="blockquote">
<p>Use this data-mining tool to export custom datasets from Ensembl.</p>
<p>‚Äì <cite><a href="http://www.ensembl.org/info/docs/tools/index.html">Ensemble docs about what Biomart is</a></cite></p>
</blockquote>
<p>I found that the ‚Äúasia‚Äù mirror of <b style="color:#EB1960">Biomart</b> gives a different error:</p>
<div id="27165d57-7f03-46c0-bc88-a60533c73fd9" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>mitgenes <span class="op">=</span> queries.mitochondrial_genes(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"drerio"</span>,</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    attrname <span class="op">=</span> <span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    host <span class="op">=</span> <span class="st">"asia.ensembl.org"</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>mitgenes</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-error">
<div class="ansi-escaped-output">
<pre><span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">HTTPError</span>                                 Traceback (most recent call last)
Cell <span class="ansi-green-fg">In[13], line 1</span>
<span class="ansi-green-fg">----&gt; 1</span> mitgenes <span style="color:rgb(98,98,98)">=</span> <span class="ansi-yellow-bg">queries</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">mitochondrial_genes</span><span class="ansi-yellow-bg">(</span>
<span class="ansi-green-fg ansi-bold">      2</span> <span class="ansi-yellow-bg">    </span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">"</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">drerio</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">"</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg ansi-bold">      3</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">attrname</span><span class="ansi-yellow-bg"> </span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg"> </span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">"</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">ensembl_gene_id</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">"</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg ansi-bold">      4</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">host</span><span class="ansi-yellow-bg"> </span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg"> </span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">"</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">asia.ensembl.org</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">"</span>
<span class="ansi-green-fg ansi-bold">      5</span> <span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg ansi-bold">      6</span> mitgenes

File <span class="ansi-green-fg">~/opt/anaconda3/envs/Blog-nbdev/lib/python3.9/site-packages/scanpy/queries/_queries.py:191</span>, in <span class="ansi-cyan-fg">mitochondrial_genes</span><span class="ansi-blue-fg">(org, attrname, host, use_cache, chromosome)</span>
<span class="ansi-green-fg ansi-bold">    156</span> <span style="color:rgb(175,0,255)">@_doc_params</span>(doc_org<span style="color:rgb(98,98,98)">=</span>_doc_org, doc_host<span style="color:rgb(98,98,98)">=</span>_doc_host, doc_use_cache<span style="color:rgb(98,98,98)">=</span>_doc_use_cache)
<span class="ansi-green-fg ansi-bold">    157</span> <span style="font-weight:bold;color:rgb(0,135,0)">def</span> <span style="color:rgb(0,0,255)">mitochondrial_genes</span>(
<span class="ansi-green-fg ansi-bold">    158</span>     org: <span style="color:rgb(0,135,0)">str</span>,
<span class="ansi-green-fg">   (...)</span>
<span class="ansi-green-fg ansi-bold">    163</span>     chromosome: <span style="color:rgb(0,135,0)">str</span> <span style="color:rgb(98,98,98)">=</span> <span style="color:rgb(175,0,0)">"</span><span style="color:rgb(175,0,0)">MT</span><span style="color:rgb(175,0,0)">"</span>,
<span class="ansi-green-fg ansi-bold">    164</span> ) <span style="color:rgb(98,98,98)">-</span><span style="color:rgb(98,98,98)">&gt;</span> pd<span style="color:rgb(98,98,98)">.</span>DataFrame:
<span class="ansi-green-fg ansi-bold">    165</span>     <span style="font-style:italic;color:rgb(175,0,0)">"""\</span>
<span class="ansi-green-fg ansi-bold">    166</span> <span style="font-style:italic;color:rgb(175,0,0)">    Mitochondrial gene symbols for specific organism through BioMart.</span>
<span class="ansi-green-fg ansi-bold">    167</span> 
<span class="ansi-green-fg">   (...)</span>
<span class="ansi-green-fg ansi-bold">    189</span> <span style="font-style:italic;color:rgb(175,0,0)">    &gt;&gt;&gt; mito_gene_names_fly = sc.queries.mitochondrial_genes("dmelanogaster", chromosome="mitochondrion_genome")</span>
<span class="ansi-green-fg ansi-bold">    190</span> <span style="font-style:italic;color:rgb(175,0,0)">    """</span>
<span class="ansi-green-fg">--&gt; 191</span>     <span style="font-weight:bold;color:rgb(0,135,0)">return</span> <span class="ansi-yellow-bg">simple_query</span><span class="ansi-yellow-bg">(</span>
<span class="ansi-green-fg ansi-bold">    192</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">org</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg ansi-bold">    193</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">attrs</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">[</span><span class="ansi-yellow-bg">attrname</span><span class="ansi-yellow-bg">]</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg ansi-bold">    194</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">filters</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">{</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">"</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">chromosome_name</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">"</span><span class="ansi-yellow-bg">:</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg">[</span><span class="ansi-yellow-bg">chromosome</span><span class="ansi-yellow-bg">]</span><span class="ansi-yellow-bg">}</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg ansi-bold">    195</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">host</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">host</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg ansi-bold">    196</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg">use_cache</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">=</span><span class="ansi-yellow-bg">use_cache</span><span class="ansi-yellow-bg">,</span>
<span class="ansi-green-fg ansi-bold">    197</span> <span class="ansi-yellow-bg">    </span><span class="ansi-yellow-bg">)</span>

File <span class="ansi-green-fg">~/opt/anaconda3/envs/Blog-nbdev/lib/python3.9/site-packages/scanpy/queries/_queries.py:67</span>, in <span class="ansi-cyan-fg">simple_query</span><span class="ansi-blue-fg">(org, attrs, filters, host, use_cache)</span>
<span class="ansi-green-fg ansi-bold">     63</span>     <span style="font-weight:bold;color:rgb(0,135,0)">raise</span> <span style="font-weight:bold;color:rgb(215,95,95)">ImportError</span>(
<span class="ansi-green-fg ansi-bold">     64</span>         <span style="color:rgb(175,0,0)">"</span><span style="color:rgb(175,0,0)">This method requires the `pybiomart` module to be installed.</span><span style="color:rgb(175,0,0)">"</span>
<span class="ansi-green-fg ansi-bold">     65</span>     )
<span class="ansi-green-fg ansi-bold">     66</span> server <span style="color:rgb(98,98,98)">=</span> Server(host, use_cache<span style="color:rgb(98,98,98)">=</span>use_cache)
<span class="ansi-green-fg">---&gt; 67</span> dataset <span style="color:rgb(98,98,98)">=</span> <span class="ansi-yellow-bg">server</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">marts</span>[<span style="color:rgb(175,0,0)">"</span><span style="color:rgb(175,0,0)">ENSEMBL_MART_ENSEMBL</span><span style="color:rgb(175,0,0)">"</span>]<span style="color:rgb(98,98,98)">.</span>datasets[
<span class="ansi-green-fg ansi-bold">     68</span>     <span style="color:rgb(175,0,0)">"</span><span style="font-weight:bold;color:rgb(175,95,135)">{}</span><span style="color:rgb(175,0,0)">_gene_ensembl</span><span style="color:rgb(175,0,0)">"</span><span style="color:rgb(98,98,98)">.</span>format(org)
<span class="ansi-green-fg ansi-bold">     69</span> ]
<span class="ansi-green-fg ansi-bold">     70</span> res <span style="color:rgb(98,98,98)">=</span> dataset<span style="color:rgb(98,98,98)">.</span>query(attributes<span style="color:rgb(98,98,98)">=</span>attrs, filters<span style="color:rgb(98,98,98)">=</span>filters, use_attr_names<span style="color:rgb(98,98,98)">=</span><span style="font-weight:bold;color:rgb(0,135,0)">True</span>)
<span class="ansi-green-fg ansi-bold">     71</span> <span style="font-weight:bold;color:rgb(0,135,0)">return</span> res

File <span class="ansi-green-fg">~/opt/anaconda3/envs/Blog-nbdev/lib/python3.9/site-packages/pybiomart/server.py:61</span>, in <span class="ansi-cyan-fg">Server.marts</span><span class="ansi-blue-fg">(self)</span>
<span class="ansi-green-fg ansi-bold">     59</span> <span style="font-style:italic;color:rgb(175,0,0)">"""List of available marts."""</span>
<span class="ansi-green-fg ansi-bold">     60</span> <span style="font-weight:bold;color:rgb(0,135,0)">if</span> <span style="color:rgb(0,135,0)">self</span><span style="color:rgb(98,98,98)">.</span>_marts <span style="font-weight:bold;color:rgb(175,0,255)">is</span> <span style="font-weight:bold;color:rgb(0,135,0)">None</span>:
<span class="ansi-green-fg">---&gt; 61</span>     <span style="color:rgb(0,135,0)">self</span><span style="color:rgb(98,98,98)">.</span>_marts <span style="color:rgb(98,98,98)">=</span> <span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">_fetch_marts</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg ansi-bold">     62</span> <span style="font-weight:bold;color:rgb(0,135,0)">return</span> <span style="color:rgb(0,135,0)">self</span><span style="color:rgb(98,98,98)">.</span>_marts

File <span class="ansi-green-fg">~/opt/anaconda3/envs/Blog-nbdev/lib/python3.9/site-packages/pybiomart/server.py:79</span>, in <span class="ansi-cyan-fg">Server._fetch_marts</span><span class="ansi-blue-fg">(self)</span>
<span class="ansi-green-fg ansi-bold">     78</span> <span style="font-weight:bold;color:rgb(0,135,0)">def</span> <span style="color:rgb(0,0,255)">_fetch_marts</span>(<span style="color:rgb(0,135,0)">self</span>):
<span class="ansi-green-fg">---&gt; 79</span>     response <span style="color:rgb(98,98,98)">=</span> <span style="color:rgb(0,135,0)" class="ansi-yellow-bg">self</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">get</span><span class="ansi-yellow-bg">(</span><span style="color:rgb(0,135,0)" class="ansi-yellow-bg">type</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">=</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">'</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">registry</span><span style="color:rgb(175,0,0)" class="ansi-yellow-bg">'</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg ansi-bold">     81</span>     xml <span style="color:rgb(98,98,98)">=</span> xml_from_string(response<span style="color:rgb(98,98,98)">.</span>content)
<span class="ansi-green-fg ansi-bold">     82</span>     marts <span style="color:rgb(98,98,98)">=</span> [
<span class="ansi-green-fg ansi-bold">     83</span>         <span style="color:rgb(0,135,0)">self</span><span style="color:rgb(98,98,98)">.</span>_mart_from_xml(child)
<span class="ansi-green-fg ansi-bold">     84</span>         <span style="font-weight:bold;color:rgb(0,135,0)">for</span> child <span style="font-weight:bold;color:rgb(175,0,255)">in</span> xml<span style="color:rgb(98,98,98)">.</span>findall(<span style="color:rgb(175,0,0)">'</span><span style="color:rgb(175,0,0)">MartURLLocation</span><span style="color:rgb(175,0,0)">'</span>)
<span class="ansi-green-fg ansi-bold">     85</span>     ]

File <span class="ansi-green-fg">~/opt/anaconda3/envs/Blog-nbdev/lib/python3.9/site-packages/pybiomart/base.py:111</span>, in <span class="ansi-cyan-fg">ServerBase.get</span><span class="ansi-blue-fg">(self, **params)</span>
<span class="ansi-green-fg ansi-bold">    109</span>     <span style="font-weight:bold;color:rgb(0,135,0)">with</span> requests_cache<span style="color:rgb(98,98,98)">.</span>disabled():
<span class="ansi-green-fg ansi-bold">    110</span>         r <span style="color:rgb(98,98,98)">=</span> requests<span style="color:rgb(98,98,98)">.</span>get(<span style="color:rgb(0,135,0)">self</span><span style="color:rgb(98,98,98)">.</span>url, params<span style="color:rgb(98,98,98)">=</span>params)
<span class="ansi-green-fg">--&gt; 111</span> <span class="ansi-yellow-bg">r</span><span style="color:rgb(98,98,98)" class="ansi-yellow-bg">.</span><span class="ansi-yellow-bg">raise_for_status</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-fg ansi-bold">    112</span> <span style="font-weight:bold;color:rgb(0,135,0)">return</span> r

File <span class="ansi-green-fg">~/opt/anaconda3/envs/Blog-nbdev/lib/python3.9/site-packages/requests/models.py:1021</span>, in <span class="ansi-cyan-fg">Response.raise_for_status</span><span class="ansi-blue-fg">(self)</span>
<span class="ansi-green-fg ansi-bold">   1016</span>     http_error_msg <span style="color:rgb(98,98,98)">=</span> (
<span class="ansi-green-fg ansi-bold">   1017</span>         <span style="color:rgb(175,0,0)">f</span><span style="color:rgb(175,0,0)">"</span><span style="font-weight:bold;color:rgb(175,95,135)">{</span><span style="color:rgb(0,135,0)">self</span><span style="color:rgb(98,98,98)">.</span>status_code<span style="font-weight:bold;color:rgb(175,95,135)">}</span><span style="color:rgb(175,0,0)"> Server Error: </span><span style="font-weight:bold;color:rgb(175,95,135)">{</span>reason<span style="font-weight:bold;color:rgb(175,95,135)">}</span><span style="color:rgb(175,0,0)"> for url: </span><span style="font-weight:bold;color:rgb(175,95,135)">{</span><span style="color:rgb(0,135,0)">self</span><span style="color:rgb(98,98,98)">.</span>url<span style="font-weight:bold;color:rgb(175,95,135)">}</span><span style="color:rgb(175,0,0)">"</span>
<span class="ansi-green-fg ansi-bold">   1018</span>     )
<span class="ansi-green-fg ansi-bold">   1020</span> <span style="font-weight:bold;color:rgb(0,135,0)">if</span> http_error_msg:
<span class="ansi-green-fg">-&gt; 1021</span>     <span style="font-weight:bold;color:rgb(0,135,0)">raise</span> HTTPError(http_error_msg, response<span style="color:rgb(98,98,98)">=</span><span style="color:rgb(0,135,0)">self</span>)

<span class="ansi-red-fg">HTTPError</span>: 504 Server Error: Gateway Time-out for url: http://asia.ensembl.org:80/biomart/martservice?type=registry</pre>
</div>
</div>
</div>
</details>
<p>Another way to get the data is to search by location. We can play around with their region-searcher to see that the mitochondrial dna of a zebrafish is just over 16 kilobases long (16,596 bases to be exact).</p>
<p><img src="./images/loc-search.png" class="img-fluid"></p>
<p>If we do that, we can find this cute overview of the mitochondrial dna:</p>
<p><img src="./images/mitdna.png" class="img-fluid"></p>
<p>Not relevant to us, though - we want to click on the ‚Äúexport‚Äù button, which will give us this popup:</p>
<p><img src="./images/export.png" class="img-fluid"></p>
<p>You can then download all the mitochondrial data! It‚Äôs a rather small file that looks like this:</p>
<pre><code>seqname,source,feature,start,end,score,strand,frame,hid,hstart,hend,genscan,gene_id,transcript_id,exon_id,gene_type,variation_name,probe_name
MT,EVA,variation,113,113,.,+,.,,,,,,,,,rs508804888,
[...]
MT,Ensembl,gene,951,1019,.,+,0,,,,,ENSDARG00000083480.3,ENSDART00000116162.3,ENSDARE00000880048,Mt_tRNA,,
[...]</code></pre>
<div id="cb112adf-d06d-4c31-9b5c-c0a6f363df2d" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="9947df64-280c-4fef-ae39-8e29f35c1d57" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>mito_data <span class="op">=</span> pd.read_csv(<span class="st">"./localdata/mt-danio-rerio.txt"</span>, sep<span class="op">=</span><span class="st">","</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>mito_data</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">seqname</th>
<th data-quarto-table-cell-role="th">source</th>
<th data-quarto-table-cell-role="th">feature</th>
<th data-quarto-table-cell-role="th">start</th>
<th data-quarto-table-cell-role="th">end</th>
<th data-quarto-table-cell-role="th">score</th>
<th data-quarto-table-cell-role="th">strand</th>
<th data-quarto-table-cell-role="th">frame</th>
<th data-quarto-table-cell-role="th">hid</th>
<th data-quarto-table-cell-role="th">hstart</th>
<th data-quarto-table-cell-role="th">hend</th>
<th data-quarto-table-cell-role="th">genscan</th>
<th data-quarto-table-cell-role="th">gene_id</th>
<th data-quarto-table-cell-role="th">transcript_id</th>
<th data-quarto-table-cell-role="th">exon_id</th>
<th data-quarto-table-cell-role="th">gene_type</th>
<th data-quarto-table-cell-role="th">variation_name</th>
<th data-quarto-table-cell-role="th">probe_name</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th">0</th>
<td>MT</td>
<td>EVA</td>
<td>variation</td>
<td>113</td>
<td>113</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>rs508804888</td>
<td>NaN</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">1</th>
<td>MT</td>
<td>EVA</td>
<td>variation</td>
<td>239</td>
<td>239</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>rs513784503</td>
<td>NaN</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">2</th>
<td>MT</td>
<td>EVA</td>
<td>variation</td>
<td>314</td>
<td>314</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>rs504526537</td>
<td>NaN</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">3</th>
<td>MT</td>
<td>EVA</td>
<td>variation</td>
<td>339</td>
<td>339</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>rs511578098</td>
<td>NaN</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">4</th>
<td>MT</td>
<td>EVA</td>
<td>variation</td>
<td>438</td>
<td>438</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>rs514602887</td>
<td>NaN</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">...</th>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">174</th>
<td>MT</td>
<td>Ensembl</td>
<td>gene</td>
<td>14714</td>
<td>15232</td>
<td>.</td>
<td>-</td>
<td>0</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>ENSDARG00000063922.3</td>
<td>ENSDART00000093623.3</td>
<td>ENSDARE00000685766</td>
<td>protein_coding</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">175</th>
<td>MT</td>
<td>Ensembl</td>
<td>gene</td>
<td>15233</td>
<td>15301</td>
<td>.</td>
<td>-</td>
<td>0</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>ENSDARG00000083312.3</td>
<td>ENSDART00000116823.3</td>
<td>ENSDARE00000882905</td>
<td>Mt_tRNA</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">176</th>
<td>MT</td>
<td>Ensembl</td>
<td>gene</td>
<td>15308</td>
<td>16448</td>
<td>.</td>
<td>+</td>
<td>0</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>ENSDARG00000063924.3</td>
<td>ENSDART00000093625.3</td>
<td>ENSDARE00000685768</td>
<td>protein_coding</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">177</th>
<td>MT</td>
<td>Ensembl</td>
<td>gene</td>
<td>16449</td>
<td>16520</td>
<td>.</td>
<td>+</td>
<td>0</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>ENSDARG00000083462.3</td>
<td>ENSDART00000116552.3</td>
<td>ENSDARE00000881627</td>
<td>Mt_tRNA</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">178</th>
<td>MT</td>
<td>Ensembl</td>
<td>gene</td>
<td>16527</td>
<td>16596</td>
<td>.</td>
<td>-</td>
<td>1</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>ENSDARG00000081475.3</td>
<td>ENSDART00000115546.3</td>
<td>ENSDARE00000882281</td>
<td>Mt_tRNA</td>
<td>NaN</td>
<td>NaN</td>
</tr>
</tbody>
</table>

<p>179 rows √ó 18 columns</p>
</div>
</div>
</div>
<div id="62e34077-80f4-4971-a9f6-603af56493da" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Grab all the genes, get their ids, and chop off the '.3' </span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># ending which indicates the version</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>mito_genes <span class="op">=</span> mito_data[mito_data[<span class="st">"feature"</span>] <span class="op">==</span> <span class="st">"gene"</span>][<span class="st">"gene_id"</span>].<span class="bu">apply</span>(</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    <span class="kw">lambda</span> x: x[:<span class="op">-</span><span class="dv">2</span>]</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>mito_genes</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>142    ENSDARG00000083480
143    ENSDARG00000082753
144    ENSDARG00000081443
145    ENSDARG00000080337
146    ENSDARG00000083046
147    ENSDARG00000063895
148    ENSDARG00000083118
149    ENSDARG00000080630
150    ENSDARG00000082084
151    ENSDARG00000063899
152    ENSDARG00000080718
153    ENSDARG00000080401
154    ENSDARG00000081938
155    ENSDARG00000082789
156    ENSDARG00000080128
157    ENSDARG00000063905
158    ENSDARG00000081369
159    ENSDARG00000083519
160    ENSDARG00000063908
161    ENSDARG00000080151
162    ENSDARG00000063910
163    ENSDARG00000063911
164    ENSDARG00000063912
165    ENSDARG00000081758
166    ENSDARG00000063914
167    ENSDARG00000080329
168    ENSDARG00000063916
169    ENSDARG00000063917
170    ENSDARG00000082716
171    ENSDARG00000082123
172    ENSDARG00000081280
173    ENSDARG00000063921
174    ENSDARG00000063922
175    ENSDARG00000083312
176    ENSDARG00000063924
177    ENSDARG00000083462
178    ENSDARG00000081475
Name: gene_id, dtype: object</code></pre>
</div>
</div>
<div id="c44a29d9-d933-46e4-8de8-43139b7efd20" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>mito_genes.to_csv(<span class="st">"./localdata/mito-genes.csv"</span>, index<span class="op">=</span><span class="va">False</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>So, uh, yeah - success!</p>



</main> <!-- /main -->
<script>
    // Remove background
    elements = document.querySelectorAll('[id=clean-collapse]').forEach(
        function(element) {
            element.classList.remove("callout-style-default");
        }
    )
    
    // Add arrow to front to signify dropdownyness
    arrow = "‚ñ∂";
    elements = document.querySelectorAll('.callout-caption-container').forEach(
        function(element) {
            element.innerHTML = arrow + element.innerHTML;
        }
    )
</script>
<script src="https://giscus.app/client.js" data-repo="baileyandrew/blog" data-repo-id="R_kgDOInJwKg" data-category="Announcements" data-category-id="DIC_kwDOInJwKs4CTGOQ" data-mapping="title" data-strict="0" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="dark_protanopia" data-lang="en" data-loading="lazy" crossorigin="anonymous" async="">
</script>
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "Óßã";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/BaileyAndrew\.github\.io\/Blog");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions"><ul><li><a href="https://github.com/BaileyAndrew/Blog/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>