<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Bailey Andrew">
<meta name="dcterms.date" content="2023-01-11">
<meta name="description" content="An overview of the different sequencing protocols in scRNA-seq.">

<title>Lorem Ipsomething - Overview of scRNA-seq Data Creation – Sequencing Protocol</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
<link rel="stylesheet" href="../../html_scripts/collapse.css">
<link rel="stylesheet" href="../../html_scripts/pretty_shortcuts.css">
<meta property="og:title" content="Lorem Ipsomething - Overview of scRNA-seq Data Creation – Sequencing Protocol">
<meta property="og:description" content="An overview of the different sequencing protocols in scRNA-seq.">
<meta property="og:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/why-amplification.jpeg">
<meta property="og:site-name" content="Lorem Ipsomething">
<meta name="twitter:title" content="Lorem Ipsomething - Overview of scRNA-seq Data Creation – Sequencing Protocol">
<meta name="twitter:description" content="An overview of the different sequencing protocols in scRNA-seq.">
<meta name="twitter:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/why-amplification.jpeg">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Lorem Ipsomething</span>
    </a>
  </div>
        <div class="quarto-navbar-tools ms-auto">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Overview of scRNA-seq Data Creation – Sequencing Protocol</li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Home</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../blog.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Blog</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../staticposts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Static Posts</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Puzzles</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../500_Puzzles/bnv.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Boulders in Valleys</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slideshowindex.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slideshows</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../otherblogs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Other People’s Blogs</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#what-is-a-protocol" id="toc-what-is-a-protocol" class="nav-link active" data-scroll-target="#what-is-a-protocol">What is a <b style="color:#A6A440">protocol</b>?</a></li>
  <li><a href="#details-about-protocols" id="toc-details-about-protocols" class="nav-link" data-scroll-target="#details-about-protocols">Details about <b style="color:#A6A440">protocols</b></a>
  <ul class="collapse">
  <li><a href="#reverse-transcription" id="toc-reverse-transcription" class="nav-link" data-scroll-target="#reverse-transcription"><b style="color:#537FBF">Reverse Transcription</b></a></li>
  <li><a href="#nd-strand-synthesis" id="toc-nd-strand-synthesis" class="nav-link" data-scroll-target="#nd-strand-synthesis"><b style="color:#537FBF">2nd Strand Synthesis</b></a></li>
  <li><a href="#amplification" id="toc-amplification" class="nav-link" data-scroll-target="#amplification"><b style="color:#537FBF">Amplification</b></a></li>
  <li><a href="#pcr" id="toc-pcr" class="nav-link" data-scroll-target="#pcr"><b style="color:#537FBF">PCR</b></a></li>
  <li><a href="#ivt" id="toc-ivt" class="nav-link" data-scroll-target="#ivt"><b style="color:#537FBF">IVT</b></a></li>
  <li><a href="#rna-fragmentation" id="toc-rna-fragmentation" class="nav-link" data-scroll-target="#rna-fragmentation"><b style="color:#537FBF">RNA fragmentation</b></a></li>
  <li><a href="#tagmentation" id="toc-tagmentation" class="nav-link" data-scroll-target="#tagmentation"><b style="color:#537FBF">Tagmentation</b></a></li>
  <li><a href="#primers" id="toc-primers" class="nav-link" data-scroll-target="#primers"><b style="color:#A6A440">Primers</b></a></li>
  <li><a href="#enrichment" id="toc-enrichment" class="nav-link" data-scroll-target="#enrichment"><b style="color:#537FBF">3’ Enrichment</b></a></li>
  <li><a href="#adapter-ligation" id="toc-adapter-ligation" class="nav-link" data-scroll-target="#adapter-ligation"><b style="color:#537FBF">Adapter Ligation</b></a></li>
  <li><a href="#umi-vs-full-length" id="toc-umi-vs-full-length" class="nav-link" data-scroll-target="#umi-vs-full-length"><b style="color:#A6A440">UMI</b> vs <b style="color:#A6A440">Full Length</b></a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/BaileyAndrew/Blog/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Overview of scRNA-seq Data Creation – Sequencing Protocol</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Work</div>
    <div class="quarto-category">Omics</div>
    <div class="quarto-category">Useful</div>
  </div>
  </div>

<div>
  <div class="description">
    An overview of the different sequencing protocols in scRNA-seq.
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Bailey Andrew </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 11, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p>I’ve seen a few “<b style="color:#A6A440">protocols</b>” pop up, such as <b style="color:#A6A440">Smart-seq2</b> and <b style="color:#A6A440">10x Chromium</b>. This blog aims to answer the following questions:</p>
<ul>
<li>What are protocols</li>
<li>What affect do they have on downstream analysis</li>
<li>Which are the common protocols</li>
</ul>
<p>I am indebted to <span class="citation" data-cites="protocols">Ziegenhain et al. (<a href="#ref-protocols" role="doc-biblioref">2017</a>)</span> for much of my understanding.</p>
<section id="what-is-a-protocol" class="level1">
<h1>What is a <b style="color:#A6A440">protocol</b>?</h1>
<p>An algorithm for biologicians; it’s a recorded sequence of steps to follow in the lab to produce scRNA-seq data.</p>
</section>
<section id="details-about-protocols" class="level1">
<h1>Details about <b style="color:#A6A440">protocols</b></h1>
<blockquote class="blockquote">
<p>One type of technical variable is the <b style="color:#A6A440">sensitivity</b> of a scRNA-seq method (i.e., <b style="color:#C0CF96">the probability to capture and convert a particular mRNA transcript present in a single cell into a cDNA molecule present in the library</b>). Another variable of interest is the <b style="color:#A6A440">accuracy</b> (i.e., <b style="color:#C0CF96">how well the read quantification corresponds to the actual concentration of mRNAs</b>), and a third type is the <b style="color:#A6A440">precision with which this amplification occurs</b> (i.e., the technical variation of the quantification). The combination of <b style="color:#A6A440">sensitivity</b>, <b style="color:#A6A440">precision</b>, and number of cells analyzed determines the power to detect relative differences in expression levels.</p>
<p>– <span class="citation" data-cites="protocols">Ziegenhain et al. (<a href="#ref-protocols" role="doc-biblioref">2017</a>)</span></p>
</blockquote>
<p>The difference between <b style="color:#A6A440">sensitivity</b> and <b style="color:#A6A440">accuracy</b> is a bit confusing. I <strong style="color:#757575">think</strong> that <b style="color:#A6A440">sensitivity</b> here is on a per-trascript level (how likely it is to be captured) whereas the <b style="color:#A6A440">accuracy</b> is about the relative proportions of the transcripts. If you had 5 instances of <b style="color:#EB1960">mt-co1</b> and 10 instances of <b style="color:#EB1960">mt-cyb</b> in your results but had 10 <b style="color:#EB1960">mt-co1</b>/20 <b style="color:#EB1960">mt-cyb</b> in reality, the precision would be 50% but the accuracy would be 100%. I’m not sure what they mean by <b style="color:#A6A440">precision</b>, but I feel confident in assuming it’s the probability that a reported transcript was actually present in the sample.</p>
<details>
<summary>
<b class="sidequests">Sidequest</b><b style="color:#C0CF96">: What is <b style="color:#EB1960">mt-co1</b>?</b>
</summary>
<p>I’ve been trying to “<b style="color:#C0CF96">think more like a bioinformatician</b>”. Now, I don’t really think that bioinformatician spend much time learning about random genes, but I think the tools I have to use to learn about them will be tools that bioinformaticians use and hence learning about random genes will still be useful for me. <b style="color:#EB1960">mt-co1</b> is a random gene, which is why I picked it out here. I specifically picked it because it’s present in zebrafish (<i style="color:#EB1960">Danio rerio</i>), which is what I was reading about a lot lately. However, it is also present in humans and mice. In fact it’s probably present in a lot of places, this was just what came up on the first page of my <b style="color:#EB1960">Ensembl</b> search.</p>
<ul>
<li>For zebrafish, the <b style="color:#A6A440">Ensembl ID</b> is <b style="color:#757575">ENSDARG00000063905</b></li>
<li>For humans, the <b style="color:#A6A440">Ensembl ID</b> is <b style="color:#757575">ENSG00000198804</b></li>
<li>For mice, the <b style="color:#A6A440">Ensembl ID</b> is <b style="color:#757575">ENSMUSG00000064351</b></li>
</ul>
<p>In general it has quite a lot of <b style="color:#A6A440">orthologs</b> (genes in similar species sharing a common ancestor). However, there are also quite a lot of species (98) for which <b style="color:#EB1960">Ensembl</b> reports no <b style="color:#A6A440">orthologs</b>, such as <i style="color:#EB1960">Balaenoptera musculus</i>, the blue whale. Unfortunately, it’s not clear to me whether <b style="color:#EB1960">Ensembl</b> reports no orthologs because it is confident none exist, or because it does not have the data; there are 314 species in <b style="color:#EB1960">Ensembl</b> as far as I can tell, 200 of which have <b style="color:#A6A440">orthologs</b> and 98 of which don’t. Presumably the remaining 16 are unsure? But it’d be weird that such a high percentage of species, spread out over many domains in the tree of life, lack a mitochondrial gene that is very widespread. I’m not sure.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/human-zebra-m1-co1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Human vs zebrafish mitochondrial dna near <b style="color:#EB1960">mt-co1</b></figcaption>
</figure>
</div>
<p>We can see that this region seems to be <em style="color:#C0CF96">highly</em> conserved between species; all the protein-coding genes are in the same order along the chromosome. Zebrafish and humans are just about as far away as you can get among vertebrates so if it’s conserved amongst them, we’d expect it to be conserved amongst much closer relatives as well.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/whale-genome.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption"><b style="color:#EB1960">Ensembl</b> genome for the blue whale</figcaption>
</figure>
</div>
<p>The above graphic indicates that <b style="color:#EB1960">Ensembl</b> does not have the mitochondrial DNA for the blue whale (or else we’d expect an “MT” chromosome like in humans). This seals the deal for me that a lack of <b style="color:#A6A440">orthologs</b> in <b style="color:#EB1960">Ensembl</b> does not denote confidence in a lack of <b style="color:#A6A440">orthologs</b> irl, merely that no orthologs have been found.</p>
<p>Despite my current fascination with zebrafish, I’ll mostly talk about its role in humans (but again, it’s probably quite similar).</p>
<p>To continue, <b style="color:#EB1960">mt-co1</b> stands for “mitochondrially encoded cytochrome c oxidase I”.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/variants.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Table of possible gene variants</figcaption>
</figure>
</div>
<p>The above table is quite interesting; it’s a list of known variants to the gene that have occured. I’ll go through column-by-column to explain what they are.</p>
<ul>
<li><b style="color:#537FBF">Variant ID</b>: The unique identifier of this gene variant.</li>
<li><b style="color:#537FBF">Chr: bp</b>: The location, in terms of base pairs, along the chromosome.</li>
<li><b style="color:#537FBF">Alleles</b>: The base pair before the slash is the default, the ones after the slash are possible mutations.</li>
<li><b style="color:#537FBF">Global MAF</b>: “The frequency of the second most common allele in the global population, defined in human by the 1000 Genomes Project phase 3.</li>
<li><b style="color:#537FBF">Class</b>: The type of variant it is (SNP or SNV for example)</li>
<li><b style="color:#537FBF">Source</b>: The database this is from</li>
<li><b style="color:#537FBF">Evidence</b>: Evidence for the existence of this variant</li>
<li><b style="color:#537FBF">Clinical Significance</b>: Clinical significance</li>
<li><b style="color:#537FBF">Consequence Type</b>: What will happen when the cell/mitochondria try to transcribe this DNA? How will it affect the mRNA and the protein?</li>
<li><b style="color:#537FBF">AA</b>: The resulting amino acid (before the slash is original, after is the variant - it’s possible for synonymous variants to exist in which case there’s no slash).</li>
<li><b style="color:#537FBF">AA coord</b>: Not sure, but I think this is its location in the chain of amino acids in the generated protein.</li>
<li><b style="color:#537FBF">SIFT</b>: Score of likelihood for whether we predict the function of the generated protein to change.</li>
<li><b style="color:#537FBF">Poly-Phen</b>: (I think “phen” is “phenotype”, as in “multiple phenotypes”). Another score of likelihood of predicted function change, calculated by looking at different characteristics.</li>
<li><b style="color:#537FBF">CADD</b>: Scores the “deleteriousness” of SNPs</li>
<li><b style="color:#537FBF">REVEL</b>: Predicts “pathogeneity” of of SNVs</li>
<li><b style="color:#537FBF">Meta LR</b>: Same as above</li>
<li><b style="color:#537FBF">Mutation Assessor</b>: Similar to <b style="color:#537FBF">SIFT</b> and <b style="color:#537FBF">Poly-Phen</b>. For SNVs.</li>
<li><b style="color:#537FBF">Transcript</b>: How the gene manifests in RNA</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/variant-info.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">We can look at more details by clicking on the variant ID! But this is for another time.</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/transcript-info.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">And more details about the transcript! But this is also for another time.</figcaption>
</figure>
</div>
<p>The stuff about clinical significance might be quite useful to look at for future projects.</p>
<p>We could look into ontology terms and pathways as well, but I’ve spent quite a long time on this <b class="sidequests">sidequest</b>.</p>
</details>
<details>
<summary>
<b class="sidequests">Sidequest</b><b style="color:#C0CF96">: What is <b style="color:#EB1960">mt-cyb</b>?</b>
</summary>
<p><b style="color:#EB1960">mt-cyb</b>, or <b style="color:#757575">ENSG00000198727</b>, or even “mitochondrially encoded cytochrome b”, is another mitochondrial gene. It is involved in “respiratory electron transport”, metabolism, and <b style="color:#A6A440">the citric acid cycle</b>, but has no more known pathways in <b style="color:#EB1960">Ensembl</b>. All of the aforementioned pathways were also shared by <b style="color:#EB1960">mt-co1</b>, but <b style="color:#EB1960">mt-co1</b> had much more pathways overall.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/phenotypes.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The phenotypes page on <b style="color:#EB1960">Ensembl</b></figcaption>
</figure>
</div>
We can see that variants in this gene can be related to certain diseases. Based on the GO terms, it is an “integral component of membrane”. That’s all for now, folks!
</details>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/protocol-compare.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">A graphical overview from <span class="citation" data-cites="protocols">Ziegenhain et al. (<a href="#ref-protocols" role="doc-biblioref">2017</a>)</span>. Not my own work.</figcaption>
</figure>
</div>
<p>There’s a lot to unpack in this image. I’ll only pick out a few.</p>
<section id="reverse-transcription" class="level2">
<h2 class="anchored" data-anchor-id="reverse-transcription"><b style="color:#537FBF">Reverse Transcription</b></h2>
<p>The creation of cDNA from RNA (called so because DNA-&gt;RNA is <b style="color:#537FBF">transcription</b>). cDNA is more stable than RNA, so it’s easier to work with.</p>
</section>
<section id="nd-strand-synthesis" class="level2">
<h2 class="anchored" data-anchor-id="nd-strand-synthesis"><b style="color:#537FBF">2nd Strand Synthesis</b></h2>
<p>The synthesis of the second strand of cDNA or RNA (😉). (Since <b style="color:#537FBF">reverse transcription</b> is just creating one strand of cDNA)</p>
</section>
<section id="amplification" class="level2">
<h2 class="anchored" data-anchor-id="amplification"><b style="color:#537FBF">Amplification</b></h2>
<p>Because of limited <b style="color:#A6A440">read length</b>, <b style="color:#537FBF">fragmentation</b> is required (see the section on it for details). However, it can be hard to reconstruct the original transcript with confidence. If we <b style="color:#537FBF">amplify</b> the RNA/cDNA first then after <b style="color:#537FBF">fragmentation</b> this problem will be mitigated. <b style="color:#537FBF">Amplification</b> is just the process of creating many copies of RNA/cDNA already extant in your sample.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/why-amplification.jpeg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption"><b style="color:#537FBF">Amplification</b> is beneficial because it will give us more <b style="color:#A6A440">reads</b> to work with during <b style="color:#537FBF">sequencing</b>, if <b style="color:#537FBF">fragmentation</b> is performed.</figcaption>
</figure>
</div>
<p>Fun fact: you don’t need amplification when using <b style="color:#A6A440">nanopores</b>.</p>
<blockquote class="blockquote">
<p>The facility of <b style="color:#A6A440">nanopore</b> technology to analyse native DNA, without the requirement for <b style="color:#537FBF">amplification</b>, eliminates <b style="color:#A6A440">PCR bias</b> and allows the identification of base modifications alongside nucleotide sequence — with no requirement for time-consuming, harsh, and, often inefficient, chemical conversion (e.g.&nbsp;bisulfite conversion).</p>
<p>– <b style="color:#EB1960">Oxford Nanopore, about their products</b></p>
</blockquote>
</section>
<section id="pcr" class="level2">
<h2 class="anchored" data-anchor-id="pcr"><b style="color:#537FBF">PCR</b></h2>
<p><b style="color:#537FBF">PCR</b> stands for <b style="color:#537FBF">Polymerase Chain Reaction</b>.</p>
<blockquote class="blockquote">
<p>Sometimes called “molecular photocopying,” the <b style="color:#537FBF">polymerase chain reaction (PCR)</b> is a fast and inexpensive technique used to “<b style="color:#537FBF">amplify</b>” - copy - small segments of DNA.</p>
<p>– <b style="color:#EB1960"><span class="citation" data-cites="pcr"><span>“Polymerase Chain Reaction Fact Sheet”</span> (<a href="#ref-pcr" role="doc-biblioref">2020</a>)</span></b></p>
</blockquote>
<blockquote class="blockquote">
<p>In the <b style="color:#537FBF">PCR method</b>, a pair of <b style="color:#A6A440">primers</b> hybridizes with the sample DNA and defines the region that will be <b style="color:#537FBF">amplified</b>, resulting in millions and millions of copies in a very short timeframe.</p>
<p>– <b style="color:#EB1960"><span class="citation" data-cites="primers">Shchelochkov (<a href="#ref-primers" role="doc-biblioref">2023</a>)</span></b></p>
</blockquote>
</section>
<section id="ivt" class="level2">
<h2 class="anchored" data-anchor-id="ivt"><b style="color:#537FBF">IVT</b></h2>
<p><b style="color:#537FBF">IVT</b> stands for <b style="color:#537FBF">In Vitro Transcription</b>, and is a method of <b style="color:#537FBF">amplification</b> by making the previously produced cDNA molecules be transcribed into RNA. Note that this is why <b style="color:#EB1960">CEL-seq2/C1</b> and <b style="color:#EB1960">MARS-seq</b> in the diagram above go from RNA to cDNA and then back to RNA (because they both use <b style="color:#537FBF">IVT</b>.</p>
<blockquote class="blockquote">
<p><b style="color:#537FBF">In vitro transcription</b> is a simple procedure that allows for template-directed synthesis of RNA molecules of any sequence from short <b style="color:#A6A440">oligonucleotides</b> to those of several kilobases in μg to mg quantities. It is based on the engineering of a template that includes a <b style="color:#A6A440">bacteriophage promoter sequence</b> (e.g.&nbsp;from the T7 coliphage) upstream of the sequence of interest followed by transcription using the corresponding RNA polymerase.</p>
<p>– <b style="color:#EB1960"><span class="citation" data-cites="ivt">Beckert and Masquida (<a href="#ref-ivt" role="doc-biblioref">2011</a>)</span></b></p>
</blockquote>
</section>
<section id="rna-fragmentation" class="level2">
<h2 class="anchored" data-anchor-id="rna-fragmentation"><b style="color:#537FBF">RNA fragmentation</b></h2>
<blockquote class="blockquote">
<p>After poly(A) + selection or rRNA depletion, RNA samples are typically subject to RNA fragmentation to a certain size range before RT. <strong style="color:#C0CF96">This is necessary because of the size limitation of most current sequencing platforms</strong>, e.g., &lt;600 bp on Illumina sequencers.</p>
<p>– <b style="color:#EB1960"> Fragmentation Subsection; General Aspects OF RNA-Seq Section; <span class="citation" data-cites="rna-seq-methods">Hrdlickova, Toloue, and Tian (<a href="#ref-rna-seq-methods" role="doc-biblioref">2017</a>)</span> </b></p>
</blockquote>
<p>Fragmentation is done when there is a limited <b style="color:#A6A440">read length</b> in your <b style="color:#A6A440">flow cell</b>. If we did not fragment the RNA, then we would need large read length to read the base pairs in the middle of a long RNA strand. <b style="color:#537FBF">Fragmentation</b> is not necessary if your read length is longer than the largest RNA molecule (or unlimited). <b style="color:#537FBF">Fragmentation</b> puts extra strain on downstream analysis, because we need to reconstruct the original sequence from the fragments. Furthermore, <b style="color:#537FBF">fragmentation</b> isn’t random; the techniques used to break up the RNA may preferentially cause breaks in some places compared to others. <span class="citation" data-cites="rna-seq-methods">Hrdlickova, Toloue, and Tian (<a href="#ref-rna-seq-methods" role="doc-biblioref">2017</a>)</span> talk about about the biases for different <b style="color:#537FBF">fragmentation</b> methods.</p>
<p><b style="color:#537FBF">Fragmentation</b> can also be used on the cDNA, but it can be harder to automate. However, <b style="color:#537FBF">tagmentation</b> is done on cDNA.</p>
</section>
<section id="tagmentation" class="level2">
<h2 class="anchored" data-anchor-id="tagmentation"><b style="color:#537FBF">Tagmentation</b></h2>
<blockquote class="blockquote">
<p>Recent development in using transposon-based, so-called <b style="color:#537FBF">tagmentation</b> method has made it simple to <b style="color:#537FBF">fragment cDNA</b> and <strong style="color:#537FBF">add adapter sequences at the same time</strong>. In this method, an active variant of the Tn5 transposase mediates the <b style="color:#537FBF">fragmentation</b> of double-stranded DNA and <b style="color:#537FBF">ligates</b> adapter <b style="color:#A6A440">oligonucleotides</b> at both ends in a quick reaction (~5 min). However, it is notable that Tn5 and other enzyme-based cDNA fragmentation methods require a precise enzyme:DNA ratio, <strong style="color:#C0CF96">making method optimization less straightforward than RNA fragmentation</strong>. Consequently, fragmenting RNA is currently still the most frequently used approach in RNA-Seq library preparation.</p>
<p>– <b style="color:#EB1960"> Fragmentation Subsection; General Aspects OF RNA-Seq Section; <span class="citation" data-cites="rna-seq-methods">Hrdlickova, Toloue, and Tian (<a href="#ref-rna-seq-methods" role="doc-biblioref">2017</a>)</span> </b></p>
</blockquote>
<p><b style="color:#537FBF">Ligation</b> is the joining of two nucleic acid fragments through the action of an enzyme - at least, according to Wikipedia. It seems that this is just the combination of <b style="color:#537FBF">fragmentation</b> and tagging with the adapter <b style="color:#A6A440">oligonucleotides</b>. These processes used to need to be done sequentially.</p>
<details>
<summary>
<b style="color:#C0CF96">Why do we want to tag the adapter with</b> <b style="color:#A6A440">oligonucleotides</b>?
</summary>
<p>The tags are necessary to attach <b style="color:#A6A440">primers</b> to your cDNA <span class="citation" data-cites="tag-why">(<a href="#ref-tag-why" role="doc-biblioref">Salamat 2018</a>)</span>. This is described well during the following explanation of <b style="color:#537FBF">Index PCR</b>:</p>
<blockquote class="blockquote">
<p>A single Illumina <b style="color:#A6A440">flow cell</b> can sequence multiple samples as long as the expected reads will not saturate (exceed) the capacity of the flow cell. However, since the samples are pooled and loaded as one into the flow cell, <b style="color:#C0CF96">there must be a mechanism to which to distinguish sequences from one sample to another</b>. This is accomplished via a process called <b style="color:#537FBF">Index PCR</b>. Here, custom <b style="color:#A6A440">oligonucleotides</b> called <b style="color:#A6A440">index primers</b> are used to <b style="color:#537FBF">amplify</b> and ‘barcode’ the fragments. Each <b style="color:#A6A440">index primer</b> contains the following: * <b style="color:#A6A440">Read (1 or 2) sequencing primer</b>: <b style="color:#C0CF96">a segment complementary to the ‘tag’ introduced during tagmentation</b>, * <b style="color:#A6A440">Indexing sequence</b>: a unique DNA sequence for identification / barcoding of samples * <b style="color:#A6A440">Sequencing Anchor (P5/P7)</b>: sequences complementary to the oligos in the flow cell. <b style="color:#C0CF96">These allow the index PCR products (libraries) to bind to the Illumina flow cell for sequencing</b>.</p>
<p><b style="color:#EB1960">STEP 4: LIBRARY PREPARATION: AMPLIFICATION; <span class="citation" data-cites="tag-why">Salamat (<a href="#ref-tag-why" role="doc-biblioref">2018</a>)</span></b></p>
</blockquote>
<p>For an explanation on all the details here, see the section on <b style="color:#537FBF">Index PCR</b>; essentially, the tags are necessary to hook an <b style="color:#A6A440">index primer</b> into your cDNA, which is then necessary to hook it up with the <b style="color:#A6A440">oligos</b> in your <b style="color:#A6A440">flow cell</b>.</p>
</details>
</section>
<section id="primers" class="level2">
<h2 class="anchored" data-anchor-id="primers"><b style="color:#A6A440">Primers</b></h2>
<blockquote class="blockquote">
<p>A <b style="color:#A6A440">primer</b>, as related to genomics, is a short single-stranded DNA fragment used in certain laboratory techniques, such as the <b style="color:#537FBF">polymerase chain reaction (PCR)</b>. In the <b style="color:#537FBF">PCR method</b>, a pair of <b style="color:#A6A440">primers</b> hybridizes with the sample DNA and defines the region that will be <b style="color:#537FBF">amplified</b>, resulting in millions and millions of copies in a very short timeframe. <b style="color:#A6A440">Primers</b> are also used in DNA sequencing and other experimental processes.</p>
<p>– <b style="color:#EB1960"><span class="citation" data-cites="primers">Shchelochkov (<a href="#ref-primers" role="doc-biblioref">2023</a>)</span></b></p>
</blockquote>
<p><b style="color:#A6A440">Primers</b> are used to attach to the R/DNA in as a marker for <b style="color:#537FBF">amplification</b>.</p>
</section>
<section id="enrichment" class="level2">
<h2 class="anchored" data-anchor-id="enrichment"><b style="color:#537FBF">3’ Enrichment</b></h2>
<p><b style="color:#537FBF">Enrichment</b> is the process of increasing the proportion of the <b style="color:#A6A440">thingamajig of interest</b> (RNA/cDNA in this case) relative to the rest of stuff in the sample. It differs from <b style="color:#537FBF">amplification</b> in that it’s not increasing the <b style="color:#A6A440">thingamajig of interest</b> but rather getting rid of the <b style="color:#A6A440">thingamajigs of disinterest</b>. (So it’s more like <b style="color:#537FBF">depoorment</b>). <b style="color:#537FBF">3’ enrichment</b> specifically utilizes the 3’ end of the RNA/cDNA to do the <b style="color:#537FBF">enrichment</b>.</p>
</section>
<section id="adapter-ligation" class="level2">
<h2 class="anchored" data-anchor-id="adapter-ligation"><b style="color:#537FBF">Adapter Ligation</b></h2>
<blockquote class="blockquote">
<p>In a standard RNA-Seq library <b style="color:#A6A440">protocol</b>, cDNAs of a desired size generated from <b style="color:#537FBF">RT</b> of fragmented RNAs with random hexamer <b style="color:#A6A440">primers</b> or from fragmented <b style="color:#A6A440">full-length cDNAs</b> are <b style="color:#537FBF">ligated</b> to DNA adapters before amplification and sequencing. While simple, <b style="color:#C0CF96">this approach loses the information about which DNA strand corresponds to the sense strand of RNA</b>. Lack of strand specificity would make it difficult to identify <b style="color:#A6A440">antisense</b> and novel RNA species and cause inaccurate measurement of sense RNA expression. Several methods have been developed to capture the directionality of RNA in cDNA libraries.</p>
<p>– <b style="color:#EB1960"> Adapters and Directionality Subsection; General Aspects OF RNA-Seq Section; <span class="citation" data-cites="rna-seq-methods">Hrdlickova, Toloue, and Tian (<a href="#ref-rna-seq-methods" role="doc-biblioref">2017</a>)</span> </b></p>
</blockquote>
<p><b style="color:#537FBF">RT</b> stands for <b style="color:#537FBF">Reverse Transcription</b>, described earlier. <b style="color:#A6A440">Antisense RNA</b> is RNA that is complementary to mRNA. <b style="color:#A6A440">Adapters</b> are useful for sequencing as they allow the transcripts to bind to the <b style="color:#A6A440">oligos</b> in <b style="color:#537FBF">SBS</b> <b style="color:#A6A440">flow cells</b>.</p>
</section>
<section id="umi-vs-full-length" class="level2">
<h2 class="anchored" data-anchor-id="umi-vs-full-length"><b style="color:#A6A440">UMI</b> vs <b style="color:#A6A440">Full Length</b></h2>
<blockquote class="blockquote">
<p><b style="color:#A6A440">UMI</b> is an acronym for <b style="color:#A6A440">Unique Molecular Identifier</b>. <b style="color:#A6A440">UMIs</b> are complex indices added to sequencing libraries before any <b style="color:#537FBF">PCR amplification</b> steps, enabling the accurate bioinformatic identification of <b style="color:#537FBF">PCR</b> duplicates. <b style="color:#A6A440">UMIs</b> are also known as “Molecular Barcodes” or “Random Barcodes”</p>
<p>– <b style="color:#EB1960"><span class="citation" data-cites="umi">Tech (<a href="#ref-umi" role="doc-biblioref">n.d.</a>)</span></b></p>
</blockquote>
<p><b style="color:#A6A440">UMIs</b> are applied <strong style="color:#C0CF96">to individual strands of cDNA</strong> - i.e they are not applied to identify cells (<b style="color:#537FBF">barcoding</b>) or samples (<b style="color:#A6A440">Unique Dual Indices (UDIs)</b>), but individual transcripts of the original RNA. This can help accurately piece <b style="color:#A6A440">reads</b> together when <b style="color:#537FBF">sequencing</b>. The <b style="color:#A6A440">UMI</b> will only be on one end of the strand and hence many <b style="color:#A6A440">reads</b> will not contain the <b style="color:#A6A440">UMI</b> on them, but when piecing them together post-<b style="color:#537FBF">fragmentation</b> it can provide more information on what can be matched with what.</p>
<blockquote class="blockquote">
<p>The main disadvantage of both <b style="color:#EB1960">Smart-seq</b> <b style="color:#A6A440">protocols</b> is that the generation of <b style="color:#A6A440">full-length</b> cDNA libraries precludes an early <b style="color:#537FBF">barcoding</b> step and the incorporation of <b style="color:#A6A440">UMIs</b>.</p>
<p>– <b style="color:#EB1960"> Fragmentation Subsection; General Aspects OF RNA-Seq Section; <span class="citation" data-cites="rna-seq-methods">Hrdlickova, Toloue, and Tian (<a href="#ref-rna-seq-methods" role="doc-biblioref">2017</a>)</span> </b></p>
</blockquote>
<blockquote class="blockquote">
<p>Other protocols have sacrificed full-length coverage in order to sequence part of the <b style="color:#A6A440">primer</b> used for cDNA generation. This enables early <b style="color:#537FBF">barcoding</b> of libraries (i.e., the incorporation of cell-specific barcodes), allowing for multiplexing the cDNA <b style="color:#537FBF">amplification</b> and thereby increasing the throughput of scRNA-seq library generation by one to three orders of magnitude</p>
<p>– <b style="color:#EB1960"> Introduction Section; <span class="citation" data-cites="rna-seq-methods">Hrdlickova, Toloue, and Tian (<a href="#ref-rna-seq-methods" role="doc-biblioref">2017</a>)</span> </b></p>
</blockquote>
<p>Not all methods give full coverage of the transcripts; when they do, it is called <b style="color:#A6A440">full length</b>.</p>
<div class="References">

</div>



</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-ivt" class="csl-entry" role="listitem">
Beckert, Bertrand, and Benoît Masquida. 2011. <span>“Synthesis of RNA by in Vitro Transcription.”</span> <em>Methods Mol Biol.</em> <a href="https://pubmed.ncbi.nlm.nih.gov/21125481/">https://pubmed.ncbi.nlm.nih.gov/21125481/</a>.
</div>
<div id="ref-rna-seq-methods" class="csl-entry" role="listitem">
Hrdlickova, Radmila, Masoud Toloue, and Bin Tian. 2017. <span>“RNA-Seq Methods for Transcriptome Analysis.”</span> <em>Wiley Interdiscip Rev RNA</em>.
</div>
<div id="ref-pcr" class="csl-entry" role="listitem">
<span>“Polymerase Chain Reaction Fact Sheet.”</span> 2020. National Human Gene Research Institute. <a href="https://www.genome.gov/about-genomics/fact-sheets/Polymerase-Chain-Reaction-Fact-Sheet">https://www.genome.gov/about-genomics/fact-sheets/Polymerase-Chain-Reaction-Fact-Sheet</a>.
</div>
<div id="ref-tag-why" class="csl-entry" role="listitem">
Salamat, Ezra. 2018. <span>“DNA SEQUENCING: ILLUMINA WORKFLOW PART-2 (DNA TAGMENTATION, AMPLIFICATION, AND CLEANUP).”</span> <a href="https://upcvmda-pl480.weebly.com/educational-articles/dna-sequencing-illumina-workflow-part-2-dna-tagmentation-amplification-and-cleanup">https://upcvmda-pl480.weebly.com/educational-articles/dna-sequencing-illumina-workflow-part-2-dna-tagmentation-amplification-and-cleanup</a>.
</div>
<div id="ref-primers" class="csl-entry" role="listitem">
Shchelochkov, Oleg. 2023. <span>“Primer.”</span> National Human Gene Research Institute. <a href="https://www.genome.gov/genetics-glossary/Primer">https://www.genome.gov/genetics-glossary/Primer</a>.
</div>
<div id="ref-umi" class="csl-entry" role="listitem">
Tech, DNA. n.d. <span>“What Are UMIs and Why Are They Used in High-Throughput Sequencing?”</span> UC Davis. <a href="https://dnatech.genomecenter.ucdavis.edu/faqs/what-are-umis-and-why-are-they-used-in-high-throughput-sequencing/">https://dnatech.genomecenter.ucdavis.edu/faqs/what-are-umis-and-why-are-they-used-in-high-throughput-sequencing/</a>.
</div>
<div id="ref-protocols" class="csl-entry" role="listitem">
Ziegenhain, Christoph, Beate Vieth, Swati Parekh, Björn Reinius, Amy Guillaumet-Adkins, Martha Smets, Heinrich Leonhardt, Holger Heyn, Ines Hellmann, and Wolfgang Enard. 2017. <span>“Comparative Analysis of Single-Cell RNA Sequencing Methods.”</span> <em>Molecular Cell</em> 65 (4): 631–643.e4. https://doi.org/<a href="https://doi.org/10.1016/j.molcel.2017.01.023">https://doi.org/10.1016/j.molcel.2017.01.023</a>.
</div>
</div></section></div></main> <!-- /main -->
<script>
    // Remove background
    elements = document.querySelectorAll('[id=clean-collapse]').forEach(
        function(element) {
            element.classList.remove("callout-style-default");
        }
    )
    
    // Add arrow to front to signify dropdownyness
    arrow = "▶";
    elements = document.querySelectorAll('.callout-caption-container').forEach(
        function(element) {
            element.innerHTML = arrow + element.innerHTML;
        }
    )
</script>
<script src="https://giscus.app/client.js" data-repo="baileyandrew/blog" data-repo-id="R_kgDOInJwKg" data-category="Announcements" data-category-id="DIC_kwDOInJwKs4CTGOQ" data-mapping="title" data-strict="0" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="dark_protanopia" data-lang="en" data-loading="lazy" crossorigin="anonymous" async="">
</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>