<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.27">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Bailey Andrew">
<meta name="dcterms.date" content="2023-01-20">
<meta name="description" content="Another scRNA paper">

<title>A Living Biobank of Ovarian Cancer – Figure 5a – Lorem Ipsomething</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.min.js" integrity="sha384-ZvpUoO/+PpLXR1lu4jmpXWu80pZlYUAfxl5NsBMWOEPSjUn/6Z/hRTt8+pR6L4N2" crossorigin="anonymous"></script><script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-4d9afe2b8d18ee9fa5d0d57b5ed4214d.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-10892d65c249a5d8f93b275ef5a48de5.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdn.jsdelivr.net/npm/requirejs@2.3.6/require.min.js" integrity="sha384-c9c+LnTbwQ3aujuU7ULEPVvgLs+Fn6fJUvIGTsuu1ZcCf11fiEubah0ttpca4ntM sha384-6V1/AdqZRWk1KAlWbKBlGhN7VG4iE/yAZcO6NZPMF8od0vukrvr0tg4qY6NSrItx" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../styles.css">
<link rel="stylesheet" href="../../html_scripts/collapse.css">
<link rel="stylesheet" href="../../html_scripts/pretty_shortcuts.css">
<meta property="og:title" content="A Living Biobank of Ovarian Cancer – Figure 5a – Lorem Ipsomething">
<meta property="og:description" content="Another scRNA paper">
<meta property="og:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/tsne-reproduce-ovarian.png">
<meta property="og:site_name" content="Lorem Ipsomething">
<meta property="og:image:height" content="838">
<meta property="og:image:width" content="842">
<meta name="twitter:title" content="A Living Biobank of Ovarian Cancer – Figure 5a – Lorem Ipsomething">
<meta name="twitter:description" content="Another scRNA paper">
<meta name="twitter:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/tsne-reproduce-ovarian.png">
<meta name="twitter:image-height" content="838">
<meta name="twitter:image-width" content="842">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed slimcontent quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Lorem Ipsomething</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">A Living Biobank of Ovarian Cancer – Figure 5a</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Home</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../blog.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Blog</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../staticposts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Static Posts</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Puzzles</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../500_Puzzles/bnv.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Boulders in Valleys</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slideshowindex.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slideshows</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../otherblogs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Other People’s Blogs</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#filtering-and-normalization" id="toc-filtering-and-normalization" class="nav-link active" data-scroll-target="#filtering-and-normalization"><b style="color:#537FBF">Filtering</b> and <b style="color:#537FBF">Normalization</b></a>
  <ul class="collapse">
  <li><a href="#library-size-filtering" id="toc-library-size-filtering" class="nav-link" data-scroll-target="#library-size-filtering"><b style="color:#A6A440">Library Size</b> <b style="color:#537FBF">Filtering</b></a></li>
  <li><a href="#gene-count-filtering" id="toc-gene-count-filtering" class="nav-link" data-scroll-target="#gene-count-filtering"><b style="color:#A6A440">Gene Count</b> <b style="color:#537FBF">Filtering</b></a></li>
  <li><a href="#mitochondrial-filtering" id="toc-mitochondrial-filtering" class="nav-link" data-scroll-target="#mitochondrial-filtering">Mitochondrial <b style="color:#537FBF">Filtering</b></a></li>
  <li><a href="#multiplets" id="toc-multiplets" class="nav-link" data-scroll-target="#multiplets"><b style="color:#A6A440">Multiplets</b></a></li>
  <li><a href="#normalization-and-miscellaneous-filtering" id="toc-normalization-and-miscellaneous-filtering" class="nav-link" data-scroll-target="#normalization-and-miscellaneous-filtering"><b style="color:#537FBF">Normalization</b> and miscellaneous <b style="color:#537FBF">filtering</b></a></li>
  </ul></li>
  <li><a href="#highly-variable-genes" id="toc-highly-variable-genes" class="nav-link" data-scroll-target="#highly-variable-genes"><b style="color:#A6A440">Highly Variable Genes</b></a></li>
  <li><a href="#figure-5a" id="toc-figure-5a" class="nav-link" data-scroll-target="#figure-5a"><b style="color:#EB1960">Figure 5a</b></a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/BaileyAndrew/Blog/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">A Living Biobank of Ovarian Cancer – Figure 5a</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Work</div>
    <div class="quarto-category">Omics</div>
    <div class="quarto-category">Useful</div>
    <div class="quarto-category">Paper Replication</div>
  </div>
  </div>

<div>
  <div class="description">
    Another scRNA paper
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Bailey Andrew </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 20, 2023</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p>Last time on the blog, we were looking at the paper “<b style="color:#EB1960">A living biobank of ovarian cancer ex vivo models reveals profound mitotic heterogeneity</b>” <span class="citation" data-cites="ovarian-cancer-living-biobank">(<a href="#ref-ovarian-cancer-living-biobank" role="doc-biblioref">Nelson et al. 2020</a>)</span>. We’ll pick up where we left off, this time trying to recreate <b style="color:#EB1960">Figure 5</b>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/ovarian-fig5.png" class="img-fluid figure-img"></p>
<figcaption><b style="color:#EB1960">Figure 5</b>. Not my own work</figcaption>
</figure>
</div>
<details>
<summary>
<b class="setups">Recap</b><b style="color:#C0CF96">: Loading the Data</b>
</summary>
<div id="de5f06e8-0b9d-493e-ab38-f26a3bec1b03" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>file.path <span class="ot">=</span> <span class="st">'./localdata/E-MTAB-8559-quantification-raw-files/'</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="6a2958cc-8bec-4162-9a22-5e70d6cac08e" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load in the matrix of expected raw counts</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>raw.counts <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">readMM</span>(</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste</span>(</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>        file.path,</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">'E-MTAB-8559.aggregated_filtered_counts.mtx'</span>,</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">sep=</span><span class="st">''</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Load in the row names (genes)</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>row.info <span class="ot">&lt;-</span> <span class="fu">read.table</span>(</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste</span>(</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>        file.path,</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>        <span class="st">'E-MTAB-8559.aggregated_filtered_counts.mtx_rows'</span>,</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">sep=</span><span class="st">''</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">header=</span><span class="cn">FALSE</span>,</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">col.names=</span><span class="fu">c</span>(<span class="st">"Ensembl.ID"</span>, <span class="st">"Redundant"</span>)</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">all</span>(row.info[<span class="st">'Ensembl.ID'</span>] <span class="sc">!=</span> row.info[<span class="st">'Redundant'</span>])) {</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>    <span class="co"># I don't understand why these datasets tend to write</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>    <span class="co"># the same name twice in each row???</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="st">"Warning: second column is not actually redundant!"</span>)</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Drop duplicate column from genes</span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>row.info <span class="ot">&lt;-</span> row.info[<span class="st">'Ensembl.ID'</span>]</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Load in the column names (single cells)</span></span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>col.info <span class="ot">&lt;-</span> <span class="fu">read.table</span>(</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste</span>(</span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>        file.path,</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>        <span class="st">'E-MTAB-8559.aggregated_filtered_counts.mtx_cols'</span>,</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>        <span class="at">sep=</span><span class="st">''</span></span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>    <span class="at">header=</span><span class="cn">FALSE</span>,</span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a>    <span class="at">sep=</span><span class="st">'-'</span>,</span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a>    <span class="at">col.names=</span><span class="fu">c</span>(<span class="st">'Sample.Info'</span>, <span class="st">'Cell.Barcode'</span>)</span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-42"><a href="#cb2-42" aria-hidden="true" tabindex="-1"></a>col.info[<span class="st">'Full.ID'</span>] <span class="ot">&lt;-</span> <span class="fu">apply</span>(</span>
<span id="cb2-43"><a href="#cb2-43" aria-hidden="true" tabindex="-1"></a>    col.info,</span>
<span id="cb2-44"><a href="#cb2-44" aria-hidden="true" tabindex="-1"></a>    <span class="dv">1</span>, <span class="co"># Rows</span></span>
<span id="cb2-45"><a href="#cb2-45" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(row) {</span>
<span id="cb2-46"><a href="#cb2-46" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste</span>(</span>
<span id="cb2-47"><a href="#cb2-47" aria-hidden="true" tabindex="-1"></a>        row[<span class="st">'Sample.Info'</span>],</span>
<span id="cb2-48"><a href="#cb2-48" aria-hidden="true" tabindex="-1"></a>        row[<span class="st">'Cell.Barcode'</span>],</span>
<span id="cb2-49"><a href="#cb2-49" aria-hidden="true" tabindex="-1"></a>        <span class="at">sep=</span><span class="st">'-'</span></span>
<span id="cb2-50"><a href="#cb2-50" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-51"><a href="#cb2-51" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb2-52"><a href="#cb2-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-53"><a href="#cb2-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Load in the experimental design matrix</span></span>
<span id="cb2-54"><a href="#cb2-54" aria-hidden="true" tabindex="-1"></a>exp.design.table <span class="ot">&lt;-</span> <span class="fu">read.table</span>(</span>
<span id="cb2-55"><a href="#cb2-55" aria-hidden="true" tabindex="-1"></a>    <span class="st">'./localdata/ExpDesign-E-MTAB-8559.tsv'</span>,</span>
<span id="cb2-56"><a href="#cb2-56" aria-hidden="true" tabindex="-1"></a>    <span class="at">header=</span><span class="cn">TRUE</span>,</span>
<span id="cb2-57"><a href="#cb2-57" aria-hidden="true" tabindex="-1"></a>    <span class="at">sep=</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span></span>
<span id="cb2-58"><a href="#cb2-58" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</details>
<section id="filtering-and-normalization" class="level1 page-columns page-full">
<h1><b style="color:#537FBF">Filtering</b> and <b style="color:#537FBF">Normalization</b></h1>
<div id="fccf8a9f-0871-46c3-a55e-1eee56fe3e48" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(raw.counts)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>23284</li><li>19880</li></ol>
</div>
</div>
<p>There are 23,284 genes and 19,880 cells - which is quite a lot! So we should probably <b style="color:#537FBF">filter</b> some things out.</p>
<blockquote class="blockquote">
<p>We used three measures to identify and remove the low-quality cells. Namely, the <strong style="color:#A6A440">library size</strong>; the <strong style="color:#C0CF96">number of expressed genes</strong>; and <strong style="color:#C0CF96">the proportion of reads mapped to mitochondrial genes</strong> in all four samples. Cells exhibiting a library size lower than three <b style="color:#A6A440">Median Absolute Deviations</b> (<b style="color:#A6A440">MAD</b>) were <b style="color:#537FBF">filtered</b> out. Also, cells expressing a gene count lower than three <b style="color:#A6A440">MAD</b> were filtered out. For mitochondrial read proportions, we filtered out the cells that displayed a percentage of reads mapping to mitochondrial genes greater than three <b style="color:#A6A440">MAD</b>.</p>
</blockquote>
<section id="library-size-filtering" class="level2">
<h2 class="anchored" data-anchor-id="library-size-filtering"><b style="color:#A6A440">Library Size</b> <b style="color:#537FBF">Filtering</b></h2>
<div id="191bb9d0-b042-4e74-a3de-c65cf7b315e4" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>library.sizes <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(raw.counts)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="d6353a46-bc6f-4826-8fe6-0a874f3e2c9e" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>library.median.size <span class="ot">&lt;-</span> <span class="fu">median</span>(library.sizes)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>median.absolute.deviation <span class="ot">&lt;-</span> <span class="fu">median</span>(<span class="fu">abs</span>(library.sizes <span class="sc">-</span> library.median.size))</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>library.lower.bound <span class="ot">&lt;-</span> library.median.size <span class="sc">-</span> <span class="dv">3</span><span class="sc">*</span>median.absolute.deviation</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(library.sizes <span class="sc">&lt;</span> library.lower.bound)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
FALSE
</div>
</div>
<p>It seems that none of the cells have a sufficiently low <b style="color:#A6A440">library size</b> to be <b style="color:#537FBF">filtered out</b>.</p>
</section>
<section id="gene-count-filtering" class="level2">
<h2 class="anchored" data-anchor-id="gene-count-filtering"><b style="color:#A6A440">Gene Count</b> <b style="color:#537FBF">Filtering</b></h2>
<div id="9659dd0a-0e4f-47bd-aa0c-696c4805693d" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>gene.counts <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(raw.counts <span class="sc">&gt;</span> <span class="dv">0</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="2ab8eccf-b6d3-49c2-8ed6-87b08566939b" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>gene.median.counts <span class="ot">&lt;-</span> <span class="fu">median</span>(gene.counts)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>median.absolute.deviation <span class="ot">&lt;-</span> <span class="fu">median</span>(<span class="fu">abs</span>(gene.counts <span class="sc">-</span> gene.median.counts))</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>gene.lower.bound <span class="ot">&lt;-</span> gene.median.counts <span class="sc">-</span> <span class="dv">3</span><span class="sc">*</span>median.absolute.deviation</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>gene.filter.index <span class="ot">&lt;-</span> gene.counts <span class="sc">&lt;</span> gene.lower.bound</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(gene.filter.index)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
TRUE
</div>
</div>
<div id="b8665f53-26f9-4b59-8882-1da99d64fdb0" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(gene.filter.index)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
557
</div>
</div>
<p>We’ll <b style="color:#537FBF">filter out</b> these 557 cells.</p>
<div id="abf4b630-70cc-4aa7-a522-644582150317" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># gf for gene filtering</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>filtered.counts.gf <span class="ot">&lt;-</span> raw.counts[, <span class="sc">!</span>gene.filter.index]</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>filtered.row.info.gf <span class="ot">&lt;-</span> row.info</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>filtered.col.info.gf <span class="ot">&lt;-</span> col.info[<span class="sc">!</span>gene.filter.index, ]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="mitochondrial-filtering" class="level2">
<h2 class="anchored" data-anchor-id="mitochondrial-filtering">Mitochondrial <b style="color:#537FBF">Filtering</b></h2>
<p>For this we’ll use the <code>biomart</code> package.</p>
<details>
<summary>
<b class="setups">Install</b><b style="color:#C0CF96">: biomaRt</b>
</summary>
<pre class="{bash}"><code>conda search bioconda::r-biomart
conda install bioconda::bioconductor-biomart=2.54.0 -y</code></pre>
<p>This is the first time I’ve had to use the <code>bioconda</code> channel, but <a href="https://bioconda.github.io/#:~:text=The%20bioconda%20channel%20enables%20installation,is%20set%20to%20highest%20priority.">their website</a> clearly shows they’re built on top of <code>conda-forge</code> so I know I’m not going to get burned by channel conflicts as has happened before. Unfortunately it doesn’t work, all versions of <code>biomart</code> have dependency conflicts somewhere (so maybe I have been burned by channel conflicts…)</p>
<p>So now I will install <code>biocmanager</code> and install it the in-R way.</p>
<pre class="{bash}"><code># Bash
conda search conda-forge::r-biocmanager
conda install conda-forge::r-biocmanager=1.30.19 -y</code></pre>
<pre class="{R}"><code># R
BiocManager::install("biomaRt")</code></pre>
</details>
<div id="c3d1b52d-bc4d-4995-a1d0-f272513caca3" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"biomaRt"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="330ca076-b415-4b7b-9b2b-6e9cf9c8f908" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Configure biomart to use human info (as opposed to other species)</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>mart <span class="ot">&lt;-</span> <span class="fu">useDataset</span>(<span class="st">"hsapiens_gene_ensembl"</span>, <span class="fu">useMart</span>(<span class="st">"ensembl"</span>))</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>gene.to.symbol.map <span class="ot">&lt;-</span> <span class="fu">getBM</span>(</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">filters=</span><span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">attributes=</span><span class="fu">c</span>(</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>        <span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>        <span class="st">"hgnc_symbol"</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">values=</span>filtered.row.info.gf<span class="sc">$</span>Ensembl.ID,</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">mart=</span>mart</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Annoyingly, <code>getBM</code> silently drops IDs when it can’t find a matching attribute, so we loose twenty genes:</p>
<div id="baa42d21-f1e6-4ce7-9d91-38d0fb92ebee" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(gene.to.symbol.map<span class="sc">$</span>ensembl_gene_id)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(filtered.row.info.gf<span class="sc">$</span>Ensembl.ID)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
23264
</div>
<div class="cell-output cell-output-display">
23284
</div>
</div>
<div id="d96105fd-369f-4225-ac67-26201d1345b9" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>the.lost.genes <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>    filtered.row.info.gf<span class="sc">$</span>Ensembl.ID,</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>    gene.to.symbol.map<span class="sc">$</span>ensembl_gene_id</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>We’ll <b style="color:#537FBF">filter</b> these genes out:</p>
<div id="91b72187-ecc0-488a-bcce-6a4c582975fe" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>non.missing.gene.index <span class="ot">&lt;-</span> <span class="sc">!</span>(filtered.row.info.gf<span class="sc">$</span>Ensembl.ID <span class="sc">%in%</span> the.lost.genes)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>filtered.counts.mito <span class="ot">&lt;-</span> filtered.counts.gf[non.missing.gene.index,]</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>filtered.row.info.mito <span class="ot">&lt;-</span> filtered.row.info.gf[non.missing.gene.index,]</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>filtered.col.info.mito <span class="ot">&lt;-</span> filtered.col.info.gf</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>And then continue with our attempt to <b style="color:#537FBF">filter out</b> highly-mitochondrial cells.</p>
<div id="a8ebc629-b740-4425-909e-9ea29ff94a20" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For some reason, r is behaving really strangely with `strsplit`</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co"># which is why this function is so strangely complicated</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>mito.genes.index <span class="ot">&lt;-</span> <span class="fu">sapply</span>(</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>    gene.to.symbol.map<span class="sc">$</span>hgnc_symbol,</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(x) {</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (x <span class="sc">==</span> <span class="st">""</span>) {</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>            <span class="fu">return</span>(<span class="cn">FALSE</span>)</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>        split <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(x, <span class="st">'-'</span>)</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="fu">length</span>(split) <span class="sc">&gt;=</span> <span class="dv">1</span>) {</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>            split. <span class="ot">&lt;-</span> split[[<span class="dv">1</span>]]</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> (<span class="fu">length</span>(split.) <span class="sc">&gt;</span> <span class="dv">1</span>) {</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>                split. <span class="ot">&lt;-</span> split.[[<span class="dv">1</span>]]</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>            out <span class="ot">&lt;-</span> <span class="fu">tolower</span>(split.) <span class="sc">==</span> <span class="st">'mt'</span></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>            <span class="fu">return</span>(out)</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span> {</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>            <span class="fu">return</span>(<span class="cn">FALSE</span>)</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="bd94141b-7400-4d42-ad84-a1ab5954c0f4" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>gene.to.symbol.map[mito.genes.index, ]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small">
<caption>A data.frame: 13 × 2</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">ensembl_gene_id</th>
<th data-quarto-table-cell-role="th" scope="col">hgnc_symbol</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">15116</th>
<td>ENSG00000198695</td>
<td>MT-ND6</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">15119</th>
<td>ENSG00000198712</td>
<td>MT-CO2</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">15127</th>
<td>ENSG00000198727</td>
<td>MT-CYB</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">15146</th>
<td>ENSG00000198763</td>
<td>MT-ND2</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">15153</th>
<td>ENSG00000198786</td>
<td>MT-ND5</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">15162</th>
<td>ENSG00000198804</td>
<td>MT-CO1</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">15183</th>
<td>ENSG00000198840</td>
<td>MT-ND3</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">15203</th>
<td>ENSG00000198886</td>
<td>MT-ND4</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">15205</th>
<td>ENSG00000198888</td>
<td>MT-ND1</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">15211</th>
<td>ENSG00000198899</td>
<td>MT-ATP6</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">15234</th>
<td>ENSG00000198938</td>
<td>MT-CO3</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">15790</th>
<td>ENSG00000212907</td>
<td>MT-ND4L</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">17277</th>
<td>ENSG00000228253</td>
<td>MT-ATP8</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>So now we have a list of all mitochondrial genes.</p>
<div id="d627611a-0708-4818-9646-c7fe5d4f77bb" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>mito.counts <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(filtered.counts.mito <span class="sc">*</span> mito.genes.index)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>not.mito.counts <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(filtered.counts.mito <span class="sc">*</span> <span class="sc">!</span>mito.genes.index)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="3189a880-3bc8-4b55-b747-901800b64aaf" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>mito.percent <span class="ot">&lt;-</span> mito.counts <span class="sc">/</span> not.mito.counts</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="b87be4bf-98f7-4e53-9897-de4c0eed7ad7" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>mito.median.percent <span class="ot">&lt;-</span> <span class="fu">median</span>(mito.percent)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>median.absolute.deviation <span class="ot">&lt;-</span> <span class="fu">median</span>(<span class="fu">abs</span>(mito.percent <span class="sc">-</span> mito.median.percent))</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>mito.lower.bound <span class="ot">&lt;-</span> mito.median.percent <span class="sc">-</span> <span class="dv">3</span><span class="sc">*</span>median.absolute.deviation</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>mito.index <span class="ot">&lt;-</span> mito.percent <span class="sc">&lt;</span> mito.lower.bound</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(mito.index)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
TRUE
</div>
</div>
<div id="40f93447-c6f6-48ab-a9ac-b009db9162b4" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb23"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(mito.index)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
99
</div>
</div>
<div id="52c530c3-9c5c-4a70-a16d-ee3ed33b4c04" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>filtered.counts.mito<span class="fl">.2</span> <span class="ot">&lt;-</span> filtered.counts.mito[, <span class="sc">!</span>mito.index]</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>filtered.row.info.mito<span class="fl">.2</span> <span class="ot">&lt;-</span> filtered.row.info.mito</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>filtered.col.info.mito<span class="fl">.2</span> <span class="ot">&lt;-</span> filtered.col.info.mito[<span class="sc">!</span>mito.index, ]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="multiplets" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="multiplets"><b style="color:#A6A440">Multiplets</b></h2>
<blockquote class="blockquote">
<p>After this <b style="color:#537FBF">filtering</b> two cells still showed an outlier distribution on the <b style="color:#A6A440">library size</b>. These two “cells” were assumed to be <b style="color:#A6A440">muliplets</b> and were excluded from further analyses.</p>
</blockquote>
<p>To see the <b style="color:#A6A440">multiplets</b>, we can make a <b style="color:#A6A440">box and whisker plot</b>.</p>
<div id="074504e1-d709-49a0-b54a-742ddb1417e6" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb25"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggplot2"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="42dcd43f-1021-41d5-b10f-e4fd2718c17f" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We need to recreate the sizes as things have been filtered since then</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>library.sizes <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(filtered.counts.mito<span class="fl">.2</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>sizes.to.plot <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(library.sizes)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="9b5caaf3-6977-4fd1-ae6e-e750914863a2" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(sizes.to.plot, <span class="fu">aes</span>(<span class="at">y=</span>library.sizes)) <span class="sc">+</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_boxplot</span>(<span class="at">outlier.colour=</span><span class="st">"black"</span>, <span class="at">outlier.shape=</span><span class="dv">16</span>,</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">outlier.size=</span><span class="dv">3</span>, <span class="at">notch=</span><span class="cn">FALSE</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="011_Cancer2_files/figure-html/cell-25-output-1.png" width="420" height="420" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>This <b style="color:#EB1960">box-and-whiskers plot</b> shows more than two outliers, although there are two cells that are clearly more outlier-y than the rest. We <em style="color:#C0CF96">should</em> just filter out the two obvious ones, to adhere to the paper, but we’ll filter out all of them for the sake of reducing our dataset further.</p>
<blockquote class="blockquote">
<p>An outlier is an observation that is numerically distant from the rest of the data. When reviewing a boxplot, an outlier is defined as a data point that is located outside the fences (“whiskers”) of the boxplot (e.g: outside 1.5 times the interquartile range above the upper quartile and bellow the lower quartile). – <b style="color:#EB1960">R Statistics Blog</b>, <a href="https://www.r-statistics.com/tag/boxplot-outlier/#:~:text=An%20outlier%20is%20an%20observation,and%20bellow%20the%20lower%20quartile">available here</a>.</p>
</blockquote>
<div id="857abb73-73b5-43c2-b880-ceec08cd5e2f" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>offset <span class="ot">&lt;-</span> <span class="fl">1.5</span> <span class="sc">*</span> <span class="fu">IQR</span>(library.sizes)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>quants <span class="ot">&lt;-</span> <span class="fu">quantile</span>(library.sizes, <span class="at">probs=</span><span class="fu">c</span>(<span class="fl">0.25</span>, <span class="fl">0.75</span>))</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>quants[<span class="st">"25%"</span>] <span class="ot">&lt;-</span> quants[<span class="st">"25%"</span>] <span class="sc">-</span> offset</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>quants[<span class="st">"75%"</span>] <span class="ot">&lt;-</span> quants[<span class="st">"75%"</span>] <span class="sc">+</span> offset</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>outliers.index <span class="ot">&lt;-</span> library.sizes <span class="sc">&lt;</span> quants[<span class="st">"25%"</span>] <span class="sc">|</span> library.sizes <span class="sc">&gt;</span> quants[<span class="st">"75%"</span>]</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(outliers.index) <span class="sc">/</span> <span class="fu">length</span>(outliers.index)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
0.0242925509779442
</div>
</div>
<p>About 2.5% of the data is an outlier! Which to me feels as if they’re not-so-outlier-y after all! (It goes up to 8% if you use the unfiltered data).</p>
<details>
<summary>
<b class="sidequests">Sidequest</b><b style="color:#C0CF96">: What is the largest percent of a dataset that we can make be outliers?</b>
</summary>
<p>Well, assuming you have a dataset with 4 elements, the IQR is determined by the middle two elements; you can move the outer two as far away from it as possible. Thus you can force them to be outliers, giving you an outlier percentage of 50%.</p>
<p>This should be optimal, because at least 50% of the data has to lie within the IQR by the definition of the IQR.</p>
</details>

<div class="no-row-height column-margin column-container"><div class="">
<p><b style="color:#C0CF96">50%</b></p>
</div></div><p>Normally I’d be reluctant to remove these, but we have so much data that it’d be nice to shave some of it off<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>.</p>
<div id="b842bd2a-c1c2-4552-8fe1-6961afc39ae3" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb29"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>filtered.counts.iqr <span class="ot">&lt;-</span> filtered.counts.mito<span class="fl">.2</span>[, <span class="sc">!</span>outliers.index]</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>filtered.row.info.iqr <span class="ot">&lt;-</span> filtered.row.info.mito<span class="fl">.2</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>filtered.col.info.iqr <span class="ot">&lt;-</span> filtered.col.info.mito<span class="fl">.2</span>[<span class="sc">!</span>outliers.index, ]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="f1ea3f21-0df7-47b7-97a7-006dd11869db" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(filtered.counts.iqr)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>23264</li><li>18757</li></ol>
</div>
</div>
</section>
<section id="normalization-and-miscellaneous-filtering" class="level2">
<h2 class="anchored" data-anchor-id="normalization-and-miscellaneous-filtering"><b style="color:#537FBF">Normalization</b> and miscellaneous <b style="color:#537FBF">filtering</b></h2>
<blockquote class="blockquote">
<p><b style="color:#C0CF96">Raw counts</b> of the remaining cells were then <b style="color:#537FBF">normalised</b> using the <b style="color:#537FBF">deconvolution-based method</b> and then <b style="color:#537FBF">log-transformed</b>. We also <b style="color:#537FBF">filtered out</b> the genes with average counts below 0.01 assuming these low-abundance genes to be unreliable for statistical inference.</p>
</blockquote>
<p>The question is - what is this <b style="color:#537FBF">deconvolution-based method</b>? I found <a href="http://bioconductor.org/books/3.13/OSCA.basic/normalization.html#normalization-by-deconvolution">this helpful resource</a> which describes it well. Essentially, we want to be able to normalize each cell by their library size, but due to the sparsity of scRNA-seq we need to pool cells together to achieve sufficient mass. We then “<b style="color:#537FBF">deconvolve</b>” the pooled values to find them on a per-cell basis.</p>
<details>
<summary>
<b class="setups">Install</b><b style="color:#C0CF96">: scran</b>
</summary>
<pre class="{R}"><code>BiocManager::install("scran")</code></pre>
</details>
<div id="f06825b1-5771-421c-b5f7-f30bdaebda07" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scran)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="299c667e-f761-4b26-bc18-4bb60d038f5b" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb33"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>quick.clusters <span class="ot">&lt;-</span> <span class="fu">quickCluster</span>(filtered.counts.iqr)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message in regularize.values(x, y, ties, missing(ties), na.rm = na.rm):
“collapsing to unique 'x' values”</code></pre>
</div>
</div>
<div id="2894e6f9-1261-4e7e-bdf4-7128a7ef03b5" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb35"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(quick.clusters)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>quick.clusters
   1    2    3    4    5    6    7    8    9   10   11   12 
1807 1803 3116 1371 2731 1939 1382 1620  150 2608  126  104 </code></pre>
</div>
</div>
<div id="0b9163f8-7fef-438a-bd0d-084cff8ee8e7" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb37"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>deconvoluted.counts <span class="ot">&lt;-</span> <span class="fu">calculateSumFactors</span>(</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>    filtered.counts.iqr,</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster=</span>quick.clusters</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(deconvoluted.counts)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.1732  0.6554  0.9378  1.0000  1.2619  3.5655 </code></pre>
</div>
</div>
<p>So, in short, the output is a length of factors to divide the cells by.</p>
<div id="82d771be-ac28-48de-8333-4c5868c07709" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb39"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(filtered.counts.iqr)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(deconvoluted.counts)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>23264</li><li>18757</li></ol>
</div>
<div class="cell-output cell-output-display">
18757
</div>
</div>
<div id="9ed70943-36f4-4d90-ad96-49cd8d38b146" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>filtered.counts.deconv <span class="ot">&lt;-</span> filtered.counts.iqr <span class="sc">/</span> deconvoluted.counts</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>filtered.row.info.deconv <span class="ot">&lt;-</span> filtered.row.info.iqr</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>filtered.col.info.deconv <span class="ot">&lt;-</span> filtered.col.info.iqr</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Now we’ll <b style="color:#537FBF">filter out</b> the genes with small mean counts.</p>
<div id="1fe5049d-76ac-48fc-b339-564af89ecc8e" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb41"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>small.mean.genes.index <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(filtered.counts.deconv) <span class="sc">&lt;</span> <span class="fl">0.01</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="921ebf6e-7e32-4ffa-bd3b-e2bbdb750cc8" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>filtered.counts.gene.mean <span class="ot">&lt;-</span> filtered.counts.deconv[<span class="sc">!</span>small.mean.genes.index, ]</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>filtered.row.info.gene.mean <span class="ot">&lt;-</span> filtered.row.info.deconv[<span class="sc">!</span>small.mean.genes.index]</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>filtered.col.info.gene.mean <span class="ot">&lt;-</span> filtered.col.info.deconv</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="0fb9c496-9b0e-4d93-96fa-57d794c27bf3" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb43"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(filtered.counts.gene.mean)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>14872</li><li>18757</li></ol>
</div>
</div>
<p>Finally, we’ll <b style="color:#537FBF">log-transform</b> the data. We’ll leave the 0 values as 0 for the sake of our RAM, which means we need to peak into the internal workings of the sparse matrix class.</p>
<div id="fbd1bce7-fe59-4d29-a4a7-1dd81fef9eea" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb44"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(filtered.counts.gene.mean)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
'dgCMatrix'
</div>
</div>
<div id="c0cdaf77-5f4d-4489-9a74-d5cc18910d44" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb45"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>filtered.log.counts <span class="ot">&lt;-</span> filtered.counts.gene.mean</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>filtered.log.row.info <span class="ot">&lt;-</span> filtered.row.info.gene.mean</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>filtered.log.col.info <span class="ot">&lt;-</span> filtered.col.info.gene.mean</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>filtered.log.counts<span class="sc">@</span>x <span class="ot">&lt;-</span> <span class="fu">log</span>(filtered.log.counts<span class="sc">@</span>x)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="a7e8c931-a563-4a31-88be-339b9efad860" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb46"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(filtered.log.counts)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>14872</li><li>18757</li></ol>
</div>
</div>
</section>
</section>
<section id="highly-variable-genes" class="level1">
<h1><b style="color:#A6A440">Highly Variable Genes</b></h1>
<blockquote class="blockquote">
<p>For visualisation and clustering we first selected the <b style="color:#A6A440">Highly Variable Genes</b> (<b style="color:#A6A440">HVGs</b>). For this we first decomposed the variance of expression in each gene to <b style="color:#A6A440">technical</b> and <b style="color:#A6A440">biological components</b> and identified the genes as <b style="color:#A6A440">HVGs</b> where the <b style="color:#A6A440">biological components</b> were significantly greater than zero.</p>
<p>– <span class="citation" data-cites="ovarian-cancer-living-biobank">Nelson et al. (<a href="#ref-ovarian-cancer-living-biobank" role="doc-biblioref">2020</a>)</span></p>
</blockquote>
<p>I’ll continue following the <a href="http://bioconductor.org/books/3.13/OSCA.basic/feature-selection.html">bioconductor guide</a> on this as I’ve not done it before and would like to do a blog post in the future about it.</p>
<div id="5825a619-18ec-42e5-ac89-02974ac63fcf" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb47"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>gene.var.data <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(filtered.log.counts)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="95e7bc0e-6876-4153-80a7-60087ee91562" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb48"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualizing the fit:</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>gene.var.metadata <span class="ot">&lt;-</span> <span class="fu">metadata</span>(gene.var.data)</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(<span class="fu">data.frame</span>(gene.var.metadata<span class="sc">$</span>mean)),</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">x=</span>gene.var.metadata<span class="sc">$</span>mean</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>    <span class="co">#geom_line(aes(y=gene.var.metadata$var)) +</span></span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stat_function</span>(<span class="at">fun =</span> gene.var.metadata<span class="sc">$</span>trend, <span class="at">size=</span><span class="dv">3</span>, <span class="at">colour=</span><span class="st">'blue'</span>) <span class="sc">+</span> </span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y=</span>gene.var.metadata<span class="sc">$</span>var)) <span class="sc">+</span></span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">"Mean of log-expression"</span>) <span class="sc">+</span></span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Variance of log-expression"</span>) <span class="sc">+</span></span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="fu">max</span>(gene.var.metadata<span class="sc">$</span>mean)) <span class="sc">+</span></span>
<span id="cb48-15"><a href="#cb48-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"The equivalent plot as the bioconductor book"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Warning message:

“Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.

<span class="ansi-cyan-fg">ℹ</span> Please use `linewidth` instead.”

Warning message:

“Removed 2594 rows containing missing values (`geom_point()`).”
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="011_Cancer2_files/figure-html/cell-42-output-2.png" width="420" height="420" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We now need to indentify the genes where the <b style="color:#A6A440">biological component</b> is <b style="color:#C0CF96">significantly</b><a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a> greater than zero.</p>
<div id="834f9c48-38c2-47cd-8440-87d7e9737408" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb49"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>signif.genes <span class="ot">&lt;-</span> gene.var.data<span class="sc">$</span>p.value <span class="sc">&lt;</span> <span class="fl">0.05</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>signif.genes[<span class="fu">is.na</span>(signif.genes)] <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(signif.genes)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
6454
</div>
</div>
<div id="08cb5b46-259e-4d2c-bd31-70a58e1690e7" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb50"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>hgv.counts <span class="ot">&lt;-</span> filtered.counts.gene.mean[signif.genes,]</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>hgv.log.counts <span class="ot">&lt;-</span> filtered.log.counts[signif.genes,]</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>hgv.log.row.info <span class="ot">&lt;-</span> filtered.log.row.info[signif.genes]</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>hgv.log.col.info <span class="ot">&lt;-</span> filtered.log.col.info</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="8acda698-cb28-4039-ad13-f1e157d05401" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb51"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(hgv.log.counts)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(hgv.counts)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>6454</li><li>18757</li></ol>
</div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>6454</li><li>18757</li></ol>
</div>
</div>
</section>
<section id="figure-5a" class="level1 page-columns page-full">
<h1><b style="color:#EB1960">Figure 5a</b></h1>
<blockquote class="blockquote">
<p><b style="color:#EB1960">5a)</b> <b style="color:#537FBF">t-Stochastic neighbour embedding (t-SNE)</b> plot showing <b style="color:#537FBF">clustering</b> of single cells from four <b style="color:#A6A440">OCM pairs</b>, with <b style="color:#A6A440">tumour</b> and <b style="color:#A6A440">stromal cells</b> mixed 3:1.</p>
<p>– <b style="color:#C0CF96">Caption of</b> <b style="color:#EB1960">Figure 5a</b></p>
</blockquote>
<details>
<summary>
<b class="false-starts">False Start</b><b style="color:#C0CF96">: Which cells are <b style="color:#A6A440">tumour</b> and which cells are <b style="color:#A6A440">stromal</b>?</b>
</summary>
<p>At first I thought it would be easy to tell which cells are which via hand-labeling, however we can see from below that there is no column that corresponds to cell type other <code>Sample.Characteristic.cell.type.</code>, and as we’ll see it’s not a very informative column.</p>
<div id="1f1a6721-440f-451a-a849-cd567771b365" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb52"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># To pretty-print `colnames`, we pasted them together with \n</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Note that we need to use `collapse`, not `sep` - I'm pretty</span></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="co"># sure this is because `colnames` is a vector rather than</span></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a><span class="co"># multiple arguments, so we need to "collapse" it.</span></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Finally, to render the \ns rather than printing "\n", we need</span></span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a><span class="co"># to use `cat`.  All-in-all this was far more complicated</span></span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a><span class="co"># than I thought it would be to get pretty-print working.</span></span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste</span>(</span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">colnames</span>(exp.design.table),</span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">collapse=</span><span class="st">'</span><span class="sc">\n</span><span class="st">'</span></span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb52-14"><a href="#cb52-14" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Assay
Sample.Characteristic.organism.
Sample.Characteristic.Ontology.Term.organism.
Sample.Characteristic.developmental.stage.
Sample.Characteristic.Ontology.Term.developmental.stage.
Sample.Characteristic.age.
Sample.Characteristic.Ontology.Term.age.
Sample.Characteristic.sex.
Sample.Characteristic.Ontology.Term.sex.
Sample.Characteristic.individual.
Sample.Characteristic.Ontology.Term.individual.
Sample.Characteristic.disease.
Sample.Characteristic.Ontology.Term.disease.
Sample.Characteristic.tumor.grading.
Sample.Characteristic.Ontology.Term.tumor.grading.
Sample.Characteristic.metastatic.site.
Sample.Characteristic.Ontology.Term.metastatic.site.
Sample.Characteristic.organism.part.
Sample.Characteristic.Ontology.Term.organism.part.
Sample.Characteristic.cell.type.
Sample.Characteristic.Ontology.Term.cell.type.
Sample.Characteristic.growth.condition.
Sample.Characteristic.Ontology.Term.growth.condition.
Factor.Value.individual.
Factor.Value.Ontology.Term.individual.</code></pre>
</div>
</div>
<div id="fccfb446-90b0-4fe9-a7e0-37b87c247cbc" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb54"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(exp.design.table<span class="sc">$</span>Sample.Characteristic.cell.type.)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
'mix of stromal fibroblasts and epithelial tumour cells'
</div>
</div>
<p>Thus every cell is either a <b style="color:#A6A440">stromal fibroblast</b> or an <b style="color:#A6A440">epithelial tumour cell</b>, but we’re not given any indication which is which.</p>
<blockquote class="blockquote">
<p><b style="color:#A6A440">Tumour</b> and <b style="color:#A6A440">stromal</b> cells from the four pairs were mixed 3:1 and analysed in parallel. <b style="color:#537FBF">tSNE plots</b> showed that the majority of cells from each sample formed distinct clusters, whereas smaller fractions formed an overlapping cluster (Fig. 5a). Based on the 3:1 mix, we reasoned that the large distinct clusters represented the <b style="color:#A6A440">tumour cells</b> while the overlapping cluster corresponded to the <b style="color:#A6A440">stromal cells</b>.</p>
</blockquote>
<p>So it seems that they didn’t keep track of which cells were which (which makes sense, because it’d probably be pretty hard… Maybe they could have tagged the cells somehow, but maybe that’d be expensive or inaccurate - it’s hard for me to say as I don’t have the background). Rather, they just inferred them from <b style="color:#537FBF">tSNE plots</b>.</p>
</details>

<div class="no-row-height column-margin column-container"><div class="">
<p>Spoiler: We don’t know which!</p>
</div></div><p>So we need to run <b style="color:#537FBF">tSNE</b> - ideally, we’d like to use the exact same packages they used for analysis, but they used <code>scanpy</code>, a Python package, and this is an R notebook<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>. So, instead, we’ll continue to use the <code>scran</code> package.</p>
<p>However, before we run <b style="color:#537FBF">tSNE</b> we have do to (yet another) dimensionality reduction:</p>
<blockquote class="blockquote">
<p>These <b style="color:#A6A440">HVG genes</b> were then used to reduce the dimensions of the dataset using <b style="color:#537FBF">PCA</b>. <b style="color:#537FBF">T-SNE</b> plots were then generated by taking 1–14 components of the <b style="color:#537FBF">PCA</b>.</p>
</blockquote>
<div id="415c15d4-6e3e-41c4-9c99-f7057b0bd3bd" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb55"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>sce.data <span class="ot">&lt;-</span> <span class="fu">SingleCellExperiment</span>(</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">assays=</span><span class="fu">list</span>(</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">counts=</span>hgv.counts,</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">logcounts=</span>hgv.log.counts</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="66aaacf4-de60-4d84-8577-a06d10084600" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb56"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>sce.data.pca <span class="ot">&lt;-</span> <span class="fu">fixedPCA</span>(</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>    sce.data,</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">rank=</span><span class="dv">14</span>,</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">subset.row=</span><span class="cn">NULL</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>dimnames(.) &lt;- NULL translated to
dimnames(.) &lt;- list(NULL,NULL)

dimnames(.) &lt;- NULL translated to
dimnames(.) &lt;- list(NULL,NULL)
</code></pre>
</div>
</div>
<div id="5f18cebb-9cec-4004-9fd2-7b65538bf019" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb58"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">reducedDim</span>(sce.data.pca, <span class="st">"PCA"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>18757</li><li>14</li></ol>
</div>
</div>
<p>To run tSNE, we need the <code>scater</code> library - note that there is only one “t” in its name!</p>
<details>
<summary>
<b class="setups">Install</b><b style="color:#C0CF96">: scater</b>
</summary>
<pre class="{R}"><code>BiocManager::install("scater", type="mac.binary")</code></pre>
<p>You should probably remove the <code>mac.binary</code> bit if not on a Mac, but if you’re on a Mac you might need it - without it I got the following uninformative error:</p>
<pre><code>'getOption("repos")' replaces Bioconductor standard repositories, see
'?repositories' for details

replacement repositories:
    CRAN: https://cran.r-project.org


Bioconductor version 3.16 (BiocManager 1.30.19), R 4.2.1 (2022-06-23)

Installing package(s) 'scater'

also installing the dependencies ‘textshaping’, ‘Cairo’, ‘ragg’, ‘ggrastr’


Warning message in install.packages(...):
“installation of package ‘textshaping’ had non-zero exit status”
Warning message in install.packages(...):
“installation of package ‘Cairo’ had non-zero exit status”
Warning message in install.packages(...):
“installation of package ‘ragg’ had non-zero exit status”
Warning message in install.packages(...):
“installation of package ‘ggrastr’ had non-zero exit status”
Warning message in install.packages(...):
“installation of package ‘scater’ had non-zero exit status”
Updating HTML index of packages in '.Library'

Making 'packages.html' ...
 done

Old packages: 'boot', 'broom', 'cli', 'cpp11', 'crayon', 'desc', 'digest',
  'evaluate', 'forcats', 'generics', 'gtable', 'hms', 'IRkernel', 'isoband',
  'jsonlite', 'mitml', 'nlme', 'pillar', 'pkgload', 'repr', 'rstudioapi',
  'scales', 'statmod', 'stringi', 'stringr', 'survival', 'testthat',
  'viridisLite'</code></pre>
</details>
<div id="225912c8-b861-4f66-85b2-c4f33659c9e6" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb61"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“package ‘scater’ was built under R version 4.2.2”</code></pre>
</div>
</div>
<div id="9e802779-638c-43eb-b20d-ce5f0b130e5a" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb63"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>sce.data.tsne <span class="ot">&lt;-</span> <span class="fu">runTSNE</span>(sce.data.pca, <span class="at">dimred=</span><span class="st">"PCA"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="dec0ec01-dd7e-4440-991f-7c53b2536bee" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb64"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(sce.data.tsne, <span class="at">dimred=</span><span class="st">"TSNE"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="011_Cancer2_files/figure-html/cell-53-output-1.png" width="420" height="420" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The only thing remaining is to color by patient.</p>
<div id="148b28ba-dde7-434f-9b86-aa6e4a07c6ef" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb65"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>exp.design.index <span class="ot">&lt;-</span> exp.design.table<span class="sc">$</span>Assay <span class="sc">%in%</span> hgv.log.col.info<span class="sc">$</span>Full.ID</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="01de80b4-024f-46e2-a34f-f10836a870ea" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb66"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>sce.data.tsne<span class="sc">$</span>patient <span class="ot">&lt;-</span> exp.design.table[exp.design.index, ]<span class="sc">$</span>Sample.Characteristic.individual</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="e0412b3c-42c1-4ec8-933d-c53a128ba23b" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb67"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(sce.data.tsne, <span class="at">dimred=</span><span class="st">"TSNE"</span>, <span class="at">colour_by=</span><span class="st">"patient"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="011_Cancer2_files/figure-html/cell-56-output-1.png" width="420" height="420" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Voilà, it’s done! That took me ages<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a> - mainly because of my lack of experience with R and its ecosystem<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>. But, it’s done. Our graph does not look as nice as theirs, but I’m sure it could do if we were to mess with the <b style="color:#537FBF">tSNE</b> parameters. There are clear patient-related clusters, and then there are sum that are much more mixed which are presumably the <b style="color:#A6A440">stromal</b> clusters.</p>
<p>Since it took some time to run the normalization steps, and because we’ll be continuing where we left off in the next post, let’s save our data:</p>
<div id="3a44e27b-d30a-407f-8ced-88685f3417cd" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb68"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>Matrix<span class="sc">::</span><span class="fu">writeMM</span>(hgv.counts, <span class="st">"./localdata/ovarian-cancer-hgv-counts.mtx"</span>)</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(hgv.log.col.info<span class="sc">$</span>Full.ID, <span class="st">"./localdata/ovarian-cancer-hgv-columns.csv"</span>)</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(hgv.log.row.info, <span class="st">"./localdata/ovarian-cancer-hgv-rows.csv"</span>)</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(exp.design.table[exp.design.index, ], <span class="st">"./localdata/ovarian-cancer-hgv-metadata.csv"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>NULL</code></pre>
</div>
</div>



</section>


<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-ovarian-cancer-living-biobank" class="csl-entry" role="listitem">
Nelson, Louisa, Anthony Tighe, Anya Golder, Samantha Littler, Bjorn Bakker, Daniela Moralli, Syed Murtuza Baker, et al. 2020. <span>“A Living Biobank of Ovarian Cancer Ex Vivo Models Reveals Profound Mitotic Heterogeneity.”</span> <em>Nature Communications</em> 11 (1): 822. <a href="https://doi.org/10.1038/s41467-020-14551-2">https://doi.org/10.1038/s41467-020-14551-2</a>.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>I know that’s not the normal attitude, I just don’t want to have to deal with any code that takes longer than a second to run 😅. Although I chose R over numpy so clearly I don’t care that much about speed… Although I guess I’m sure that R can be fast too if you understand how you’re supposed to use it.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>In the statistical sense.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>If only <code>sos-notebook</code> didn’t ruin syntax highlighting… This might require a <b class="sidequests">Sidequest</b> sometime soon.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p>Two days, to be precise. It’s 20:54 Jan 21 and I started in the morning of Jan 20 (with many distractions, to be fair)<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn5"><p>Shoutout to my hour-long adventure trying to install <code>scater</code>.<a href="#fnref5" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script>
    // Remove background
    elements = document.querySelectorAll('[id=clean-collapse]').forEach(
        function(element) {
            element.classList.remove("callout-style-default");
        }
    )
    
    // Add arrow to front to signify dropdownyness
    arrow = "▶";
    elements = document.querySelectorAll('.callout-caption-container').forEach(
        function(element) {
            element.innerHTML = arrow + element.innerHTML;
        }
    )
</script>
<script src="https://giscus.app/client.js" data-repo="baileyandrew/blog" data-repo-id="R_kgDOInJwKg" data-category="Announcements" data-category-id="DIC_kwDOInJwKs4CTGOQ" data-mapping="title" data-strict="0" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="dark_protanopia" data-lang="en" data-loading="lazy" crossorigin="anonymous" async="">
</script>
<!--
Assuming `_quarto.yml` has `navbar: pinned: true`,
this will make it so that the navbar scrolls with the rest of the website.
-->
<script>
    document.getElementById('quarto-header').classList.remove('fixed-top');
</script>
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/BaileyAndrew\.github\.io\/Blog");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/BaileyAndrew/Blog/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>