<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.27">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Bailey Andrew">
<meta name="dcterms.date" content="2023-01-22">
<meta name="description" content="Another scRNA paper">

<title>A Living Biobank of Ovarian Cancer ‚Äì Figure 5b-c ‚Äì Lorem Ipsomething</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.min.js" integrity="sha384-ZvpUoO/+PpLXR1lu4jmpXWu80pZlYUAfxl5NsBMWOEPSjUn/6Z/hRTt8+pR6L4N2" crossorigin="anonymous"></script><script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-4d9afe2b8d18ee9fa5d0d57b5ed4214d.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-10892d65c249a5d8f93b275ef5a48de5.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdn.jsdelivr.net/npm/requirejs@2.3.6/require.min.js" integrity="sha384-c9c+LnTbwQ3aujuU7ULEPVvgLs+Fn6fJUvIGTsuu1ZcCf11fiEubah0ttpca4ntM sha384-6V1/AdqZRWk1KAlWbKBlGhN7VG4iE/yAZcO6NZPMF8od0vukrvr0tg4qY6NSrItx" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../styles.css">
<link rel="stylesheet" href="../../html_scripts/collapse.css">
<link rel="stylesheet" href="../../html_scripts/pretty_shortcuts.css">
<meta property="og:title" content="A Living Biobank of Ovarian Cancer ‚Äì Figure 5b-c ‚Äì Lorem Ipsomething">
<meta property="og:description" content="Another scRNA paper">
<meta property="og:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/tsne-reproduce-ovarian-fig5c.png">
<meta property="og:site_name" content="Lorem Ipsomething">
<meta property="og:image:height" content="842">
<meta property="og:image:width" content="840">
<meta name="twitter:title" content="A Living Biobank of Ovarian Cancer ‚Äì Figure 5b-c ‚Äì Lorem Ipsomething">
<meta name="twitter:description" content="Another scRNA paper">
<meta name="twitter:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/tsne-reproduce-ovarian-fig5c.png">
<meta name="twitter:image-height" content="842">
<meta name="twitter:image-width" content="840">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed slimcontent quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Lorem Ipsomething</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">A Living Biobank of Ovarian Cancer ‚Äì Figure 5b-c</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Home</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../blog.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Blog</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../staticposts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Static Posts</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Puzzles</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../500_Puzzles/bnv.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Boulders in Valleys</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slideshowindex.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slideshows</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../otherblogs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Other People‚Äôs Blogs</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#loading-the-data" id="toc-loading-the-data" class="nav-link active" data-scroll-target="#loading-the-data">Loading the Data</a></li>
  <li><a href="#identifying-stromals" id="toc-identifying-stromals" class="nav-link" data-scroll-target="#identifying-stromals">Identifying <b style="color:#A6A440">stromals</b></a></li>
  <li><a href="#fig-5b" id="toc-fig-5b" class="nav-link" data-scroll-target="#fig-5b"><b style="color:#EB1960">Fig 5b</b></a></li>
  <li><a href="#fig-5c" id="toc-fig-5c" class="nav-link" data-scroll-target="#fig-5c"><b style="color:#EB1960">Fig 5c</b></a></li>
  <li><a href="#going-offscript" id="toc-going-offscript" class="nav-link" data-scroll-target="#going-offscript">Going Offscript</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/BaileyAndrew/Blog/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">A Living Biobank of Ovarian Cancer ‚Äì Figure 5b-c</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Work</div>
    <div class="quarto-category">Omics</div>
    <div class="quarto-category">Useful</div>
    <div class="quarto-category">Paper Replication</div>
  </div>
  </div>

<div>
  <div class="description">
    Another scRNA paper
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Bailey Andrew </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 22, 2023</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p>We‚Äôll continue where we left off in the last post, trying to recreate <b style="color:#EB1960">Figure 5</b> from a paper by <span class="citation" data-cites="ovarian-cancer-living-biobank">Nelson et al. (<a href="#ref-ovarian-cancer-living-biobank" role="doc-biblioref">2020</a>)</span>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/ovarian-fig5.png" class="img-fluid figure-img"></p>
<figcaption><b style="color:#EB1960">Figure 5</b>. Not my own work</figcaption>
</figure>
</div>
<p>This time, we‚Äôll focus on <b style="color:#EB1960">Figure 5b</b>, which entails deciding whether a cell is <b style="color:#A6A440">stromal</b> or <b style="color:#A6A440">tumour</b>.</p>
<blockquote class="blockquote">
<p><b style="color:#EB1960">5b)</b> Dot plots quantitating the percentage of cells in the <b style="color:#EB1960">stromal</b> and <b style="color:#EB1960">tumour</b> clusters. Line represents the mean (N = 4 biological independent samples i.e.&nbsp;n = 1 for each of the four <b style="color:#A6A440">OCMsM</b>).</p>
<p>‚Äì <span class="citation" data-cites="ovarian-cancer-living-biobank">Nelson et al. (<a href="#ref-ovarian-cancer-living-biobank" role="doc-biblioref">2020</a>)</span></p>
</blockquote>
<section id="loading-the-data" class="level1 page-columns page-full">
<h1>Loading the Data</h1>
<p>To get started, we‚Äôll load in our data and then create a <code>SingleCellExperiment</code> class from the eponymous <code>SingleCellExperiment</code> package. This is a dependency of <code>scran</code> and <code>scater</code> which were installed in the last post; so we‚Äôll go ahead and the latter two (and we‚Äôll explicitly load <code>SingleCellExperiment</code> this time for clarity).</p>
<div id="38644897-bc50-41d1-89db-03f8d6bd2ca3" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For reproducibility</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">0</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="1b0e534d-2914-40f0-a14e-bcd8848d6b80" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleCellExperiment)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="d64a13df-f32b-4b91-9f62-717365fc2309" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>counts.mat <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">readMM</span>(<span class="st">'./localdata/ovarian-cancer-hgv-counts.mtx'</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="33d4d329-7b05-4fd0-9bec-844123cc4741" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>counts.rows <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">'./localdata/ovarian-cancer-hgv-rows.csv'</span>, <span class="at">sep=</span><span class="st">','</span>)[,<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>counts.cols <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">'./localdata/ovarian-cancer-hgv-columns.csv'</span>, <span class="at">sep=</span><span class="st">','</span>)[,<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>counts.meta <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">'./localdata/ovarian-cancer-hgv-metadata.csv'</span>, <span class="at">sep=</span><span class="st">','</span>)[,<span class="sc">-</span><span class="dv">1</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="ba0771cf-fbf2-4db6-9838-08613dd425a4" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(counts.mat) <span class="ot">&lt;-</span> counts.rows</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(counts.mat) <span class="ot">&lt;-</span> counts.cols</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>From documentation, it seems like the convention is to abbreviate <code>SingleCellExperiment</code> as <code>sce</code>, so we‚Äôll name our data<code>ovarian.sce</code>.</p>
<div id="70bd0b21-c763-4b35-a028-2b8e5b223c78" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>ovarian.sce <span class="ot">&lt;-</span> <span class="fu">SingleCellExperiment</span>(</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">assays=</span><span class="fu">list</span>(<span class="at">counts=</span>counts.mat),</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="33b12b1f-c05f-4886-a413-d4627ef7be04" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add patient metadata</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>ovarian.sce<span class="sc">$</span>patient <span class="ot">&lt;-</span> counts.meta<span class="sc">$</span>Sample.Characteristic.individual.</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="4da4ce6c-3ceb-4714-9c6d-4490baf8770a" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a peak at internal structure</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>ovarian.sce</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>class: SingleCellExperiment 
dim: 6454 18757 
metadata(0):
assays(1): counts
rownames(6454): ENSG00000000457 ENSG00000000460 ... ENSG00000288504
  ENSG00000288649
rowData names(0):
colnames(18757): SAMEA6492740-AAACCCACAGTTAGGG
  SAMEA6492740-AAACCCACATGTGTCA ... SAMEA6492743-TTTGTTGGTCCTGGTG
  SAMEA6492743-TTTGTTGTCAGATTGC
colData names(1): patient
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
<p>While R somewhat markets itself as a functional<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> language<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>, the way data is used tends to look a tad imperative. We‚Äôll be writing a lot of lines that look like:</p>
<pre class="{R}"><code>ovarian.sce &lt;- f(ovarian.sce)</code></pre>
<p>Which, yeah, is genuinely functional:</p>
<pre class="{R}"><code># This does not change `ovarian.sce`
ovarian.sce.f &lt;- f(ovarian.sce)</code></pre>
<p>But it feels like it‚Äôs functional being used with an imperative mindset üòÖ. It feels a bit weird to me! But it‚Äôs not a big deal.</p>
<p>Let‚Äôs take the log of our data. We‚Äôll use <code>logNormCounts</code> from the <code>scuttle</code> package<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>. This will have the side effect of normalizing our data, but since our data was already normalized I don‚Äôt think this will cause a problem? ü§∑‚Äç‚ôÇÔ∏è I‚Äôm just being quick and dirty here to get some experience using the <code>SingleCellExperiment</code> class, we saw an example of manually normalizing in the last blog post so we won‚Äôt worry too much here.</p>
<div id="9f2d0340-ed66-4d4c-ba24-812ba73ef725" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>ovarian.sce <span class="ot">&lt;-</span> scuttle<span class="sc">::</span><span class="fu">logNormCounts</span>(ovarian.sce)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>We can see that we now have a new <b style="color:#A6A440">assay</b>, <code>logcounts</code>, as well as more <code>colData</code> ‚Äú<code>sizeFactor</code>‚Äù. The latter is the value we used to scale counts for normalization.</p>
<div id="2aaa3b14-480b-4105-97a0-5006a5554529" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>ovarian.sce</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>class: SingleCellExperiment 
dim: 6454 18757 
metadata(0):
assays(2): counts logcounts
rownames(6454): ENSG00000000457 ENSG00000000460 ... ENSG00000288504
  ENSG00000288649
rowData names(0):
colnames(18757): SAMEA6492740-AAACCCACAGTTAGGG
  SAMEA6492740-AAACCCACATGTGTCA ... SAMEA6492743-TTTGTTGGTCCTGGTG
  SAMEA6492743-TTTGTTGTCAGATTGC
colData names(2): patient sizeFactor
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
<div id="0bc19fe0-7794-4f6c-b56d-648873198518" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(ovarian.sce)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>DataFrame with 18757 rows and 2 columns
                                  patient sizeFactor
                              &lt;character&gt;  &lt;numeric&gt;
SAMEA6492740-AAACCCACAGTTAGGG         38b   1.127861
SAMEA6492740-AAACCCACATGTGTCA         38b   1.110393
SAMEA6492740-AAACCCAGTCGCATGC         38b   1.701412
SAMEA6492740-AAACCCAGTCTTTCAT         38b   1.058486
SAMEA6492740-AAACCCATCCGTGTCT         38b   0.586412
...                                   ...        ...
SAMEA6492743-TTTGGTTTCATCGTAG          79   1.364770
SAMEA6492743-TTTGTTGCAACCGACC          79   0.249135
SAMEA6492743-TTTGTTGGTCAGTCCG          79   1.374554
SAMEA6492743-TTTGTTGGTCCTGGTG          79   1.357728
SAMEA6492743-TTTGTTGTCAGATTGC          79   1.021525</code></pre>
</div>
</div>
<p>Since <b style="color:#EB1960">Figue 5b</b> is dependent on knowing the <b style="color:#A6A440">tumor</b> and <b style="color:#A6A440">stromal</b> cells, which were deduced using <b style="color:#537FBF">tSNE</b> by <span class="citation" data-cites="ovarian-cancer-living-biobank">Nelson et al. (<a href="#ref-ovarian-cancer-living-biobank" role="doc-biblioref">2020</a>)</span>, we‚Äôll get set up with that as well.</p>
<div id="ac4ea876-f81d-47dd-9fcd-e2a0be03e73f" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># They only ran tSNE on the first 14 dimensions of PCA-transformed data.</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="co"># The `assay.type` defaults to "logcounts" but I wanted to be explicit</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># since it's not obvious that it would (although it is sensible that it does).</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fu">tryCatch</span>(</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    ovarian.sce <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">fixedPCA</span>(</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>        ovarian.sce,</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">rank=</span><span class="dv">14</span>,</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">subset.row=</span><span class="cn">NULL</span>,</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">assay.type=</span><span class="st">"logcounts"</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">error=</span><span class="cf">function</span>(e){<span class="fu">print</span>(e)}</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;simpleError in MatrixGenerics:::.load_next_suggested_package_to_search(x): Failed to find a colVars() method for dgTMatrix objects.&gt;</code></pre>
</div>
</div>
<p>This error is a nuisance, but it seems like it‚Äôs an issue with <code>dgTMatrix</code> being for <strong style="color:#C0CF96">dweebs</strong>. We should use <code>dgCMatrix</code> instead (<a href="https://support.bioconductor.org/p/100954/">thanks internet!</a>).</p>
<div id="18b9d8b0-e8d3-475a-b2e4-3fb836331a6c" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert assays to dgCMatrix</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(ovarian.sce, <span class="st">"counts"</span>) <span class="ot">&lt;-</span> <span class="fu">as</span>(</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">assay</span>(ovarian.sce, <span class="st">"counts"</span>),</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CsparseMatrix"</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(ovarian.sce, <span class="st">"logcounts"</span>) <span class="ot">&lt;-</span> <span class="fu">as</span>(</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">assay</span>(ovarian.sce, <span class="st">"logcounts"</span>),</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CsparseMatrix"</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="62697c3a-1cdd-491b-9dc2-482b1211dc14" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>ovarian.sce <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">fixedPCA</span>(</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>    ovarian.sce,</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">rank=</span><span class="dv">14</span>,</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">subset.row=</span><span class="cn">NULL</span>,</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">assay.type=</span><span class="st">"logcounts"</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="afc4e1b9-45e4-4993-a600-3f748382ff72" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>ovarian.sce <span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">runTSNE</span>(</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>    ovarian.sce,</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">dimred=</span><span class="st">"PCA"</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="7a98299b-aef8-4ee3-9193-06f249381326" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>ovarian.sce</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>class: SingleCellExperiment 
dim: 6454 18757 
metadata(0):
assays(2): counts logcounts
rownames(6454): ENSG00000000457 ENSG00000000460 ... ENSG00000288504
  ENSG00000288649
rowData names(0):
colnames(18757): SAMEA6492740-AAACCCACAGTTAGGG
  SAMEA6492740-AAACCCACATGTGTCA ... SAMEA6492743-TTTGTTGGTCCTGGTG
  SAMEA6492743-TTTGTTGTCAGATTGC
colData names(2): patient sizeFactor
reducedDimNames(2): PCA TSNE
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
<div id="de225b58-fef5-450e-8a13-a460934b25ee" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(ovarian.sce, <span class="at">dimred=</span><span class="st">"TSNE"</span>, <span class="at">colour_by=</span><span class="st">"patient"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="012_Cancer3_files/figure-html/cell-18-output-1.png" width="420" height="420" class="figure-img"></p>
</figure>
</div>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<p>This looks much better than last post‚Ä¶ Which probably means I made a mistake with the normalization somewhere there üò¨</p>
</div></div></section>
<section id="identifying-stromals" class="level1">
<h1>Identifying <b style="color:#A6A440">stromals</b></h1>
<p>The paper claims that the <b style="color:#A6A440">tumor</b> cells are the ones in the single-patient clusters and the <b style="color:#A6A440">stromal</b> cells are the ones in the multi-patient cluster. To partition them ourself, we need a clustering method! We‚Äôll do <code>scran::clusterCells</code> because <code>scran</code> hasn‚Äôt failed us yet.</p>
<div id="0d19c4ec-6f03-4831-8e21-85910ccc8381" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb25"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>ovarian.sce<span class="sc">$</span>tumor.stromal.clusters <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">clusterCells</span>(</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>    ovarian.sce,</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.dimred=</span><span class="st">"TSNE"</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="408a0659-9984-4640-a2b8-7abed7ac0cf9" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>    ovarian.sce,</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"TSNE"</span>,</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">colour_by=</span><span class="st">"tumor.stromal.clusters"</span>,</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">shape_by=</span><span class="st">"patient"</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="012_Cancer3_files/figure-html/cell-20-output-1.png" width="420" height="420" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Annoyingly, this is way too many clusters! So we‚Äôll have to tune the parameters a bit.</p>
<div id="c66119aa-2fb0-4d50-8cad-669cf6328958" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This library is needed to specify graph parameters</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(bluster)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="25b6113e-91e1-4e83-83c6-0cc7a80ff068" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>ovarian.sce<span class="sc">$</span>tumor.stromal.clusters <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">clusterCells</span>(</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>    ovarian.sce,</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.dimred=</span><span class="st">"TSNE"</span>,</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">BLUSPARAM=</span>bluster<span class="sc">::</span><span class="fu">KmeansParam</span>(<span class="at">centers=</span><span class="dv">5</span>)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>    ovarian.sce,</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"TSNE"</span>,</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">colour_by=</span><span class="st">"tumor.stromal.clusters"</span>,</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">shape_by=</span><span class="st">"patient"</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="012_Cancer3_files/figure-html/cell-22-output-1.png" width="420" height="420" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>There we go! It seems that either Cluster 1 or 5 is the cluster we want to remove - it‚Äôs hard for me to tell because I‚Äôm colorblind<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>. So let‚Äôs see how heterogenous the candidate clusters are:</p>
<div id="08f73c07-3e58-4e88-82f3-27d380b3c160" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb29"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(ovarian.sce[,ovarian.sce<span class="sc">$</span>tumor.stromal.clusters<span class="sc">==</span><span class="dv">1</span>]<span class="sc">$</span>patient)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(ovarian.sce[,ovarian.sce<span class="sc">$</span>tumor.stromal.clusters<span class="sc">==</span><span class="dv">5</span>]<span class="sc">$</span>patient)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>'38b'</li><li>'59'</li><li>'74-1'</li><li>'79'</li></ol>
</div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>'38b'</li><li>'79'</li></ol>
</div>
</div>
<p>It seems clear that <b style="color:#C0CF96">Cluster 1</b> is our heterogenous cluster, and hence represents our <b style="color:#A6A440">stromal cells</b>.</p>
<div id="abd800fa-2258-4a61-90df-14360cfba3d3" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>stromal.index <span class="ot">&lt;-</span> ovarian.sce<span class="sc">$</span>tumor.stromal.clusters<span class="sc">==</span><span class="dv">1</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="fig-5b" class="level1">
<h1><b style="color:#EB1960">Fig 5b</b></h1>
<p>To make <b style="color:#EB1960">Fig 5b</b>, we just need to now know the relative proportions of <b style="color:#A6A440">tumor</b> and <b style="color:#A6A440">stromal</b> cells on a per-patient level.</p>
<div id="cc7eb972-860e-4dae-be76-b5001ec2e466" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb31"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>patient<span class="fl">.38</span>b.index <span class="ot">&lt;-</span> ovarian.sce<span class="sc">$</span>patient<span class="sc">==</span><span class="st">'38b'</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>patient.<span class="fl">59.</span>index <span class="ot">&lt;-</span> ovarian.sce<span class="sc">$</span>patient<span class="sc">==</span><span class="st">'59'</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>patient<span class="fl">.74</span>_1.index <span class="ot">&lt;-</span> ovarian.sce<span class="sc">$</span>patient<span class="sc">==</span><span class="st">'74-1'</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>patient.<span class="fl">79.</span>index <span class="ot">&lt;-</span> ovarian.sce<span class="sc">$</span>patient<span class="sc">==</span><span class="st">'79'</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="ae2f6cef-a8e7-4042-8f88-53c1345ec4d6" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get raw amount for each patient</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>stromal.amount <span class="ot">&lt;-</span> <span class="fu">colSums</span>(<span class="fu">data.frame</span>(</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">patient.38b=</span>patient<span class="fl">.38</span>b.index <span class="sc">&amp;</span> stromal.index,</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">patient.59=</span>patient.<span class="fl">59.</span>index <span class="sc">&amp;</span> stromal.index,</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">patient.74_1=</span>patient<span class="fl">.74</span>_1.index <span class="sc">&amp;</span> stromal.index,</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">patient.79=</span>patient.<span class="fl">79.</span>index <span class="sc">&amp;</span> stromal.index</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>))</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>tumor.amount <span class="ot">&lt;-</span> <span class="fu">colSums</span>(<span class="fu">data.frame</span>(</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">patient.38b=</span>patient<span class="fl">.38</span>b.index <span class="sc">&amp;</span> <span class="sc">!</span>stromal.index,</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">patient.59=</span>patient.<span class="fl">59.</span>index <span class="sc">&amp;</span> <span class="sc">!</span>stromal.index,</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">patient.74_1=</span>patient<span class="fl">.74</span>_1.index <span class="sc">&amp;</span> <span class="sc">!</span>stromal.index,</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">patient.79=</span>patient.<span class="fl">79.</span>index <span class="sc">&amp;</span> <span class="sc">!</span>stromal.index</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="06c3d17c-aaea-4daa-8aec-1dacf59b7540" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb33"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the percents</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>tumor.stromal.data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(stromal.amount, tumor.amount)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>tumor.stromal.data<span class="sc">$</span>patient <span class="ot">&lt;-</span> <span class="fu">rownames</span>(tumor.stromal.data)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>tumor.stromal.data<span class="sc">$</span>stromal.percent <span class="ot">&lt;-</span> (</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>    tumor.stromal.data<span class="sc">$</span>stromal.amount <span class="sc">/</span> (</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>        tumor.stromal.data<span class="sc">$</span>stromal.amount <span class="sc">+</span> tumor.stromal.data<span class="sc">$</span>tumor.amount</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>tumor.stromal.data<span class="sc">$</span>tumor.percent <span class="ot">&lt;-</span> (</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>    tumor.stromal.data<span class="sc">$</span>tumor.amount <span class="sc">/</span> (</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>        tumor.stromal.data<span class="sc">$</span>stromal.amount <span class="sc">+</span> tumor.stromal.data<span class="sc">$</span>tumor.amount</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>tumor.stromal.data</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small">
<caption>A data.frame: 4 √ó 5</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">stromal.amount</th>
<th data-quarto-table-cell-role="th" scope="col">tumor.amount</th>
<th data-quarto-table-cell-role="th" scope="col">patient</th>
<th data-quarto-table-cell-role="th" scope="col">stromal.percent</th>
<th data-quarto-table-cell-role="th" scope="col">tumor.percent</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">patient.38b</th>
<td>1230</td>
<td>3976</td>
<td>patient.38b</td>
<td>0.2362658</td>
<td>0.7637342</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">patient.59</th>
<td>693</td>
<td>3003</td>
<td>patient.59</td>
<td>0.1875000</td>
<td>0.8125000</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">patient.74_1</th>
<td>1854</td>
<td>4111</td>
<td>patient.74_1</td>
<td>0.3108131</td>
<td>0.6891869</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">patient.79</th>
<td>849</td>
<td>3041</td>
<td>patient.79</td>
<td>0.2182519</td>
<td>0.7817481</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="a9d89620-e3fe-4061-99a0-ad1b9f9cd286" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Format it to work in `geom_dotplot`</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>tumor.stromal.data.for.dotplot <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">percent=</span><span class="fu">as.integer</span>(</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">c</span>(tumor.stromal.data<span class="sc">$</span>stromal.percent, tumor.stromal.data<span class="sc">$</span>tumor.percent) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">tumor.stromal=</span><span class="fu">c</span>(<span class="st">"S"</span>, <span class="st">"S"</span>, <span class="st">"S"</span>, <span class="st">"S"</span>, <span class="st">"T"</span>, <span class="st">"T"</span>, <span class="st">"T"</span>, <span class="st">"T"</span>)</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>tumor.stromal.data.for.dotplot</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small">
<caption>A data.frame: 8 √ó 2</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th" scope="col">percent</th>
<th data-quarto-table-cell-role="th" scope="col">tumor.stromal</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>23</td>
<td>S</td>
</tr>
<tr class="even">
<td>18</td>
<td>S</td>
</tr>
<tr class="odd">
<td>31</td>
<td>S</td>
</tr>
<tr class="even">
<td>21</td>
<td>S</td>
</tr>
<tr class="odd">
<td>76</td>
<td>T</td>
</tr>
<tr class="even">
<td>81</td>
<td>T</td>
</tr>
<tr class="odd">
<td>68</td>
<td>T</td>
</tr>
<tr class="even">
<td>78</td>
<td>T</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="f323e45a-6e5e-491b-a9c4-e5aeaaa1da62" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb35"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(tumor.stromal.data.for.dotplot, <span class="fu">aes</span>(<span class="at">x=</span>tumor.stromal, <span class="at">y=</span>percent)) <span class="sc">+</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_dotplot</span>(<span class="at">binaxis=</span><span class="st">'y'</span>, <span class="at">stackdir=</span><span class="st">'center'</span>) <span class="sc">+</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">100</span>)) <span class="sc">+</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stat_summary</span>(</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">fun=</span>mean,</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">geom=</span><span class="st">"crossbar"</span>,</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">size=</span><span class="fl">0.5</span>,</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">width=</span><span class="fl">0.3</span>,</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">color=</span><span class="st">"red"</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">"Stromal vs Tumor"</span>) <span class="sc">+</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Cell Count (%)"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<div class="ansi-escaped-output">
<pre>Bin width defaults to 1/30 of the range of the data. Pick better value with `binwidth`.
</pre>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="012_Cancer3_files/figure-html/cell-29-output-2.png" width="420" height="420" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="fig-5c" class="level1">
<h1><b style="color:#EB1960">Fig 5c</b></h1>
<blockquote class="blockquote">
<p><b style="color:#EB1960">Fig 5c)</b> <b style="color:#537FBF">t-SNE plot</b> from a overlaid with <b style="color:#A6A440">XIST expression</b></p>
<p>Moreover, cells in two of the distinct clusters did not express <b style="color:#A6A440">XIST</b> (<b style="color:#EB1960">Fig. 5c</b>), consistent with loss of the <b style="color:#A6A440">inactive X chromosome</b>.</p>
<p>‚Äì What <span class="citation" data-cites="ovarian-cancer-living-biobank">Nelson et al. (<a href="#ref-ovarian-cancer-living-biobank" role="doc-biblioref">2020</a>)</span> has to say about <b style="color:#EB1960">Fig 5c</b></p>
</blockquote>
<p><a href="https://www.ensembl.org/Homo_sapiens/Gene/Ontologies/biological_process?g=ENSG00000229807;r=X:73820649-73852723">We can see</a> that <b style="color:#A6A440">XIST</b> is also known as <b style="color:#A6A440">ENSG00000229807</b>.</p>
<div id="4d4a9a8e-d368-4107-9047-ae332a6092a4" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb36"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">rownames</span>(ovarian.sce) <span class="sc">==</span> <span class="st">"ENSG00000229807"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
FALSE
</div>
</div>
<p>But, uh, it‚Äôs not in our dataset. I even checked to see if we filtered it out, but its not there. This is a fairly major issue.</p>
<p>Some replacement genes:</p>
<ul>
<li><b style="color:#A6A440">JPX</b>: <b style="color:#A6A440">ENSG00000225470</b></li>
<li><b style="color:#A6A440">FTX</b>: <b style="color:#A6A440">ENSG00000230590</b></li>
<li><b style="color:#A6A440">TSIX</b>: <b style="color:#A6A440">ENSG00000270641</b></li>
<li><b style="color:#A6A440">EPCAM</b>: <b style="color:#A6A440">ENSG00000119888</b></li>
<li><b style="color:#A6A440">TP53</b>: <b style="color:#A6A440">ENSG00000141510</b></li>
<li><b style="color:#A6A440">MYC</b>: <b style="color:#A6A440">ENSG00000136997</b></li>
</ul>
<p>I found some of these by picking apart this quote referring to <b style="color:#EB1960">Fig 4</b>:</p>
<blockquote class="blockquote">
<p>This classification was supported by interrogating specific genes, with the tumour cells expressing <b style="color:#A6A440">EPCAM</b>, <b style="color:#A6A440">TP53</b> and <b style="color:#A6A440">MYC</b> but negative for <b style="color:#A6A440">XIST</b> and <b style="color:#A6A440">TSIX</b>, consistent with loss of the inactive X chromosome (<b style="color:#EB1960">Fig. 4c</b>)</p>
</blockquote>
<p>And the rest from <a href="https://en.wikipedia.org/wiki/XIST">wikipedia</a>.</p>
<div id="a5f077bb-d0f7-4438-b409-c217062c8ea5" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb37"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">rownames</span>(ovarian.sce) <span class="sc">==</span> <span class="st">"ENSG00000225470"</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">rownames</span>(ovarian.sce) <span class="sc">==</span> <span class="st">"ENSG00000230590"</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">rownames</span>(ovarian.sce) <span class="sc">==</span> <span class="st">"ENSG00000270641"</span>)</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">rownames</span>(ovarian.sce) <span class="sc">==</span> <span class="st">"ENSG00000119888"</span>)</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">rownames</span>(ovarian.sce) <span class="sc">==</span> <span class="st">"ENSG00000141510"</span>)</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">rownames</span>(ovarian.sce) <span class="sc">==</span> <span class="st">"ENSG00000136997"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
FALSE
</div>
<div class="cell-output cell-output-display">
FALSE
</div>
<div class="cell-output cell-output-display">
FALSE
</div>
<div class="cell-output cell-output-display">
FALSE
</div>
<div class="cell-output cell-output-display">
TRUE
</div>
<div class="cell-output cell-output-display">
FALSE
</div>
</div>
<p>So it seems like, of the data the genes they claim to have been looking at, only one of them is actually in their dataset. I‚Äôm sure I‚Äôm missing something here, but there‚Äôs not much I can do about it. The description of <b style="color:#A6A440">TP53</b> is ‚Äútumor protein p53‚Äù according to <b style="color:#EB1960">Ensembl</b>, so it certainly seems relevant!</p>
<div id="fcb64d1c-e5d5-4ff5-a3de-96a4b62297e3" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb38"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Removed all zero counts to prevent clutter</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>    ovarian.sce[,<span class="fu">which</span>(<span class="fu">assay</span>(ovarian.sce, <span class="st">"counts"</span>)[<span class="st">"ENSG00000141510"</span>,] <span class="sc">&gt;</span> <span class="dv">0</span>)],</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"TSNE"</span>,</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">colour_by=</span><span class="st">"ENSG00000141510"</span>,</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">shape_by=</span><span class="st">"patient"</span></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="012_Cancer3_files/figure-html/cell-32-output-1.png" width="420" height="420" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can see nothing informative here, it seems.</p>
<p>Maybe we can salvage a pretty graph by looking at the famous <b style="color:#A6A440">BRCA1</b> (<b style="color:#A6A440">ENSG00000012048</b>) gene?</p>
<div id="3a02b5d6-b4f1-44de-b36b-0c5be4f49446" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb39"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">rownames</span>(ovarian.sce) <span class="sc">==</span> <span class="st">"ENSG00000012048"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
FALSE
</div>
</div>
<p>No<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>.</p>
</section>
<section id="going-offscript" class="level1">
<h1>Going Offscript</h1>
<p>The fact that <b style="color:#A6A440">XIST</b> does not exist is quite annoying. Let‚Äôs find another marker gene to get a pretty plot!</p>
<div id="1aecafc5-efb2-41a2-9fd9-87c55d396564" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>ovarian.sce</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>class: SingleCellExperiment 
dim: 6454 18757 
metadata(0):
assays(2): counts logcounts
rownames(6454): ENSG00000000457 ENSG00000000460 ... ENSG00000288504
  ENSG00000288649
rowData names(0):
colnames(18757): SAMEA6492740-AAACCCACAGTTAGGG
  SAMEA6492740-AAACCCACATGTGTCA ... SAMEA6492743-TTTGTTGGTCCTGGTG
  SAMEA6492743-TTTGTTGTCAGATTGC
colData names(3): patient sizeFactor tumor.stromal.clusters
reducedDimNames(2): PCA TSNE
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
<div id="195b9cd5-3fc5-4261-85dc-3c081f896f3e" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>marker.info <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">scoreMarkers</span>(ovarian.sce, ovarian.sce<span class="sc">$</span>tumor.stromal.clusters)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="5a23b1dc-48f6-402f-8f27-c0914d8075a1" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb43"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We should pick a marker for cluster 1,</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="co"># because its the only stromal cell cluster</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="co"># so we should expect it to have clearly differentially expressed genes.</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>chosen <span class="ot">&lt;-</span> marker.info[[<span class="st">"1"</span>]]</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ovarian.sce)[<span class="fu">order</span>(chosen<span class="sc">$</span>mean.AUC, <span class="at">decreasing=</span><span class="cn">TRUE</span>)[[<span class="dv">1</span>]]]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
'ENSG00000106366'
</div>
</div>
<p><b style="color:#A6A440">ENSG00000106366</b> is also known as <b style="color:#A6A440">SERPINE1</b>, which is kinda cool.</p>
<div id="c2d97fb1-26fc-459f-86af-57ba4ea94126" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb44"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For pretty display, let's label each cluster as either</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="co"># tumor or stromal</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>ovarian.sce<span class="sc">$</span>tumor.or.stromal <span class="ot">&lt;-</span> <span class="st">"Unspecified"</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>ovarian.sce<span class="sc">$</span>tumor.or.stromal[ovarian.sce<span class="sc">$</span>tumor.stromal.clusters <span class="sc">==</span> <span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">"Stromal Cell"</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>ovarian.sce<span class="sc">$</span>tumor.or.stromal[ovarian.sce<span class="sc">$</span>tumor.stromal.clusters <span class="sc">==</span> <span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="st">"Tumor Cell (1)"</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>ovarian.sce<span class="sc">$</span>tumor.or.stromal[ovarian.sce<span class="sc">$</span>tumor.stromal.clusters <span class="sc">==</span> <span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="st">"Tumor Cell (2)"</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>ovarian.sce<span class="sc">$</span>tumor.or.stromal[ovarian.sce<span class="sc">$</span>tumor.stromal.clusters <span class="sc">==</span> <span class="dv">4</span>] <span class="ot">&lt;-</span> <span class="st">"Tumor Cell (3)"</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>ovarian.sce<span class="sc">$</span>tumor.or.stromal[ovarian.sce<span class="sc">$</span>tumor.stromal.clusters <span class="sc">==</span> <span class="dv">5</span>] <span class="ot">&lt;-</span> <span class="st">"Tumor Cell (4)"</span></span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>ovarian.sce</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>class: SingleCellExperiment 
dim: 6454 18757 
metadata(0):
assays(2): counts logcounts
rownames(6454): ENSG00000000457 ENSG00000000460 ... ENSG00000288504
  ENSG00000288649
rowData names(0):
colnames(18757): SAMEA6492740-AAACCCACAGTTAGGG
  SAMEA6492740-AAACCCACATGTGTCA ... SAMEA6492743-TTTGTTGGTCCTGGTG
  SAMEA6492743-TTTGTTGTCAGATTGC
colData names(4): patient sizeFactor tumor.stromal.clusters
  tumor.or.stromal
reducedDimNames(2): PCA TSNE
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
<div id="4ddc0a20-c208-48f4-bbbd-14c600630af8" class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb46"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotReducedDim</span>(</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>    ovarian.sce,</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"TSNE"</span>,</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">colour_by=</span><span class="st">"ENSG00000106366"</span>,</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">shape_by=</span><span class="st">"patient"</span>,</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">text_by=</span><span class="st">"tumor.or.stromal"</span>,</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">text_colour=</span><span class="st">"cyan"</span></span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Ovarian Cancer Cells by SERPINE1 (ENSG00000106366) Expression"</span>) <span class="sc">+</span></span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">"darkblue"</span>),</span>
<span id="cb46-12"><a href="#cb46-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">"cyan"</span>)</span>
<span id="cb46-13"><a href="#cb46-13" aria-hidden="true" tabindex="-1"></a>    )</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="012_Cancer3_files/figure-html/cell-38-output-1.png" width="420" height="420" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>This is quite a strong result! Of course, it‚Äôs a result that‚Äôs already known:</p>
<blockquote class="blockquote">
<p><b style="color:#A6A440">SERPINE1</b> has been proven to be the most reliable biological and prognostic marker for a variety of cancers, including <b style="color:#A6A440">breast cancer</b>, <b style="color:#A6A440">ovarian cancer</b>, <b style="color:#A6A440">bladder cancer</b>, <b style="color:#A6A440">colon cancer</b>, <b style="color:#A6A440">kidney cancer</b> and <b style="color:#A6A440">non-small cell lung cancer</b>.</p>
<p>‚Äì <span class="citation" data-cites="SERPINE1">Chen et al. (<a href="#ref-SERPINE1" role="doc-biblioref">2022</a>)</span></p>
</blockquote>
<p>But still, it makes a nice plot.</p>
<p>My takeaways from these past two blog posts are:</p>
<ul>
<li>‚ÄúEvery cancer is unique‚Äù is not just a buzzphrase; the variation between cancers in different people is far greater than the variation between stromal cells between people</li>
<li><b style="color:#A6A440">SERPINE1</b> is a biomarker for cancer</li>
<li><code>SingleCellExperiment</code> is a nice way to group data</li>
<li><b style="color:#EB1960">Bioconductor</b> has amazing online resources and a nice suite of packages</li>
</ul>
<p>I quite liked going offscript at the end. I may delve further into <b style="color:#537FBF">differential expression</b> in the next blogpost, or I might take a look at <b style="color:#537FBF">16S rRNA sequencing</b>. Not sure.</p>



</section>


<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-SERPINE1" class="csl-entry" role="listitem">
Chen, S., Y. Li, Y. Zhu, J. Fei, L. Song, G. Sun, L. Guo, and X. Li. 2022. <span>‚Äú<span class="nocase"><span>S</span><span>E</span><span>R</span><span>P</span><span>I</span><span>N</span><span>E</span>1 <span>O</span>verexpression <span>P</span>romotes <span>M</span>alignant <span>P</span>rogression and <span>P</span>oor <span>P</span>rognosis of <span>G</span>astric <span>C</span>ancer</span>.‚Äù</span> <em>J Oncol</em> 2022: 2647825.
</div>
<div id="ref-ovarian-cancer-living-biobank" class="csl-entry" role="listitem">
Nelson, Louisa, Anthony Tighe, Anya Golder, Samantha Littler, Bjorn Bakker, Daniela Moralli, Syed Murtuza Baker, et al. 2020. <span>‚ÄúA Living Biobank of Ovarian Cancer Ex Vivo Models Reveals Profound Mitotic Heterogeneity.‚Äù</span> <em>Nature Communications</em> 11 (1): 822. <a href="https://doi.org/10.1038/s41467-020-14551-2">https://doi.org/10.1038/s41467-020-14551-2</a>.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>In the sense of ‚Äúfunctional programming‚Äù<a href="#fnref1" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn2"><p>Although it does allow for non-functional styles too<a href="#fnref2" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn3"><p>already loaded as a dependency of either <code>scran</code> or <code>scatter</code>, I have no idea which.<a href="#fnref3" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn4"><p>I should have thought of that before making the graph‚Ä¶<a href="#fnref4" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
<li id="fn5"><p>In their defence, <b style="color:#A6A440">BRCA1</b> is actually in the dataset, it just got filtered out before we got here.<a href="#fnref5" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script>
    // Remove background
    elements = document.querySelectorAll('[id=clean-collapse]').forEach(
        function(element) {
            element.classList.remove("callout-style-default");
        }
    )
    
    // Add arrow to front to signify dropdownyness
    arrow = "‚ñ∂";
    elements = document.querySelectorAll('.callout-caption-container').forEach(
        function(element) {
            element.innerHTML = arrow + element.innerHTML;
        }
    )
</script>
<script src="https://giscus.app/client.js" data-repo="baileyandrew/blog" data-repo-id="R_kgDOInJwKg" data-category="Announcements" data-category-id="DIC_kwDOInJwKs4CTGOQ" data-mapping="title" data-strict="0" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="dark_protanopia" data-lang="en" data-loading="lazy" crossorigin="anonymous" async="">
</script>
<!--
Assuming `_quarto.yml` has `navbar: pinned: true`,
this will make it so that the navbar scrolls with the rest of the website.
-->
<script>
    document.getElementById('quarto-header').classList.remove('fixed-top');
</script>
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "Óßã";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/BaileyAndrew\.github\.io\/Blog");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/BaileyAndrew/Blog/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>