<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.27">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Bailey Andrew">
<meta name="dcterms.date" content="2023-02-06">
<meta name="description" content="Life lesson: üêî ‚â† üêì">

<title>A tale of two chickens ‚Äì Lorem Ipsomething</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.min.js" integrity="sha384-ZvpUoO/+PpLXR1lu4jmpXWu80pZlYUAfxl5NsBMWOEPSjUn/6Z/hRTt8+pR6L4N2" crossorigin="anonymous"></script><script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-4d9afe2b8d18ee9fa5d0d57b5ed4214d.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-10892d65c249a5d8f93b275ef5a48de5.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdn.jsdelivr.net/npm/requirejs@2.3.6/require.min.js" integrity="sha384-c9c+LnTbwQ3aujuU7ULEPVvgLs+Fn6fJUvIGTsuu1ZcCf11fiEubah0ttpca4ntM sha384-6V1/AdqZRWk1KAlWbKBlGhN7VG4iE/yAZcO6NZPMF8od0vukrvr0tg4qY6NSrItx" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../styles.css">
<link rel="stylesheet" href="../../html_scripts/collapse.css">
<link rel="stylesheet" href="../../html_scripts/pretty_shortcuts.css">
<meta property="og:title" content="A tale of two chickens ‚Äì Lorem Ipsomething">
<meta property="og:description" content="Life lesson: üêî ‚â† üêì">
<meta property="og:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/a-tale-of-two-chickens.png">
<meta property="og:site_name" content="Lorem Ipsomething">
<meta property="og:image:height" content="156">
<meta property="og:image:width" content="160">
<meta name="twitter:title" content="A tale of two chickens ‚Äì Lorem Ipsomething">
<meta name="twitter:description" content="Life lesson: üêî ‚â† üêì">
<meta name="twitter:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/a-tale-of-two-chickens.png">
<meta name="twitter:image-height" content="156">
<meta name="twitter:image-width" content="160">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Lorem Ipsomething</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">A tale of two chickens</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Home</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../blog.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Blog</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../staticposts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Static Posts</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Puzzles</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../500_Puzzles/bnv.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Boulders in Valleys</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slideshowindex.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slideshows</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../otherblogs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Other People‚Äôs Blogs</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#but-why" id="toc-but-why" class="nav-link active" data-scroll-target="#but-why">But why?</a></li>
  <li><a href="#tracking-the-history-of-the-events" id="toc-tracking-the-history-of-the-events" class="nav-link" data-scroll-target="#tracking-the-history-of-the-events">Tracking the history of the events</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/BaileyAndrew/Blog/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">A tale of two chickens</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Work</div>
    <div class="quarto-category">Omics</div>
    <div class="quarto-category">Useful</div>
    <div class="quarto-category">Paper Replication</div>
  </div>
  </div>

<div>
  <div class="description">
    Life lesson: üêî ‚â† üêì
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Bailey Andrew </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">February 6, 2023</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p>Today we‚Äôll be looking at the paper ‚ÄúA single-cell transcriptomic atlas of the developing chicken limb‚Äù, by <span class="citation" data-cites="chicken-cells">Feregrino et al. (<a href="#ref-chicken-cells" role="doc-biblioref">2019</a>)</span> - specifically, we‚Äôll use <a href="https://www.ebi.ac.uk/gxa/sc/experiments/E-CURD-13/downloads">this dataset from the Single Cell Expression Atlas</a>. However, we‚Äôre not actually gonna read their paper, we‚Äôre just gonna do random stuff!</p>
<p><b class="sidequests">Spoilers</b>: We don‚Äôt get much work done, due to a mystery involving chicken breeds in Ensembl.</p>
<details>
<summary style="color:#C0CF96">
Loading the Data
</summary>
<div id="175d8473-f9a6-4cb2-8dd7-3b025f5a2c95" class="cell" data-execution_count="68">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">0</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleCellExperiment)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="a5f7215e-18b2-461f-bfc3-519c0a97ed23" class="cell" data-execution_count="41">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>file.path <span class="ot">=</span> <span class="st">'./localdata/Datasets/E-CURD-13/'</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>counts.mat <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">readMM</span>(</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>        file.path,</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">'E-CURD-13-quantification-raw-files/'</span>,</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>        <span class="st">'E-CURD-13.aggregated_filtered_counts.mtx'</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(counts.mat)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>13645</li><li>7688</li></ol>
</div>
</div>
<div id="9bc2f6b2-5328-4d94-99df-4789cfc77169" class="cell" data-execution_count="42">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>counts.rows <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>        file.path,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">'E-CURD-13-quantification-raw-files/'</span>,</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">'E-CURD-13.aggregated_filtered_counts.mtx_rows'</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">sep=</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span>,</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">header=</span><span class="cn">FALSE</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>)<span class="sc">$</span>V1</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>counts.cols <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>        file.path,</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>        <span class="st">'E-CURD-13-quantification-raw-files/'</span>,</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>        <span class="st">'E-CURD-13.aggregated_filtered_counts.mtx_cols'</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">sep=</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span>,</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">header=</span><span class="cn">FALSE</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>)<span class="sc">$</span>V1</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="ce5eb019-40a9-45e1-8ade-f14a16119317" class="cell" data-execution_count="53">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>meta.data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>        file.path,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">'ExpDesign-E-CURD-13.tsv'</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">sep=</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span>,</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">header=</span><span class="cn">TRUE</span>,</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">row.names=</span><span class="st">'Assay'</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(meta.data)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>'Sample.Characteristic.organism.'</li><li>'Sample.Characteristic.Ontology.Term.organism.'</li><li>'Sample.Characteristic.age.'</li><li>'Sample.Characteristic.Ontology.Term.age.'</li><li>'Sample.Characteristic.developmental.stage.'</li><li>'Sample.Characteristic.Ontology.Term.developmental.stage.'</li><li>'Sample.Characteristic.organism.part.'</li><li>'Sample.Characteristic.Ontology.Term.organism.part.'</li><li>'Sample.Characteristic.sampling.site.'</li><li>'Sample.Characteristic.Ontology.Term.sampling.site.'</li><li>'Sample.Characteristic.genotype.'</li><li>'Sample.Characteristic.Ontology.Term.genotype.'</li><li>'Sample.Characteristic.disease.'</li><li>'Sample.Characteristic.Ontology.Term.disease.'</li><li>'Factor.Value.replicate.'</li><li>'Factor.Value.Ontology.Term.replicate.'</li><li>'Factor.Value.inferred.cell.type...ontology.labels.'</li><li>'Factor.Value.Ontology.Term.inferred.cell.type...ontology.labels.'</li></ol>
</div>
</div>
<div id="09f78739-9980-4503-94a1-05f63bcd7222" class="cell" data-execution_count="44">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(counts.mat) <span class="ot">&lt;-</span> counts.rows</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(counts.mat) <span class="ot">&lt;-</span> counts.cols</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="40b43835-58af-4b19-8861-6adb1e50f455" class="cell" data-execution_count="52">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make sure the metadata is lined up with the count matrix</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">identical</span>(meta.data<span class="sc">$</span>Assay, <span class="fu">colnames</span>(counts.mat))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
TRUE
</div>
</div>
<div id="c395c6bb-50a1-4b39-810e-e125fa81d9a5" class="cell" data-execution_count="60">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">SingleCellExperiment</span>(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">assays=</span><span class="fu">list</span>(<span class="at">counts=</span>counts.mat),</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">colData=</span>meta.data</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="b3e12cbb-d1d2-4c7f-b0b7-926a38f3034b" class="cell" data-execution_count="61">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>sce</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<pre><code>class: SingleCellExperiment 
dim: 13645 7688 
metadata(0):
assays(1): counts
rownames(13645): ENSGALG00000000003 ENSGALG00000000011 ...
  ENSGALG00000055127 ENSGALG00000055132
rowData names(0):
colnames(7688): SAMN11526603-AAAAAAATTCAG SAMN11526603-AAAAACAAGTAG ...
  SAMN11526603-TTTTTTTGTGAG SAMN11526603-TTTTTTTTTTTT
colData names(18): Sample.Characteristic.organism.
  Sample.Characteristic.Ontology.Term.organism. ...
  Factor.Value.inferred.cell.type...ontology.labels.
  Factor.Value.Ontology.Term.inferred.cell.type...ontology.labels.
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
</details>
<p>I wanted to perform QC using mitochondrial genes. This should have been as easy as reusing my code for humans with a few tweaks to look for chicken genes instead. However, when I try to use <code>biomaRt</code> I get no result.</p>
<div id="dccc1d1d-2c94-40e9-9eaf-1c2760c95c86" class="cell" data-tags="[]" data-execution_count="86">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>mart <span class="ot">&lt;-</span> <span class="fu">useDataset</span>(<span class="st">"ggallus_gene_ensembl"</span>, <span class="fu">useMart</span>(<span class="st">"ensembl"</span>))</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>gene.to.symbol.map <span class="ot">&lt;-</span> <span class="fu">getBM</span>(</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">filters=</span><span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">attributes=</span><span class="fu">c</span>(</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>        <span class="st">"hgnc_symbol"</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">values=</span><span class="fu">rownames</span>(sce),</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">mart=</span>mart</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>gene.to.symbol.map</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small">
<caption>A data.frame: 0 √ó 2</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th" scope="col">ensembl_gene_id</th>
<th data-quarto-table-cell-role="th" scope="col">hgnc_symbol</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="col">&lt;lgl&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;lgl&gt;</th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>
</div>
<p>So I went directly to the <code>biomaRt</code> <a href="https://www.ensembl.org/biomart/martview/3d63e0dabdacc8dc33bd4bba9ed2130a">web interface</a>, and searched for a gene I knew was in the dataset: <code>ENSGALG00000000003</code></p>
<p><img src="./images/no-003.png" class="img-fluid"></p>
<p>However, you can see that in the top left corner it reports no results! If you look closely, you can see that the example search is with <code>ENSGALG00010000002</code>, which does have results - and in fact adding 10,000,000 to gene numbers tends to create a valid findable gene:</p>
<p><img src="./images/yes-003.png" class="img-fluid"></p>
<p>These genes don‚Äôt seem to be related to eachother though.</p>
<p>If we filter for mitochondrial genes:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/filters.png" class="img-fluid figure-img"></p>
<figcaption>Like this.</figcaption>
</figure>
</div>
<p>Then we do get results:</p>
<p><img src="./images/found-mitos.png" class="img-fluid"></p>
<p>Which all have ids greater than 10,000,000. This list is not comprehensive, I can find (off biomaRt) another mitochondrial gene, whose id is less than 10,000,000:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/a-missed-mito.png" class="img-fluid figure-img"></p>
<figcaption>ENSGALG00000031197</figcaption>
</figure>
</div>
<blockquote class="blockquote">
<p>The avian genome encodes the same set of genes (13 proteins, 2 rRNAs and 22 tRNAs) as do other vertebrate mitochondrial DNAs and is organized in a very similar economical fashion.</p>
<p>‚Äì ‚ÄúSequence and gene organization of the chicken mitochondrial genome: A novel gene order in higher vertebrates‚Äù, <span class="citation" data-cites="chicken-mitochondria">Desjardins and Morais (<a href="#ref-chicken-mitochondria" role="doc-biblioref">1990</a>)</span></p>
</blockquote>
<p>From the above, we can see that there should be 24 mt-rRNA and mt-tRNA in total, which is in fact what we find with biomaRt! (The previous image only shows 10, but that‚Äôs because it only shows 10 per page - the total amount found was 24).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/more-than-million.png" class="img-fluid figure-img"></p>
<figcaption>The entire mitochondrial genome when viewing it starting from a gene larger than 10,000,000.</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/less-than-million.png" class="img-fluid figure-img"></p>
<figcaption>The entire mitochondrial genome when viewing it starting from a gene less than 10,000,000.</figcaption>
</figure>
</div>
<p>These two images seem to conclusively show that the gene sets are different; the genes are in a similar place but have different names between the two tracks.</p>
<p>From this, we can hypothesize that the ids less than 10,000,000 are out of date, presenting some frustration for using the data we have. Can we prove it?</p>
<p><img src="./images/no-change.png" class="img-fluid"></p>
<p>No: the gene has never changed its ID (which is a good thing for sanity purposes but is not ideal for this specific confusion).</p>
<p>However, it seems like there are multiple chicken genomes in Ensembl:</p>
<p><img src="./images/chickens.png" class="img-fluid"></p>
<p>This proves, conclusively, that there are at least two breeds of chicken. And I was looking at the wrong one.</p>
<p>The question then becomes - how do we actually find the right one? We can get a list of all the possible biomart-queryable datasets:</p>
<div id="ef0501c5-7952-4e1c-bb96-87a25b4f7d12" class="cell" data-execution_count="113">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>all.datasets <span class="ot">&lt;-</span> <span class="fu">listDatasets</span>(<span class="at">mart =</span> mart)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>all.datasets[<span class="dv">70</span><span class="sc">:</span><span class="dv">80</span>,]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small">
<caption>A data.frame: 11 √ó 3</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">dataset</th>
<th data-quarto-table-cell-role="th" scope="col">description</th>
<th data-quarto-table-cell-role="th" scope="col">version</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;I&lt;chr&gt;&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;I&lt;chr&gt;&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;I&lt;chr&gt;&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">70</th>
<td>fheteroclitus_gene_ensembl</td>
<td>Mummichog genes (Fundulus_heteroclitus-3.0.2)</td>
<td>Fundulus_heteroclitus-3.0.2</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">71</th>
<td>gaculeatus_gene_ensembl</td>
<td>Stickleback genes (BROAD S1)</td>
<td>BROAD S1</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">72</th>
<td>gevgoodei_gene_ensembl</td>
<td>Goodes thornscrub tortoise genes (rGopEvg1_v1.p)</td>
<td>rGopEvg1_v1.p</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">73</th>
<td>gfortis_gene_ensembl</td>
<td>Medium ground-finch genes (GeoFor_1.0)</td>
<td>GeoFor_1.0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">74</th>
<td>ggallus_gene_ensembl</td>
<td>Chicken genes (bGalGal1.mat.broiler.GRCg7b)</td>
<td>bGalGal1.mat.broiler.GRCg7b</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">75</th>
<td>ggorilla_gene_ensembl</td>
<td>Gorilla genes (gorGor4)</td>
<td>gorGor4</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">76</th>
<td>gmorhua_gene_ensembl</td>
<td>Atlantic cod genes (gadMor3.0)</td>
<td>gadMor3.0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">77</th>
<td>hburtoni_gene_ensembl</td>
<td>Burton's mouthbrooder genes (AstBur1.0)</td>
<td>AstBur1.0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">78</th>
<td>hcomes_gene_ensembl</td>
<td>Tiger tail seahorse genes (H_comes_QL1_v1)</td>
<td>H_comes_QL1_v1</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">79</th>
<td>hgfemale_gene_ensembl</td>
<td>Naked mole-rat female genes (HetGla_female_1.0)</td>
<td>HetGla_female_1.0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">80</th>
<td>hhucho_gene_ensembl</td>
<td>Huchen genes (ASM331708v1)</td>
<td>ASM331708v1</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>But this does not contain the option to choose anything other than <code>bGalGal1.mat.broiler.GRCg7b</code>, the paternal White leghorn layer chicken.</p>
<p>Online, we can find an archived version of <code>biomaRt</code> that grants access:</p>
<p><img src="./images/archived-chicken.png" class="img-fluid"></p>
<p>And the results line up with prior expectations.</p>
<p>So if we use the <code>biomaRt</code> R interface with an archived host, maybe we can get the Red Junglefowl‚Äôs genes?</p>
<div id="2cf6f608-6ad0-42ad-be23-b344155beade" class="cell" data-execution_count="116">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>mart.archive <span class="ot">&lt;-</span> <span class="fu">useMart</span>(<span class="st">"ensembl"</span>, <span class="at">host=</span><span class="st">"https://apr2022.archive.ensembl.org"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="339aa9d8-17bd-4d93-b3e4-1351057b2a15" class="cell" data-execution_count="117">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>all.datasets <span class="ot">&lt;-</span> <span class="fu">listDatasets</span>(<span class="at">mart =</span> mart.archive)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>all.datasets[<span class="dv">70</span><span class="sc">:</span><span class="dv">80</span>,]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small">
<caption>A data.frame: 11 √ó 3</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">dataset</th>
<th data-quarto-table-cell-role="th" scope="col">description</th>
<th data-quarto-table-cell-role="th" scope="col">version</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;I&lt;chr&gt;&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;I&lt;chr&gt;&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;I&lt;chr&gt;&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">70</th>
<td>fheteroclitus_gene_ensembl</td>
<td>Mummichog genes (Fundulus_heteroclitus-3.0.2)</td>
<td>Fundulus_heteroclitus-3.0.2</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">71</th>
<td>gaculeatus_gene_ensembl</td>
<td>Stickleback genes (BROAD S1)</td>
<td>BROAD S1</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">72</th>
<td>gevgoodei_gene_ensembl</td>
<td>Goodes thornscrub tortoise genes (rGopEvg1_v1.p)</td>
<td>rGopEvg1_v1.p</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">73</th>
<td>gfortis_gene_ensembl</td>
<td>Medium ground-finch genes (GeoFor_1.0)</td>
<td>GeoFor_1.0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">74</th>
<td>ggallus_gene_ensembl</td>
<td>Chicken genes (GRCg6a)</td>
<td>GRCg6a</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">75</th>
<td>ggorilla_gene_ensembl</td>
<td>Gorilla genes (gorGor4)</td>
<td>gorGor4</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">76</th>
<td>gmorhua_gene_ensembl</td>
<td>Atlantic cod genes (gadMor3.0)</td>
<td>gadMor3.0</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">77</th>
<td>hburtoni_gene_ensembl</td>
<td>Burton's mouthbrooder genes (AstBur1.0)</td>
<td>AstBur1.0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">78</th>
<td>hcomes_gene_ensembl</td>
<td>Tiger tail seahorse genes (H_comes_QL1_v1)</td>
<td>H_comes_QL1_v1</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">79</th>
<td>hgfemale_gene_ensembl</td>
<td>Naked mole-rat female genes (HetGla_female_1.0)</td>
<td>HetGla_female_1.0</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">80</th>
<td>hhucho_gene_ensembl</td>
<td>Huchen genes (ASM331708v1)</td>
<td>ASM331708v1</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Victory! <code>GRCg6a</code> is the Red Junglefowl.</p>
<div id="f343a5e5-df1e-4bed-84ca-9292450b23e9" class="cell" data-execution_count="120">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>mart.archive <span class="ot">&lt;-</span> <span class="fu">useDataset</span>(</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ggallus_gene_ensembl"</span>,</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">useMart</span>(</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">"ensembl"</span>,</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">host=</span><span class="st">"https://apr2022.archive.ensembl.org"</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>gene.to.symbol.map <span class="ot">&lt;-</span> <span class="fu">getBM</span>(</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">filters=</span><span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">attributes=</span><span class="fu">c</span>(</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>        <span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>        <span class="st">"hgnc_symbol"</span></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">values=</span><span class="fu">rownames</span>(sce),</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">mart=</span>mart.archive</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<div id="84c52197-6366-4d63-8280-42b12bfeced2" class="cell" data-execution_count="121">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(gene.to.symbol.map)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small">
<caption>A data.frame: 6 √ó 2</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">ensembl_gene_id</th>
<th data-quarto-table-cell-role="th" scope="col">hgnc_symbol</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">1</th>
<td>ENSGALG00000000003</td>
<td>PANX2</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">2</th>
<td>ENSGALG00000000011</td>
<td>C10orf88</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">3</th>
<td>ENSGALG00000000038</td>
<td></td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">4</th>
<td>ENSGALG00000000044</td>
<td>WFIKKN1</td>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">5</th>
<td>ENSGALG00000000055</td>
<td></td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">6</th>
<td>ENSGALG00000000059</td>
<td></td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Double victory!</p>
<div id="dedc0bbe-a4c0-40e4-abfc-5d787561ddcb" class="cell" data-execution_count="133">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode R code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>gene.to.symbol.map[</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">apply</span>(</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>        gene.to.symbol.map[<span class="st">"hgnc_symbol"</span>],</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>        <span class="dv">1</span>,</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>        <span class="cf">function</span>(x) <span class="fu">grepl</span>(<span class="st">"^MT-"</span>, x)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-display">
<table class="dataframe caption-top table table-sm table-striped small">
<caption>A data.frame: 2 √ó 2</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">ensembl_gene_id</th>
<th data-quarto-table-cell-role="th" scope="col">hgnc_symbol</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="row">9151</th>
<td>ENSGALG00000032142</td>
<td>MT-CO1</td>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th" scope="row">11684</th>
<td>ENSGALG00000043768</td>
<td>MT-ND2</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>All that for just two genes. Oh well. We‚Äôll analyze this dataset in a future blog post.</p>
<p><b style="color:#EB1960">As a learning experience, I would recommend always specifying a specific Ensembl version when querying <code>biomaRt</code></b>, since evidently code that used to work will break if they change the default breed.</p>
<p>Also, it‚Äôs a bit frustrating that I can‚Äôt specify the breed in <code>biomaRt</code>, because Ensembl still has the data for the other breed - it‚Äôs just for some reason not accessible via <code>biomaRt</code> which only presents a single breed to query.</p>
<section id="but-why" class="level2">
<h2 class="anchored" data-anchor-id="but-why">But why?</h2>
<p>I‚Äôm left with one question: why are there two chickens? Why is the junglefowl no longer on modern BioMart? What is their relation? (Ahhhhhhhhhhhh)</p>
<p>I did manage to find a <a href="https://www.ed.ac.uk/roslin/national-avian-research-facility/avian-resources/poultry-lines/white-leghorn">summary of some chicken breeds</a>, but it did not resolve my question.</p>
<p>To investigate further, I went through the history of the Ensembl versions:</p>
<ul>
<li><a href="http://oct2022.archive.ensembl.org/Gallus_gallus/Info/Index">Version 108</a> (current, October 2022): Default chicken is <code>bGalGal1.mat.broiler.GRCg7b</code>
<ul>
<li>But when we look at the <a href="http://oct2022.archive.ensembl.org/Gallus_gallus/Info/Breeds">chicken breeds page</a> they list it as <code>bGalGal1.pat.whiteleghornlayer.GRCg7w</code></li>
</ul></li>
<li><a href="http://jul2022.archive.ensembl.org/Gallus_gallus/Info/Index">Version 107</a> (July 2022): Default chicken is <code>bGalGal1.mat.broiler.GRCg7b</code>
<ul>
<li>Although in this version they call it the ‚Äúmaternal broiler‚Äù</li>
<li>And when we look at the <a href="http://jul2022.archive.ensembl.org/Gallus_gallus/Info/Breeds">chicken breeds page</a> they list it as <code>bGalGal1.pat.whiteleghornlayer.GRCg7w</code></li>
</ul></li>
<li><a href="http://apr2022.archive.ensembl.org/Gallus_gallus/Info/Index">Version 106</a> (April 2022): Default chicken is <code>GRCg6a</code>
<ul>
<li>And there is no chicken breeds page</li>
</ul></li>
</ul>
<p>It seems like the chicken breeds pages don‚Äôt list the default breed on it, and that all of the genes for <code>GRCg7w</code> are listed with IDs beginning with 00015 (<a href="https://www.ensembl.org/Gallus_gallus_GCA_016700215.2/Gene/Summary?db=core;g=ENSGALG00015014615;r=1:140692888-140816960;t=ENSGALT00015035601">such as here</a>). This is in contrast with the 00010 beginning for <code>GRCg7b</code> and <code>GRCg6a</code>.</p>
<p><b style="color:#EB1960">Thus, we can say this:</b></p>
<ul>
<li>There are three chickens in play<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>, <code>GRCg6a</code> (‚Äúred junglefowl‚Äù), <code>GRCg7w</code> (‚Äúwhite leghorn‚Äù), and <code>GRCg7b</code> (‚Äúmaternal broiler‚Äù).</li>
<li><code>GRCg7b</code> (‚Äúmaternal broiler‚Äù) is the current default</li>
<li><code>GRCg6a</code> (‚Äúred junglefowl‚Äù) used to be the default back in version 106</li>
</ul>
<p>Looking up ‚Äúmaternal broiler‚Äù, I am able to find <a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_016699485.2/">this page</a> talking about the <code>bGalGal1.mat.broiler.GRCg7b</code> genome. It lists its breed as specifically being:</p>
<blockquote class="blockquote">
<p>Breed: Cross of Broiler mother + white leghorn layer father</p>
</blockquote>
<p>This only deepens the mystery. Why did we go from Red Junglefowl to Maternal Broiler? How does the White Leghorn come into the picture, and is it related to why <code>GRCg7w</code> is a crossbreed?</p>
<blockquote class="blockquote">
<p>The chicken has been the most studied bird. It serves as a model organism for basic and biomedical science, and is one of the most commonly consumed food sources for humans.</p>
<p>This trio of samples was collected to create a high-quality reference genome for two haplotypes.</p>
<p><b style="color:#537FBF">The female offspring (<code>bGalGal1</code>) is a cross between a modern broiler breeder mother (<code>bGalGal2</code>) and white Leghorn father (<code>bGalGal3</code>).</b></p>
<p>Breeding and sample collection was conducted by Nick Anthony at the University of Arkansas, in coordination with Wesley Warren, Ron Okimoto, Hans Cheng, and Rachel Hawken.</p>
<p>Blood from these samples were used to create high-quality reference genome assemblies of each haplotype, coordinated by Erich Jarvis, Wesley Warren, and Olivier Fedrigo, at the Rockefeller University Vertebrate Genome Lab, as part of the Vertebrate Genomes Project (VGP) and the Genome Reference Consortium (GRC).</p>
<p>‚Äì <a href="https://www.ncbi.nlm.nih.gov/data-hub/genome/GCF_016699485.2/">Webpage</a> for <code>GRCg7b</code></p>
</blockquote>
<p>The webpage identifies <code>GRCg7b</code> as coming from <code>bGalGal3</code> and <code>GRCg7w</code> as coming from <code>bGalGal2</code>.</p>
<p>It seems <code>bGalGal1</code> does not have its own assembly, as <a href="https://www.ncbi.nlm.nih.gov/data-hub/genome/?taxon=9031">the list of chicken assemblies</a> does not include it. The most information I can get as to <code>bGalGal1</code> is <a href="https://www.ncbi.nlm.nih.gov/biosample/14342210">on its sample page</a>, which is significantly less detailed than <a href="https://www.ncbi.nlm.nih.gov/biosample/SAMN15960293/">its parent‚Äôs</a>. As an aside, it seems like a lot of new chicken assemblies were made in 2022, which may be interesting for the future.</p>
<p>According to <a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_016700215.1/">the page</a> for <code>bGalGal3</code>, the <code>GRCg7w</code> assembly has been updated to <code>GRCg7w_WZ</code>. I believe this did not require getting more real-world data, but rather was a re-sequencing (re-assembling) of the collected data.</p>
</section>
<section id="tracking-the-history-of-the-events" class="level2">
<h2 class="anchored" data-anchor-id="tracking-the-history-of-the-events">Tracking the history of the events</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/chicken-history.png" class="img-fluid figure-img"></p>
<figcaption>This shows the history of the assembly chosen for the chicken, procured from <a href="https://www.ensembl.org/info/website/archives/assembly.html">here</a></figcaption>
</figure>
</div>
<p>On the Ensembl blog, we can get a changelog of what has happened:</p>
<blockquote class="blockquote">
<h3 class="anchored">
New Assemblies and/or Annotation
</h3>
<h4 class="anchored">
Vertebrates
</h4>
<ul>
<li>Reannotation of the reference assembly for pig (<code>Sscrofa11.1</code>)</li>
<li><b style="color:#C0CF96">Reannotation of chicken assembly</b> (<code>GRCg6a</code>)</li>
<li><b style="color:#C0CF96">Annotation of 2 new GRC chicken assemblies</b> (<code>GRCg7w</code> <b style="color:#C0CF96">and</b> <code>GRCg7b</code>)</li>
<li><b style="color:#C0CF96">Chicken reference will be changed from</b> <code>GRCg6a</code> <b style="color:#C0CF96">to</b> <code>GRCg7b</code></li>
<li>Assembly and Gene set will be updated for Tropical clawed frog (Xenopus tropicalis): Xenopus tropicalis v9.1 to UCB Xtro 10.0</li>
</ul>
<p>‚Äì <a href="https://www.ensembl.info/2022/05/31/whats-coming-in-ensembl-release-107-ensembl-genomes-54/">Release notes for Ensembl 107</a></p>
</blockquote>
<p>All in all, this has been a really confusing adventure, but I <em>think</em> I learned something‚Ä¶</p>



</section>


<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-chicken-mitochondria" class="csl-entry" role="listitem">
Desjardins, Paul, and R√©jean Morais. 1990. <span>‚ÄúSequence and Gene Organization of the Chicken Mitochondrial Genome: A Novel Gene Order in Higher Vertebrates.‚Äù</span> <em>Journal of Molecular Biology</em> 212 (4): 599‚Äì634. https://doi.org/<a href="https://doi.org/10.1016/0022-2836(90)90225-B">https://doi.org/10.1016/0022-2836(90)90225-B</a>.
</div>
<div id="ref-chicken-cells" class="csl-entry" role="listitem">
Feregrino, C, F Sacher, Parnas O, and P Tschopp. 2019. <span>‚ÄúA Single-Cell Transcriptomic Atlas of the Developing Chicken Limb.‚Äù</span> <em>BMC Genomics</em>.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>Despite my earlier belief in this being a two-chicken conundrum!<a href="#fnref1" class="footnote-back" role="doc-backlink">‚Ü©Ô∏é</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script>
    // Remove background
    elements = document.querySelectorAll('[id=clean-collapse]').forEach(
        function(element) {
            element.classList.remove("callout-style-default");
        }
    )
    
    // Add arrow to front to signify dropdownyness
    arrow = "‚ñ∂";
    elements = document.querySelectorAll('.callout-caption-container').forEach(
        function(element) {
            element.innerHTML = arrow + element.innerHTML;
        }
    )
</script>
<script src="https://giscus.app/client.js" data-repo="baileyandrew/blog" data-repo-id="R_kgDOInJwKg" data-category="Announcements" data-category-id="DIC_kwDOInJwKs4CTGOQ" data-mapping="title" data-strict="0" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="dark_protanopia" data-lang="en" data-loading="lazy" crossorigin="anonymous" async="">
</script>
<!--
Assuming `_quarto.yml` has `navbar: pinned: true`,
this will make it so that the navbar scrolls with the rest of the website.
-->
<script>
    document.getElementById('quarto-header').classList.remove('fixed-top');
</script>
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "Óßã";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/BaileyAndrew\.github\.io\/Blog");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/BaileyAndrew/Blog/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>