<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.27">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Bailey Andrew">
<meta name="dcterms.date" content="2023-01-06">
<meta name="description" content="A high-level overview.">

<title>Overview of scRNA-seq Data Creation – Sequencing – Lorem Ipsomething</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark-4d9afe2b8d18ee9fa5d0d57b5ed4214d.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-10892d65c249a5d8f93b275ef5a48de5.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
<meta property="og:title" content="Overview of scRNA-seq Data Creation – Sequencing – Lorem Ipsomething">
<meta property="og:description" content="A high-level overview.">
<meta property="og:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/flow_cell.jpeg">
<meta property="og:site_name" content="Lorem Ipsomething">
<meta name="twitter:title" content="Overview of scRNA-seq Data Creation – Sequencing – Lorem Ipsomething">
<meta name="twitter:description" content="A high-level overview.">
<meta name="twitter:image" content="https://BaileyAndrew.github.io/Blog/000_Blog/004_scRNA/images/flow_cell.jpeg">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Lorem Ipsomething</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Overview of scRNA-seq Data Creation – Sequencing</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Home</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../blog.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Blog</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../staticposts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Static Posts</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Puzzles</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../500_Puzzles/bnv.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Boulders in Valleys</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slideshowindex.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slideshows</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../otherblogs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Other People’s Blogs</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#scrna-seq-data" id="toc-scrna-seq-data" class="nav-link active" data-scroll-target="#scrna-seq-data">scRNA-seq Data</a>
  <ul class="collapse">
  <li><a href="#library-preparation" id="toc-library-preparation" class="nav-link" data-scroll-target="#library-preparation">Library Preparation</a></li>
  <li><a href="#sequencing" id="toc-sequencing" class="nav-link" data-scroll-target="#sequencing">Sequencing</a>
  <ul class="collapse">
  <li><a href="#flow-cells" id="toc-flow-cells" class="nav-link" data-scroll-target="#flow-cells">Flow Cells</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/BaileyAndrew/Blog/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Overview of scRNA-seq Data Creation – Sequencing</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Work</div>
    <div class="quarto-category">Useful</div>
    <div class="quarto-category">Omics</div>
  </div>
  </div>

<div>
  <div class="description">
    A high-level overview.
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Bailey Andrew </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 6, 2023</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<section id="scrna-seq-data" class="level1">
<h1>scRNA-seq Data</h1>
<p>I’m heavily indebted to Morgan for help in understanding this topic. I’m still learning about it, so don’t take my word as gospel.</p>
<section id="library-preparation" class="level2">
<h2 class="anchored" data-anchor-id="library-preparation">Library Preparation</h2>
<p><strong style="color:#537FBF">Library preparation</strong> is the collection of the bits of data that you’re actually interested in (these bits would be your ‘library’). The methods I’ve seen do this by attaching something to the mRNA molecules that we can take advantage of later. For scRNA-seq we are interested in the <strong style="color:#A6A440">transcriptome</strong> (this is the set of RNA thingamajigs in the cell). <strong style="color:#EB1960"><span class="citation" data-cites="illumina-ebook"><em>Single-Cell Sequencing Workflow: Critical Steps and Considerations</em> (<a href="#ref-illumina-ebook" role="doc-biblioref">2019</a>)</span></strong> lists a few library preparation techniques for the transcriptome:</p>
<hr>
<ul>
<li>Full-length RNA-seq</li>
<li>mRNA end-tag amplification</li>
<li>Targeted panels
<ul>
<li>Targeted towards measuring specific things</li>
</ul></li>
<li>IR-seq
<ul>
<li>Specifically used for B and T cells.</li>
</ul></li>
</ul>
<hr>
<p>These are important for me to understand so I’ll definitely be delving into this topic deeper in the future.</p>
</section>
<section id="sequencing" class="level2">
<h2 class="anchored" data-anchor-id="sequencing">Sequencing</h2>
<p>We’ve now isolated our cells and grabbed/tagged the bits that we’re interested in. Now, we want to actually be able to learn about the mRNA that we’ve measured. To do this, we need to sequence it.</p>
<section id="flow-cells" class="level3">
<h3 class="anchored" data-anchor-id="flow-cells">Flow Cells</h3>
<p>One important concept for sequencing is the <strong style="color:#A6A440">Flow Cell</strong>. Intuitively, the flow cell is a sticky plate that will grab the mRNA. <strong style="color:#EB1960"><span class="citation" data-cites="illumina-ngs"><em>An Introduction to Next-Generation Sequencing Technology</em> (<a href="#ref-illumina-ngs" role="doc-biblioref">2017</a>)</span></strong> defines them as the following:</p>
<blockquote class="blockquote">
<p><strong style="color:#A6A440">Flow Cell</strong>: A glass slide with one, two, or eight physically separated lanes, depending on the instrument platform. Each lane is coated with a lawn of surface bound, adapter-complimentary <strong style="color:#A6A440">oligos</strong>. A single library or a pool of up to 96 <strong style="color:#537FBF">multiplexed</strong> libraries can be run per lane, depending on application parameters.</p>
<p>– <cite><strong style="color:#EB1960">Illumina handbook on next generation sequencing</strong></cite> <span class="citation" data-cites="illumina-ngs">(<a href="#ref-illumina-ngs" role="doc-biblioref"><em>An Introduction to Next-Generation Sequencing Technology</em> 2017</a>)</span></p>
</blockquote>
<p>An <strong style="color:#A6A440">oligo</strong> is a short strand of synthetic DNA <span class="citation" data-cites="oligo">(<a href="#ref-oligo" role="doc-biblioref"><span>“What Is an Oligo?”</span> 2019</a>)</span>. These are the “sticky bits”. <strong style="color:#537FBF">Multiplexing</strong> is the process of sequencing multiple samples at a time; this can be useful as sequencing produces a lot of data, more than may be necessary for a single project on its own <span class="citation" data-cites="ngs-considerations">(<a href="#ref-ngs-considerations" role="doc-biblioref"><span>“NGS Considerations: Coverage, Read Length, Multiplexing,”</span> n.d.</a>)</span>. This can be done by attaching molecular barcodes to the prepared cDNA to indicate the original sample from which the cDNA hails.</p>
<section id="sequencing-by-synthesis" class="level4">
<h4 class="anchored" data-anchor-id="sequencing-by-synthesis">Sequencing by Synthesis</h4>
<p>The flow cell definition given by Illumina is specifically tuned to their flow cells - this may have been obvious due to the use of specific numbers like “96”.</p>
<details>
<summary style="color:#c0cf95">
<strong>Tangential Observation</strong>
</summary>
Surprisingly, though, the competitor company <strong style="color:#EB1960">Oxford Nanopore</strong> <a href="https://store.nanoporetech.com/uk/rapid-barcoding-kit-1.html">also uses this number</a>. I assume this is due to some chemical feature of the bardcodes, but it probably isn’t as simple as that since I know you can have more than 96 UMIs which I thought would be a similar process…
</details>
<p><br> <strong style="color:#EB1960">Illumina</strong> uses a <strong style="color:#537FBF">Sequencing by Synthesis</strong> approach; we’ll look at another approach, <strong style="color:#537FBF">Nanopore Sequencing</strong>, later. During <strong style="color:#EB1960">Illumina’s</strong> library preparation phase they convert the RNA to cDNA.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/flow_cell.jpeg" class="img-fluid figure-img"></p>
<figcaption>Example of a patterned flow cell</figcaption>
</figure>
</div>
<p>The above diagram is that of a 2D slice of a <strong style="color:#A6A440">Patterned Flow Cell</strong>. These differ from <strong style="color:#A6A440">Nonpatterned Flow Cells</strong> in the use of <strong style="color:#A6A440">nanowells</strong> (little valleys in the cell) to ideally keep fragments of DNA from binding to nearby oligos. Patterned flow cells are a more recent innovation, and for a brief overview of the differences you can read <span class="citation" data-cites="patterned-vs-nonpatterned"><span>“Calculating Percent Passing Filter for Patterned and Nonpatterned Flow Cells”</span> (<a href="#ref-patterned-vs-nonpatterned" role="doc-biblioref">2017</a>)</span> and <span class="citation" data-cites="pvn-cons">Wingett (<a href="#ref-pvn-cons" role="doc-biblioref">2017</a>)</span>.</p>
<p>The cDNA created during library preparation is added onto the flow cell. The strands of cDNA (<strong style="color:#A6A440">reads</strong>) will bind to the oligos, and then start multiplying (<strong style="color:#537FBF">“bridge amplification”</strong>) so that duplicates of the read will be bound to nearby oligos. For patterned flow cells, this should mean that the nanowells are full of copies of the same read - for unpatterned flow cells the clusters are more chaotic.</p>
<p>After you’ve prepared your flow cell, you’ll plug it into a sequencer to read the base pairs. These use chemical tricks to cause each base pair in a read to let out a flash of a specific color. The sequence and color of these flashes will inform the sequencer of what the base pair sequence is. After doing this, you’ll have raw data on the contents of each read, likely in the <code>bcl</code> file format (<code>bcl</code> is the output for <strong style="color:#EB1960">Illumina</strong> sequencers).</p>
</section>
<section id="nanopore-sequencing" class="level4">
<h4 class="anchored" data-anchor-id="nanopore-sequencing">Nanopore Sequencing</h4>
<p>While most of this blog has focused on <strong style="color:#EB1960">Illumina</strong>, <strong style="color:#EB1960">Oxford Nanopore</strong> is definitely worth talking about. They use <strong style="color:#537FBF">Nanopore Sequencing</strong> instead of <strong style="color:#537FBF">Sequencing by Synthesis</strong>. A brief comparison of the two companines is given by <span class="citation" data-cites="illumina-vs-nanopore"><span>“NGS Platforms: Illumina Vs Oxford Nanopore”</span> (<a href="#ref-illumina-vs-nanopore" role="doc-biblioref">2022</a>)</span>.</p>
<details>
<summary style="color:#C0CF95">
<strong>¿cDNA?</strong>
</summary>
It is possible to directly use RNA, instead of cDNA, depending on what is done at the library preparation stage. I’ll assume we use cDNA here. It would make an interesting blog post, I think, to delve into the intricacies as to what is going on and why we should care about cDNA vs RNA.
</details>
<p><br></p>
<p>Nanopore sequencing is a newer method; instead of sticking to an oligo, the cDNA passes through a <strong style="color:#A6A440">nanopore</strong>. To do this, the cDNA has a <strong style="color:#A6A440">motor protein</strong> attached to it during the library preparation stage. The act of passing through the nanopore creates a detectable electrical signal that is dependent on the bases in the cDNA. By measuring this signal, we can sequence the molecule.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/nanopore.jpeg" class="img-fluid figure-img"></p>
<figcaption>Nanopore sequencing. The cDNA will pass through the nanopore as it gets sequenced. The tether will grab on to the motor protein to accomplish this. <span class="citation" data-cites="nanopore">(<a href="#ref-nanopore" role="doc-biblioref"><span>“Product Brochure”</span> 2022</a>)</span></figcaption>
</figure>
</div>
<p>One interesting advantage of nanopore sequencing is the ability to detect <strong style="color:#A6A440">modified bases</strong> (which I didn’t know were a thing!) <span class="citation" data-cites="modified-bases">(<a href="#ref-modified-bases" role="doc-biblioref"><span>“Delivering the Future of Genomic Pathogen Surveillance”</span> 2022</a>)</span>. Basically, the structure that makes up your G, T, A, C, U nucleic acid bases can actually get modified in certain ways, turning them into different sub-molecules that may affect gene expression.</p>
<p><strong style="color:#EB1960">Oxford Nanopore</strong> outputs <code>FAST5</code> files, a type of <code>HDF5</code> file. These contain the raw electrical signals the nanopores measured. One can then perform <strong style="color:#537FBF">basecalling</strong> to determine the sequence of bases corresponding to the signals. Due to the aforementioned detectability of modified bases, these raw electrical signals can have value rather than just looking at an end product of everything converted to G/T/A/C/U <span class="citation" data-cites="fast5">(<a href="#ref-fast5" role="doc-biblioref">Su 2019</a>)</span>.</p>
</section>
<section id="measures-and-options" class="level4">
<h4 class="anchored" data-anchor-id="measures-and-options">Measures and Options</h4>
<p>Flow cells have two key measures (besides data quality); the amount of reads and the <strong style="color:#A6A440">read length</strong>. Read counts can be in the hundreds of millions. Read lengths may be much smaller; the flow cells paired with the <strong style="color:#EB1960">Illumina NextSeq 550</strong> can only have reads as long as 150 base pairs on average. <strong style="color:#EB1960">Oxford Nanopore</strong> (and in general any nanopore method) can have unbounded read lengths. Longer reads are desirable from the perspective of reconstructing larger sequences, as it is easier to determine if two segments have a significant overlap.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/depth.jpeg" class="img-fluid figure-img"></p>
<figcaption>My theory as to why it’s called “depth”</figcaption>
</figure>
</div>
<p>An important concept is <strong style="color:#A6A440">coverage</strong>, or the amount of times a gene has a read associated with it. Higher coverage is better as it allows us to piece together the gene more accurately and weed out incorrect bases. This is also called <strong style="color:#A6A440">sequence depth</strong>. Deep sequencing is important for detecting rare genes. This can be affected by the steps done during library preparation; immune cells, which undergo <strong style="color:#537FBF">VDJ recombination</strong> - in short, each cell has its own unique markers that need to be accounted for <span class="citation" data-cites="ngs-considerations">(<a href="#ref-ngs-considerations" role="doc-biblioref"><span>“NGS Considerations: Coverage, Read Length, Multiplexing,”</span> n.d.</a>)</span></p>
<p>For single cell applications depth refers to the amount of reads per cell instead of the reads per base pair measure used for bulk sequencing. When cell populations are more homogenous, the depth should be larger as false overlaps will be more likely.</p>
<p>Finally, there’s the decision of whether to use <strong style="color:#A6A440">paired-end</strong> or <strong style="color:#A6A440">single-end</strong> sequencing. DNA and RNA are linear structures with two different ends - the 5’ and 3’ ends (“5-prime” and “3-prime”). By reading from both the ends, we can better find errors as well as increase the read length (intuitively: instead of reading n base pairs in one direction, you can read 2n total pairs with n in each direction). These advantages decrease when you have <strong style="color:#A6A440">UMIs (unique molecular identifiers)</strong> or other feature you can take advantage of (such as VDJs), but that is out of scope for this blog post. <span class="citation" data-cites="ngs-considerations">(<a href="#ref-ngs-considerations" role="doc-biblioref"><span>“NGS Considerations: Coverage, Read Length, Multiplexing,”</span> n.d.</a>)</span> The advantages are not completely nullified, as single-end sequencing cannot detect certain types of errors (such as “indel” errors). <span class="citation" data-cites="illumina-ebook">(<a href="#ref-illumina-ebook" role="doc-biblioref"><em>Single-Cell Sequencing Workflow: Critical Steps and Considerations</em> 2019</a>)</span></p>
</section>
</section>
<section id="references" class="level3">
<h3 class="anchored" data-anchor-id="references">References</h3>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-illumina-ngs" class="csl-entry" role="listitem">
<em>An Introduction to Next-Generation Sequencing Technology</em>. 2017. Illumina. <a href="https://emea.illumina.com/content/dam/illumina-marketing/documents/products/illumina_sequencing_introduction.pdf">https://emea.illumina.com/content/dam/illumina-marketing/documents/products/illumina_sequencing_introduction.pdf</a>.
</div>
<div id="ref-patterned-vs-nonpatterned" class="csl-entry" role="listitem">
<span>“Calculating Percent Passing Filter for Patterned and Nonpatterned Flow Cells.”</span> 2017. Illumina. <a href="https://emea.illumina.com/content/dam/illumina-marketing/documents/products/technotes/hiseq-x-percent-pf-technical-note-770-2014-043.pdf">https://emea.illumina.com/content/dam/illumina-marketing/documents/products/technotes/hiseq-x-percent-pf-technical-note-770-2014-043.pdf</a>.
</div>
<div id="ref-modified-bases" class="csl-entry" role="listitem">
<span>“Delivering the Future of Genomic Pathogen Surveillance.”</span> 2022. Oxford Nanopore. <a href="https://nanoporetech.com/resource-centre/white-paper/future-of-genomic-pathogen-surveillance">https://nanoporetech.com/resource-centre/white-paper/future-of-genomic-pathogen-surveillance</a>.
</div>
<div id="ref-ngs-considerations" class="csl-entry" role="listitem">
<span>“NGS Considerations: Coverage, Read Length, Multiplexing.”</span> n.d. iRepertoire. <a href="https://irepertoire.com/ngs-considerations-coverage-read-length-multiplexing/">https://irepertoire.com/ngs-considerations-coverage-read-length-multiplexing/</a>.
</div>
<div id="ref-illumina-vs-nanopore" class="csl-entry" role="listitem">
<span>“NGS Platforms: Illumina Vs Oxford Nanopore.”</span> 2022. Fios Genomics. <a href="https://www.fiosgenomics.com/ngs-platforms-illumina-vs-oxford-nanopore/">https://www.fiosgenomics.com/ngs-platforms-illumina-vs-oxford-nanopore/</a>.
</div>
<div id="ref-nanopore" class="csl-entry" role="listitem">
<span>“Product Brochure.”</span> 2022. Oxford Nanopore. <a href="https://nanoporetech.com/sites/default/files/s3/literature/product-brochure.pdf">https://nanoporetech.com/sites/default/files/s3/literature/product-brochure.pdf</a>.
</div>
<div id="ref-illumina-ebook" class="csl-entry" role="listitem">
<em>Single-Cell Sequencing Workflow: Critical Steps and Considerations</em>. 2019. Illumina. <a href="https://emea.illumina.com/destination/multiomics-transcriptomics.html">https://emea.illumina.com/destination/multiomics-transcriptomics.html</a>.
</div>
<div id="ref-fast5" class="csl-entry" role="listitem">
Su, Shian. 2019. <span>“A Look at the Nanopore Fast5 Format,”</span> May. <a href="https://medium.com/@shiansu/a-look-at-the-nanopore-fast5-format-f711999e2ff6">https://medium.com/@shiansu/a-look-at-the-nanopore-fast5-format-f711999e2ff6</a>.
</div>
<div id="ref-oligo" class="csl-entry" role="listitem">
<span>“What Is an Oligo?”</span> 2019. Thermo Fisher Scientific; Behind the Bench. <a href="https://www.thermofisher.com/blog/behindthebench/what-is-an-oligo/">https://www.thermofisher.com/blog/behindthebench/what-is-an-oligo/</a>.
</div>
<div id="ref-pvn-cons" class="csl-entry" role="listitem">
Wingett, Steven. 2017. <span>“Illumina Patterned Flow Cells Generate Duplicated Sequences.”</span> <a href="https://sequencing.qcfail.com/articles/illumina-patterned-flow-cells-generate-duplicated-sequences/">https://sequencing.qcfail.com/articles/illumina-patterned-flow-cells-generate-duplicated-sequences/</a>.
</div>
</div>
<script src="https://giscus.app/client.js" data-repo="baileyandrew/blog" data-repo-id="R_kgDOInJwKg" data-category="Announcements" data-category-id="DIC_kwDOInJwKs4CTGOQ" data-mapping="title" data-strict="0" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="dark_protanopia" data-lang="en" data-loading="lazy" crossorigin="anonymous" async="">
</script>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/BaileyAndrew\.github\.io\/Blog");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/BaileyAndrew/Blog/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>